{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e5608d8",
   "metadata": {},
   "source": [
    "# Import Gene Annotation File"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fb9c920",
   "metadata": {},
   "source": [
    "https://github.com/hakha-most/gwas_eqtl/blob/master/gene_annotations/genes.protein_coding.v39.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b64193a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>cons</th>\n",
       "      <th>gene</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621096</td>\n",
       "      <td>622034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>622034</td>\n",
       "      <td>621096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859303</td>\n",
       "      <td>879955</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC:28706</td>\n",
       "      <td>859303</td>\n",
       "      <td>879955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879583</td>\n",
       "      <td>894689</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>HGNC:24517</td>\n",
       "      <td>894689</td>\n",
       "      <td>879583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>895964</td>\n",
       "      <td>901099</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>HGNC:24023</td>\n",
       "      <td>895964</td>\n",
       "      <td>901099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end strand  ...    gene     hgnc_id     tss     tes\n",
       "0  chr1   65419   71585      +  ...   OR4F5  HGNC:14825   65419   71585\n",
       "1  chr1  621096  622034      -  ...  OR4F16  HGNC:15079  622034  621096\n",
       "2  chr1  859303  879955      +  ...  SAMD11  HGNC:28706  859303  879955\n",
       "3  chr1  879583  894689      -  ...   NOC2L  HGNC:24517  894689  879583\n",
       "4  chr1  895964  901099      +  ...  KLHL17  HGNC:24023  895964  901099\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "\n",
    "# Local path to GTF file\n",
    "gtf_file_path = 'genes.protein_coding.v39.gtf'\n",
    "\n",
    "# Load the GTF file into a pandas DataFrame and use the first row as headers\n",
    "gtf_df = pd.read_csv(gtf_file_path, sep='\\t', header=0)\n",
    "\n",
    "# Display the first few rows of the dataframe\n",
    "gtf_df.sort_values('start')\n",
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68c645d",
   "metadata": {},
   "source": [
    "#### convert chr to int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7fd4032",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>65419</td>\n",
       "      <td>249200395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>242836139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>238446</td>\n",
       "      <td>197687071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>53180</td>\n",
       "      <td>190945522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>92283</td>\n",
       "      <td>180794269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>291630</td>\n",
       "      <td>170884383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>330136</td>\n",
       "      <td>158820866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>116086</td>\n",
       "      <td>146277851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>14475</td>\n",
       "      <td>140772234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>92828</td>\n",
       "      <td>135437399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>167784</td>\n",
       "      <td>134248398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>175933</td>\n",
       "      <td>133794898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>19747912</td>\n",
       "      <td>115079957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>19377594</td>\n",
       "      <td>105992920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>20737094</td>\n",
       "      <td>102462245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>96407</td>\n",
       "      <td>90122974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>5810</td>\n",
       "      <td>81037485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>47221</td>\n",
       "      <td>77915115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>107104</td>\n",
       "      <td>59073298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>67891</td>\n",
       "      <td>62887048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>14982098</td>\n",
       "      <td>48055507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>17071641</td>\n",
       "      <td>51205929</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr       min        max\n",
       "0     1     65419  249200395\n",
       "1     2     38814  242836139\n",
       "2     3    238446  197687071\n",
       "3     4     53180  190945522\n",
       "4     5     92283  180794269\n",
       "5     6    291630  170884383\n",
       "6     7    330136  158820866\n",
       "7     8    116086  146277851\n",
       "8     9     14475  140772234\n",
       "9    10     92828  135437399\n",
       "10   11    167784  134248398\n",
       "11   12    175933  133794898\n",
       "12   13  19747912  115079957\n",
       "13   14  19377594  105992920\n",
       "14   15  20737094  102462245\n",
       "15   16     96407   90122974\n",
       "16   17      5810   81037485\n",
       "17   18     47221   77915115\n",
       "18   19    107104   59073298\n",
       "19   20     67891   62887048\n",
       "20   21  14982098   48055507\n",
       "21   22  17071641   51205929"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df['chr'] = gtf_df['chr'].str[3:].astype(int)\n",
    "gtf_df['start'] = gtf_df['start'].astype(int)\n",
    "gtf_df.groupby('chr')['start'].agg({'min', 'max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e19dcc",
   "metadata": {},
   "source": [
    "Positions reset for each chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77808ff",
   "metadata": {},
   "source": [
    "### File Description\n",
    "\n",
    "This file contains **gene annotations** for **protein-coding genes** and provides information about the **location of genes on the genome**. Below is a description of each column:\n",
    "\n",
    "1. **chr**: The chromosome where the gene is located (e.g., `chr1`, `chr2`, etc.).\n",
    "2. **start**: The start position of the gene on the chromosome (in base pairs).\n",
    "3. **end**: The end position of the gene on the chromosome (in base pairs).\n",
    "4. **strand**: Indicates the strand on which the gene is located (`+` for forward strand, `-` for reverse strand).\n",
    "5. **GeneSymbol**: The symbol or name of the gene (e.g., `OR4F5`, `SAMD11`), typically assigned by organizations like HGNC (HUGO Gene Nomenclature Committee).\n",
    "6. **cons**: This column indicates the type of gene. In this file, all genes are classified as **protein_coding**.\n",
    "7. **gene**: The Ensembl gene ID, a unique identifier for the gene (e.g., `ENSG00000186092`).\n",
    "8. **hgnc_id**: The unique identifier for the gene assigned by the **HGNC** (HUGO Gene Nomenclature Committee).\n",
    "9. **tss**: The **transcription start site**, the position where transcription of the gene starts on the chromosome.\n",
    "10. **tes**: The **transcription end site**, the position where transcription of the gene ends on the chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e70a8e",
   "metadata": {},
   "source": [
    "# Import GWAS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7500998",
   "metadata": {},
   "source": [
    " https://docs.google.com/spreadsheets/d/1kvPoupSzsSFBNSztMzl04xMoSC3Kcx3CrjVf4yBmESU/edit?gid=178908679#gid=178908679 (row 7217)\n",
    " \n",
    " 50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "072a09d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to GWAS file\n",
    "gwas_file_path = '50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz'\n",
    "\n",
    "# Open and read the compressed .bgz file using gzip\n",
    "with gzip.open(gwas_file_path, 'rt') as f:\n",
    "    # Load the file into a pandas DataFrame\n",
    "    gwas_df = pd.read_csv(f, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ab8c875a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.440760e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003474</td>\n",
       "      <td>18.049900</td>\n",
       "      <td>178.468000</td>\n",
       "      <td>0.101138</td>\n",
       "      <td>0.919441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.761760e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>-0.087536</td>\n",
       "      <td>-0.041345</td>\n",
       "      <td>0.349577</td>\n",
       "      <td>-0.118272</td>\n",
       "      <td>0.905852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.881580e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>135.620000</td>\n",
       "      <td>-2.079970</td>\n",
       "      <td>-0.046994</td>\n",
       "      <td>0.062695</td>\n",
       "      <td>-0.749571</td>\n",
       "      <td>0.453514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.674710e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>-0.105659</td>\n",
       "      <td>-0.042056</td>\n",
       "      <td>0.349588</td>\n",
       "      <td>-0.120302</td>\n",
       "      <td>0.904244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106010e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>79718.500000</td>\n",
       "      <td>101.503000</td>\n",
       "      <td>0.000797</td>\n",
       "      <td>0.002899</td>\n",
       "      <td>0.274788</td>\n",
       "      <td>0.783479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>2.454380e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>176906.000000</td>\n",
       "      <td>-787.373000</td>\n",
       "      <td>-0.001657</td>\n",
       "      <td>0.001588</td>\n",
       "      <td>-1.044020</td>\n",
       "      <td>0.296477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.297000e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>165562.000000</td>\n",
       "      <td>-734.113000</td>\n",
       "      <td>-0.001654</td>\n",
       "      <td>0.001657</td>\n",
       "      <td>-0.998454</td>\n",
       "      <td>0.318060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.393990e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>172553.000000</td>\n",
       "      <td>-480.000000</td>\n",
       "      <td>-0.001689</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>-1.011490</td>\n",
       "      <td>0.311783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>2.458510e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>177204.000000</td>\n",
       "      <td>-706.111000</td>\n",
       "      <td>-0.001520</td>\n",
       "      <td>0.001587</td>\n",
       "      <td>-0.957369</td>\n",
       "      <td>0.338381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.955520e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>14094.900000</td>\n",
       "      <td>-265.942000</td>\n",
       "      <td>-0.013297</td>\n",
       "      <td>0.005446</td>\n",
       "      <td>-2.441660</td>\n",
       "      <td>0.014620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    variant minor_allele  ...     tstat      pval\n",
       "0               1:15791:C:T            T  ...  0.101138  0.919441\n",
       "1               1:69487:G:A            A  ... -0.118272  0.905852\n",
       "2               1:69569:T:C            C  ... -0.749571  0.453514\n",
       "3              1:139853:C:T            T  ... -0.120302  0.904244\n",
       "4             1:692794:CA:C            C  ...  0.274788  0.783479\n",
       "...                     ...          ...  ...       ...       ...\n",
       "13791462    X:154929412:C:T            T  ... -1.044020  0.296477\n",
       "13791463   X:154929637:CT:C            C  ... -0.998454  0.318060\n",
       "13791464  X:154929952:CAA:C            C  ... -1.011490  0.311783\n",
       "13791465    X:154930230:A:G            G  ... -0.957369  0.338381\n",
       "13791466    X:154930487:T:A            A  ... -2.441660  0.014620\n",
       "\n",
       "[13791467 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first few rows to inspect the structure of the GWAS file\n",
    "gwas_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1418c7f4",
   "metadata": {},
   "source": [
    "### GWAS Summary Statistics Dataset Description\n",
    "\n",
    "1. **variant**: The unique identifier for each SNP (Single Nucleotide Polymorphism). This can include information like chromosome, position, reference allele, and alternative allele (e.g., `1:12345:A:G`).\n",
    "   - **Example**: `1:10583:T:G`\n",
    "\n",
    "2. **minor_allele**: The allele that is less frequent in the population (minor allele) for this particular SNP.\n",
    "   - **Example**: `G`\n",
    "\n",
    "3. **minor_AF**: The **minor allele frequency** (AF), which represents the frequency of the minor allele in the population. It ranges from 0 to 1.\n",
    "   - **Example**: `0.35` (35% of individuals carry the minor allele)\n",
    "\n",
    "4. **low_confidence_variant**: A flag indicating whether the variant has **low confidence** due to imputation quality or other uncertainties. Values may be `TRUE` or `FALSE`.\n",
    "   - **Example**: `FALSE`\n",
    "\n",
    "5. **n_complete_samples**: The number of samples for which complete genotype data is available for this variant.\n",
    "   - **Example**: `300,000`\n",
    "\n",
    "6. **AC**: The **allele count** of the minor allele, i.e., the number of times the minor allele appears in the study population (across all samples).\n",
    "   - **Example**: `50000`\n",
    "\n",
    "7. **ytx**: Likely a placeholder for a phenotype-related statistic; depending on the dataset, this could represent something like the trait mean or effect size (its exact meaning depends on the specific analysis).\n",
    "\n",
    "8. **beta**: The **effect size** of the SNP on the trait being studied (in this case, likely height). It represents the change in the trait per additional copy of the minor allele.\n",
    "   - **Example**: `0.05` (the trait increases by 0.05 units for each additional copy of the minor allele)\n",
    "\n",
    "9. **se**: The **standard error** of the effect size (beta), indicating the precision of the estimated effect.\n",
    "   - **Example**: `0.01`\n",
    "\n",
    "10. **tstat**: The **t-statistic** for the beta estimate, which is the ratio of the beta estimate to its standard error.\n",
    "    - **Example**: `5.0` (higher values indicate more significant associations)\n",
    "\n",
    "11. **pval**: The **p-value** of the association between the SNP and the trait. This indicates the significance of the result, with smaller p-values suggesting stronger evidence that the SNP is associated with the trait.\n",
    "    - **Example**: `1.2e-6` (a very small p-value, indicating strong evidence of association)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac3adf9",
   "metadata": {},
   "source": [
    "# Find Variant Chromosome and Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ca4213d3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/z0/1hhlpnfn2mz3vk7t57zdcy0w0000gn/T/ipykernel_92009/2974836220.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_df['chr'] = gwas_df['chr'].astype(int)\n",
      "/var/folders/z0/1hhlpnfn2mz3vk7t57zdcy0w0000gn/T/ipykernel_92009/2974836220.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_df['pos'] = gwas_df['pos'].astype(int)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>249230914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10181</td>\n",
       "      <td>243082521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>60197</td>\n",
       "      <td>197900375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>13012</td>\n",
       "      <td>190922257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>11882</td>\n",
       "      <td>180736061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>202076</td>\n",
       "      <td>170934414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>27916</td>\n",
       "      <td>159128550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>34440</td>\n",
       "      <td>146303560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>40997</td>\n",
       "      <td>141104957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>61334</td>\n",
       "      <td>135503523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>131128</td>\n",
       "      <td>134946451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>163752</td>\n",
       "      <td>133840247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>19022393</td>\n",
       "      <td>115109852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>19185163</td>\n",
       "      <td>107289436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>20001226</td>\n",
       "      <td>102461014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>60455</td>\n",
       "      <td>90238127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>302</td>\n",
       "      <td>81109537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>11603</td>\n",
       "      <td>78017073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>229783</td>\n",
       "      <td>59117442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>61098</td>\n",
       "      <td>62963628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>9574185</td>\n",
       "      <td>48119634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>51237712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr       min        max\n",
       "0     1     15791  249230914\n",
       "1     2     10181  243082521\n",
       "2     3     60197  197900375\n",
       "3     4     13012  190922257\n",
       "4     5     11882  180736061\n",
       "5     6    202076  170934414\n",
       "6     7     27916  159128550\n",
       "7     8     34440  146303560\n",
       "8     9     40997  141104957\n",
       "9    10     61334  135503523\n",
       "10   11    131128  134946451\n",
       "11   12    163752  133840247\n",
       "12   13  19022393  115109852\n",
       "13   14  19185163  107289436\n",
       "14   15  20001226  102461014\n",
       "15   16     60455   90238127\n",
       "16   17       302   81109537\n",
       "17   18     11603   78017073\n",
       "18   19    229783   59117442\n",
       "19   20     61098   62963628\n",
       "20   21   9574185   48119634\n",
       "21   22  16464274   51237712"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df[['chr', 'pos', 'ref', 'alt']] = gwas_df['variant'].str.split(':', expand=True)\n",
    "gwas_df = gwas_df.loc[gwas_df['chr'] != 'X']\n",
    "gwas_df['chr'] = gwas_df['chr'].astype(int)\n",
    "gwas_df['pos'] = gwas_df['pos'].astype(int)\n",
    "\n",
    "gwas_df.groupby('chr')['pos'].agg({'min','max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764da1f1",
   "metadata": {},
   "source": [
    "It appears that position resets on each chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae3e953",
   "metadata": {},
   "source": [
    "# Join to Find 5 Nearest Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "802efe59",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def repeated_forward_merge(gwas_df, gtf_df, num_merges=3):\n",
    "    # Initial sorting of key columns for the first merge\n",
    "    gwas_sorted = gwas_df[['variant', 'pval', 'chr', 'pos']].sort_values('pos')\n",
    "    gtf_sorted = gtf_df[['chr', 'start', 'GeneSymbol', 'gene']].sort_values('start')\n",
    "\n",
    "    # Ensure both 'start' columns are of the same type (convert to float to handle NaNs)\n",
    "    gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "    gwas_sorted['pos'] = gwas_sorted['pos'].astype(float)\n",
    "\n",
    "    # First merge_asof\n",
    "    merged_df = pd.merge_asof(\n",
    "        gwas_sorted, \n",
    "        gtf_sorted, \n",
    "        by='chr', \n",
    "        left_on='pos', \n",
    "        right_on='start', \n",
    "        direction='forward',\n",
    "        suffixes=('', '_f1')  # First set of suffixes\n",
    "    )\n",
    "\n",
    "    # Replace NaN values in 'start' column and rename for subsequent merges\n",
    "    merged_df['start'] = merged_df['start'].fillna(np.inf)\n",
    "    merged_df = merged_df.rename(columns={'start': 'start_f1', 'GeneSymbol': 'GeneSymbol_f1', 'gene': 'gene_f1'})\n",
    "    merged_df = merged_df.sort_values('start_f1')\n",
    "\n",
    "    # Iterative forward merges\n",
    "    for i in range(2, num_merges + 1):\n",
    "        # Ensure both columns are of the same type\n",
    "        merged_df[f'start_f{i-1}'] = merged_df[f'start_f{i-1}'].astype(float)\n",
    "        gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "\n",
    "        # Perform the next forward merge\n",
    "        merged_df = pd.merge_asof(\n",
    "            merged_df, \n",
    "            gtf_sorted, \n",
    "            by='chr', \n",
    "            left_on=f'start_f{i-1}', \n",
    "            right_on='start', \n",
    "            direction='forward',\n",
    "            allow_exact_matches=False,\n",
    "            suffixes=(f'_f{i-1}', f'_f{i}')\n",
    "        )\n",
    "\n",
    "        # Replace NaN values in the new 'start' column and rename for next iteration\n",
    "        merged_df['start'] = merged_df['start'].fillna(np.inf)\n",
    "        merged_df = merged_df.rename(columns={\n",
    "            'start': f'start_f{i}', \n",
    "            'GeneSymbol': f'GeneSymbol_f{i}', \n",
    "            'gene': f'gene_f{i}'\n",
    "        })\n",
    "        merged_df = merged_df.sort_values(f'start_f{i}')\n",
    "\n",
    "    # After the last merge, replace all np.inf values back to NaN\n",
    "    for i in range(1, num_merges + 1):\n",
    "        merged_df[f'start_f{i}'] = merged_df[f'start_f{i}'].replace(np.inf, np.nan)\n",
    "    \n",
    "    return merged_df\n",
    "\n",
    "f_final = repeated_forward_merge(gwas_df, gtf_df, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2a8cef2f",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Adjusted function to ensure all negative infinity values are replaced with NaN in the final result\n",
    "def repeated_backward_merge_with_nan(gwas_df, gtf_df, num_merges=3):\n",
    "    # Initial sorting of key columns for the first merge\n",
    "    gwas_sorted = gwas_df.sort_values('pos')\n",
    "    gtf_sorted = gtf_df[['chr', 'start', 'GeneSymbol', 'gene']].sort_values('start')\n",
    "\n",
    "    # Ensure both 'start' columns are of the same type (convert to float to handle NaNs)\n",
    "    gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "    gwas_sorted['pos'] = gwas_sorted['pos'].astype(float)\n",
    "\n",
    "    # First backward merge_asof\n",
    "    merged_df = pd.merge_asof(\n",
    "        gwas_sorted, \n",
    "        gtf_sorted, \n",
    "        by='chr', \n",
    "        left_on='pos', \n",
    "        right_on='start', \n",
    "        direction='backward',\n",
    "        suffixes=('', '_b1')  # First set of suffixes\n",
    "    )\n",
    "\n",
    "    # Replace NaN values in 'start' column and rename for subsequent merges\n",
    "    merged_df['start'] = merged_df['start'].fillna(-np.inf)\n",
    "    merged_df = merged_df.rename(columns={'start': 'start_b1', 'GeneSymbol': 'GeneSymbol_b1', 'gene': 'gene_b1'})\n",
    "    merged_df = merged_df.sort_values('start_b1')\n",
    "\n",
    "    # Iterative backward merges\n",
    "    for i in range(2, num_merges + 1):\n",
    "        # Ensure both columns are of the same type\n",
    "        merged_df[f'start_b{i-1}'] = merged_df[f'start_b{i-1}'].astype(float)\n",
    "        gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "\n",
    "        # Perform the next backward merge\n",
    "        merged_df = pd.merge_asof(\n",
    "            merged_df, \n",
    "            gtf_sorted, \n",
    "            by='chr', \n",
    "            left_on=f'start_b{i-1}', \n",
    "            right_on='start', \n",
    "            direction='backward',\n",
    "            allow_exact_matches=False,\n",
    "            suffixes=(f'_b{i-1}', f'_b{i}')\n",
    "        )\n",
    "\n",
    "        # Replace NaN values in the new 'start' column and rename for next iteration\n",
    "        merged_df['start'] = merged_df['start'].fillna(-np.inf)\n",
    "        merged_df = merged_df.rename(columns={\n",
    "            'start': f'start_b{i}', \n",
    "            'GeneSymbol': f'GeneSymbol_b{i}', \n",
    "            'gene': f'gene_b{i}'\n",
    "        })\n",
    "        merged_df = merged_df.sort_values(f'start_b{i}')\n",
    "\n",
    "    # After the last merge, replace all -np.inf values back to NaN\n",
    "    for i in range(1, num_merges + 1):\n",
    "        merged_df[f'start_b{i}'] = merged_df[f'start_b{i}'].replace(-np.inf, np.nan)\n",
    "    \n",
    "    return merged_df\n",
    "\n",
    "# Example usage:\n",
    "f_b_final = repeated_backward_merge_with_nan(f_final, gtf_df, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "083587df",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to calculate differences between pos and all forward and backward start columns\n",
    "def calculate_differences(df, num_merges=3):\n",
    "    # Calculate the difference between pos and start_f1, start_f2, ..., start_fx for forward merges\n",
    "    for i in range(1, num_merges + 1):\n",
    "        df[f'diff_f{i}'] = df[f'start_f{i}'] - df['pos']\n",
    "    \n",
    "    # Calculate the difference between pos and start_b1, start_b2, ..., start_bx for backward merges\n",
    "    for i in range(1, num_merges + 1):\n",
    "        df[f'diff_b{i}'] = df[f'start_b{i}'] - df['pos']\n",
    "    \n",
    "    return df\n",
    "\n",
    "f_b_final_with_differences = calculate_differences(f_b_final, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb3b90fb",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "f_b_final_with_differences[['diff_f1', 'diff_f2', 'diff_f3', 'diff_b1', 'diff_b2', 'diff_b3']].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834e0d9c-d7ab-4b89-91bf-ed11977bfaee",
   "metadata": {},
   "source": [
    "## Validation of above results: Implementing a lambda function to compute the 5 nearest neighboring genes for each mutation, then compute the inverse distance weighted s_het scores. Developing this solution to cross check the entries from the merge_asof method above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87110b9f-045c-4dad-b627-9c95d1524505",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandarallel import pandarallel\n",
    "import os\n",
    "import gc\n",
    "os.chdir(\"/Users/tanvibansal/Documents/GitHub/Capstone\")\n",
    "#grab a small sample of the gwas dataframe containing mutation information \n",
    "gwas_df.head()\n",
    "gtf_df.head()\n",
    "gwas_sample = gwas_df.iloc[np.random.choice(range(len(gwas_df)),size=100000,replace=False)] \n",
    "\n",
    "#drop the genes that are not present in the s_het file as these may cause na's that will break the ldsc algorithm in the next step\n",
    "s_het_info = pd.read_excel('s_het_info.xlsx',sheet_name=1).rename(columns={\"ensg\":\"GeneSymbol\"}).set_index(\"GeneSymbol\")\n",
    "gtf_df_tidy = s_het_info.join(gtf_df.set_index(\"GeneSymbol\"),how=\"inner\")[[\"chr\",\"start\",\"end\",\"gene\"]]\n",
    "\n",
    "#write a function to find the 5 nearest genes by absolute difference in mutation position and gene start position\n",
    "def nearest_gene_finder(m):\n",
    "    #get the absolute l1 distance \n",
    "    dists = np.abs(gtf_df_tidy.loc[gtf_df_tidy.chr == m.chr,\"start\"] - m.pos)\n",
    "    top_5 = dists.sort_values()[0:5]\n",
    "    out = pd.DataFrame(m).T\n",
    "    out[[\"GeneSymbol.f1\",\"GeneSymbol.f2\",\"GeneSymbol.f3\",\"GeneSymbol.f4\",\"GeneSymbol.f5\"]] = top_5.index\n",
    "    out[[\"dist.f1\",\"dist.f2\",\"dist.f3\",\"dist.f4\",\"dist.f5\"]] = top_5.values/1000 #convert distance to kilobases by dividing through 1000\n",
    "    return(out)\n",
    "\n",
    "#create a sample of the gwas dataframe with only mutations on chr 22 for this stage\n",
    "gwas_sample_22 = gwas_df.loc[gwas_df.chr ==22]\n",
    "\n",
    "#apply the function over each row of the sample mutation df \n",
    "pandarallel.initialize(progress_bar=True)\n",
    "mutn_5_nearest_genes_sample = gwas_sample_22.parallel_apply(lambda m: nearest_gene_finder(m),axis=1)\n",
    "gc.collect() \n",
    "mutn_5_nearest_genes_sample_df = pd.concat(list(mutn_5_nearest_genes_sample))\n",
    "\n",
    "#now join the s_het values onto each of our 5 nearest genes and compute the distance weighted s_het \n",
    "s_het_w = mutn_5_nearest_genes_sample_df.copy(deep=True)\n",
    "for i in range(1,6):\n",
    "    s_het_temp = s_het_info[[\"post_mean\"]].reset_index().rename(columns = {\"GeneSymbol\":\"GeneSymbol.f%s\"%(i)}).set_index(\"GeneSymbol.f%s\"%(i))\n",
    "    s_het_w = s_het_w.reset_index().set_index(\"GeneSymbol.f%s\"%(i)).join(s_het_temp)\n",
    "    s_het_w[\"s_het.f%s\"%(i)] = s_het_w[\"post_mean\"]\n",
    "    s_het_w[\"s_het_w.f%s\"%(i)] = s_het_w[\"post_mean\"]/s_het_w[\"dist.f%s\"%(i)]\n",
    "    s_het_w = s_het_w.drop(columns=\"post_mean\")\n",
    "s_het_w = s_het_w.reset_index() \n",
    "s_het_w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "037345df-12bc-4876-91de-377e1aad0d8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol.f5</th>\n",
       "      <th>GeneSymbol.f4</th>\n",
       "      <th>GeneSymbol.f3</th>\n",
       "      <th>GeneSymbol.f2</th>\n",
       "      <th>GeneSymbol.f1</th>\n",
       "      <th>index</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>dist.f1</th>\n",
       "      <th>dist.f2</th>\n",
       "      <th>dist.f3</th>\n",
       "      <th>dist.f4</th>\n",
       "      <th>dist.f5</th>\n",
       "      <th>s_het.f1</th>\n",
       "      <th>s_het_w.f1</th>\n",
       "      <th>s_het.f2</th>\n",
       "      <th>s_het_w.f2</th>\n",
       "      <th>s_het.f3</th>\n",
       "      <th>s_het_w.f3</th>\n",
       "      <th>s_het.f4</th>\n",
       "      <th>s_het_w.f4</th>\n",
       "      <th>s_het.f5</th>\n",
       "      <th>s_het_w.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184537</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.084046</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>60578.3</td>\n",
       "      <td>-262.512</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.003183</td>\n",
       "      <td>1.07204</td>\n",
       "      <td>0.283702</td>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>607.367</td>\n",
       "      <td>800.028</td>\n",
       "      <td>978.552</td>\n",
       "      <td>1101.570</td>\n",
       "      <td>1132.913</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.546518e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184538</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.080987</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>58373.6</td>\n",
       "      <td>-220.884</td>\n",
       "      <td>0.003806</td>\n",
       "      <td>0.003111</td>\n",
       "      <td>1.22341</td>\n",
       "      <td>0.221175</td>\n",
       "      <td>22</td>\n",
       "      <td>16488635</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>583.006</td>\n",
       "      <td>775.667</td>\n",
       "      <td>954.191</td>\n",
       "      <td>1077.209</td>\n",
       "      <td>1108.552</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.611532e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184539</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.080078</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>57718.1</td>\n",
       "      <td>-250.356</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>0.003121</td>\n",
       "      <td>1.19791</td>\n",
       "      <td>0.230954</td>\n",
       "      <td>22</td>\n",
       "      <td>16488702</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>582.939</td>\n",
       "      <td>775.600</td>\n",
       "      <td>954.124</td>\n",
       "      <td>1077.142</td>\n",
       "      <td>1108.485</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.611715e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184540</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078102</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>56294.0</td>\n",
       "      <td>-249.417</td>\n",
       "      <td>0.004095</td>\n",
       "      <td>0.003069</td>\n",
       "      <td>1.33402</td>\n",
       "      <td>0.182199</td>\n",
       "      <td>22</td>\n",
       "      <td>16495833</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>575.808</td>\n",
       "      <td>768.469</td>\n",
       "      <td>946.993</td>\n",
       "      <td>1070.011</td>\n",
       "      <td>1101.354</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.631382e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184541</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.07444</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>53654.4</td>\n",
       "      <td>-203.467</td>\n",
       "      <td>0.004117</td>\n",
       "      <td>0.003224</td>\n",
       "      <td>1.27711</td>\n",
       "      <td>0.201566</td>\n",
       "      <td>22</td>\n",
       "      <td>16498458</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>573.183</td>\n",
       "      <td>765.844</td>\n",
       "      <td>944.368</td>\n",
       "      <td>1067.386</td>\n",
       "      <td>1098.729</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.638696e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364298</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.29612</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>213436.0</td>\n",
       "      <td>-437.663</td>\n",
       "      <td>-0.002442</td>\n",
       "      <td>0.002004</td>\n",
       "      <td>-1.21861</td>\n",
       "      <td>0.222992</td>\n",
       "      <td>22</td>\n",
       "      <td>51229717</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>23.788</td>\n",
       "      <td>53.093</td>\n",
       "      <td>116.874</td>\n",
       "      <td>168.535</td>\n",
       "      <td>190.495</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.122508e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364299</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.072999</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>52615.7</td>\n",
       "      <td>-172.528</td>\n",
       "      <td>0.001794</td>\n",
       "      <td>0.003184</td>\n",
       "      <td>0.563533</td>\n",
       "      <td>0.573073</td>\n",
       "      <td>22</td>\n",
       "      <td>51229805</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>23.876</td>\n",
       "      <td>53.181</td>\n",
       "      <td>116.962</td>\n",
       "      <td>168.623</td>\n",
       "      <td>190.583</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.120911e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179763</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364300</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.053676</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>38688.2</td>\n",
       "      <td>-20.4974</td>\n",
       "      <td>0.002084</td>\n",
       "      <td>0.003916</td>\n",
       "      <td>0.532177</td>\n",
       "      <td>0.594604</td>\n",
       "      <td>22</td>\n",
       "      <td>51231220</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25.291</td>\n",
       "      <td>54.596</td>\n",
       "      <td>118.377</td>\n",
       "      <td>170.038</td>\n",
       "      <td>191.998</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.095559e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179764</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364301</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.299014</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>215522.0</td>\n",
       "      <td>-260.204</td>\n",
       "      <td>-0.001741</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>-0.895286</td>\n",
       "      <td>0.370635</td>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>31.134</td>\n",
       "      <td>60.439</td>\n",
       "      <td>124.220</td>\n",
       "      <td>175.881</td>\n",
       "      <td>197.841</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>1.996989e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179765</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364302</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.05515</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>39750.5</td>\n",
       "      <td>-23.8471</td>\n",
       "      <td>0.001338</td>\n",
       "      <td>0.003885</td>\n",
       "      <td>0.344285</td>\n",
       "      <td>0.730632</td>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31.783</td>\n",
       "      <td>61.088</td>\n",
       "      <td>124.869</td>\n",
       "      <td>176.530</td>\n",
       "      <td>198.490</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>1.986610e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003557</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneSymbol.f5    GeneSymbol.f4  ...  s_het.f5 s_het_w.f5\n",
       "0       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000105\n",
       "1       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000107\n",
       "2       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000107\n",
       "3       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000108\n",
       "4       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000108\n",
       "...                 ...              ...  ...       ...        ...\n",
       "179761  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003706\n",
       "179762  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003705\n",
       "179763  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003677\n",
       "179764  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003569\n",
       "179765  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003557\n",
       "\n",
       "[179766 rows x 36 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutn_5_nearest_genes_sample_df = pd.concat(list(mutn_5_nearest_genes_sample))\n",
    "\n",
    "#now join the s_het values onto each of our 5 nearest genes and compute the distance weighted s_het \n",
    "s_het_w = mutn_5_nearest_genes_sample_df.copy(deep=True)\n",
    "for i in range(1,6):\n",
    "    s_het_temp = s_het_info[[\"post_mean\"]].reset_index().rename(columns = {\"GeneSymbol\":\"GeneSymbol.f%s\"%(i)}).set_index(\"GeneSymbol.f%s\"%(i))\n",
    "    s_het_w = s_het_w.reset_index().set_index(\"GeneSymbol.f%s\"%(i)).join(s_het_temp)\n",
    "    s_het_w[\"s_het.f%s\"%(i)] = s_het_w[\"post_mean\"]\n",
    "    s_het_w[\"s_het_w.f%s\"%(i)] = s_het_w[\"post_mean\"]/s_het_w[\"dist.f%s\"%(i)]\n",
    "    s_het_w = s_het_w.drop(columns=\"post_mean\")\n",
    "s_het_w = s_het_w.reset_index() \n",
    "s_het_w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "3096565f-419a-4ece-bb24-1b62d9423906",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>GeneSymbol.f5</th>\n",
       "      <th>GeneSymbol.f4</th>\n",
       "      <th>GeneSymbol.f3</th>\n",
       "      <th>GeneSymbol.f2</th>\n",
       "      <th>GeneSymbol.f1</th>\n",
       "      <th>index</th>\n",
       "      <th>SNP</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>dist.f1</th>\n",
       "      <th>dist.f2</th>\n",
       "      <th>dist.f3</th>\n",
       "      <th>dist.f4</th>\n",
       "      <th>dist.f5</th>\n",
       "      <th>s_het_w.f1</th>\n",
       "      <th>s_het_w.f2</th>\n",
       "      <th>s_het_w.f3</th>\n",
       "      <th>s_het_w.f4</th>\n",
       "      <th>s_het_w.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35779</th>\n",
       "      <td>35779</td>\n",
       "      <td>ENSG00000100014</td>\n",
       "      <td>ENSG00000100028</td>\n",
       "      <td>ENSG00000138867</td>\n",
       "      <td>ENSG00000100024</td>\n",
       "      <td>ENSG00000128271</td>\n",
       "      <td>13220316</td>\n",
       "      <td>22:24813847:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.025098</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>18090.4</td>\n",
       "      <td>144.518</td>\n",
       "      <td>0.009787</td>\n",
       "      <td>0.005276</td>\n",
       "      <td>1.85511</td>\n",
       "      <td>0.063582</td>\n",
       "      <td>22</td>\n",
       "      <td>24813847</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000</td>\n",
       "      <td>76.228</td>\n",
       "      <td>122.544</td>\n",
       "      <td>138.079</td>\n",
       "      <td>147.062</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.002741</td>\n",
       "      <td>0.000696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76336</th>\n",
       "      <td>76336</td>\n",
       "      <td>ENSG00000185666</td>\n",
       "      <td>ENSG00000100225</td>\n",
       "      <td>ENSG00000128276</td>\n",
       "      <td>ENSG00000100220</td>\n",
       "      <td>ENSG00000184459</td>\n",
       "      <td>13260873</td>\n",
       "      <td>22:32809832:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.363284</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>261847.0</td>\n",
       "      <td>8.60544</td>\n",
       "      <td>-0.001623</td>\n",
       "      <td>0.001715</td>\n",
       "      <td>-0.94638</td>\n",
       "      <td>0.343956</td>\n",
       "      <td>22</td>\n",
       "      <td>32809832</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000</td>\n",
       "      <td>26.263</td>\n",
       "      <td>58.960</td>\n",
       "      <td>60.831</td>\n",
       "      <td>93.975</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000960</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107911</th>\n",
       "      <td>107911</td>\n",
       "      <td>ENSG00000100146</td>\n",
       "      <td>ENSG00000100151</td>\n",
       "      <td>ENSG00000184381</td>\n",
       "      <td>ENSG00000100156</td>\n",
       "      <td>ENSG00000128298</td>\n",
       "      <td>13292448</td>\n",
       "      <td>22:38480659:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.139847</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>100799.0</td>\n",
       "      <td>-586.088</td>\n",
       "      <td>-0.006443</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>-2.57693</td>\n",
       "      <td>0.009969</td>\n",
       "      <td>22</td>\n",
       "      <td>38480659</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000</td>\n",
       "      <td>6.518</td>\n",
       "      <td>26.843</td>\n",
       "      <td>28.341</td>\n",
       "      <td>113.966</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>0.001480</td>\n",
       "      <td>0.002077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62361</th>\n",
       "      <td>62361</td>\n",
       "      <td>ENSG00000100330</td>\n",
       "      <td>ENSG00000100314</td>\n",
       "      <td>ENSG00000100319</td>\n",
       "      <td>ENSG00000184076</td>\n",
       "      <td>ENSG00000100325</td>\n",
       "      <td>13246898</td>\n",
       "      <td>22:30184599:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.386249</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>278399.0</td>\n",
       "      <td>-1209.98</td>\n",
       "      <td>-0.009549</td>\n",
       "      <td>0.001696</td>\n",
       "      <td>-5.63143</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22</td>\n",
       "      <td>30184599</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001</td>\n",
       "      <td>21.241</td>\n",
       "      <td>57.654</td>\n",
       "      <td>68.607</td>\n",
       "      <td>94.559</td>\n",
       "      <td>32.282600</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65901</th>\n",
       "      <td>65901</td>\n",
       "      <td>ENSG00000100036</td>\n",
       "      <td>ENSG00000214491</td>\n",
       "      <td>ENSG00000185339</td>\n",
       "      <td>ENSG00000128242</td>\n",
       "      <td>ENSG00000100029</td>\n",
       "      <td>13250438</td>\n",
       "      <td>22:30972611:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003476</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>2505.56</td>\n",
       "      <td>-11.0614</td>\n",
       "      <td>0.018132</td>\n",
       "      <td>0.015571</td>\n",
       "      <td>1.16447</td>\n",
       "      <td>0.244234</td>\n",
       "      <td>22</td>\n",
       "      <td>30972611</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>21.989</td>\n",
       "      <td>30.379</td>\n",
       "      <td>53.825</td>\n",
       "      <td>59.157</td>\n",
       "      <td>58.204200</td>\n",
       "      <td>0.001769</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.000137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85681</th>\n",
       "      <td>85681</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100284</td>\n",
       "      <td>ENSG00000100281</td>\n",
       "      <td>ENSG00000175329</td>\n",
       "      <td>ENSG00000133424</td>\n",
       "      <td>13270218</td>\n",
       "      <td>22:34509679:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.025321</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>18251.0</td>\n",
       "      <td>-63.455</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.005245</td>\n",
       "      <td>0.019564</td>\n",
       "      <td>0.984391</td>\n",
       "      <td>22</td>\n",
       "      <td>34509679</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>951.467</td>\n",
       "      <td>952.450</td>\n",
       "      <td>1143.766</td>\n",
       "      <td>1185.589</td>\n",
       "      <td>1266.675</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85685</th>\n",
       "      <td>85685</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100284</td>\n",
       "      <td>ENSG00000100281</td>\n",
       "      <td>ENSG00000133424</td>\n",
       "      <td>ENSG00000175329</td>\n",
       "      <td>13270222</td>\n",
       "      <td>22:34510635:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001863</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>1342.61</td>\n",
       "      <td>8.71801</td>\n",
       "      <td>0.007962</td>\n",
       "      <td>0.020621</td>\n",
       "      <td>0.386107</td>\n",
       "      <td>0.699417</td>\n",
       "      <td>22</td>\n",
       "      <td>34510635</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>951.494</td>\n",
       "      <td>952.423</td>\n",
       "      <td>1142.810</td>\n",
       "      <td>1184.633</td>\n",
       "      <td>1265.719</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85684</th>\n",
       "      <td>85684</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100284</td>\n",
       "      <td>ENSG00000100281</td>\n",
       "      <td>ENSG00000133424</td>\n",
       "      <td>ENSG00000175329</td>\n",
       "      <td>13270221</td>\n",
       "      <td>22:34510633:C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.374701</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>450701.0</td>\n",
       "      <td>1128.13</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>0.001709</td>\n",
       "      <td>3.47398</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>22</td>\n",
       "      <td>34510633</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>951.496</td>\n",
       "      <td>952.421</td>\n",
       "      <td>1142.812</td>\n",
       "      <td>1184.635</td>\n",
       "      <td>1265.721</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85682</th>\n",
       "      <td>85682</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100284</td>\n",
       "      <td>ENSG00000100281</td>\n",
       "      <td>ENSG00000175329</td>\n",
       "      <td>ENSG00000133424</td>\n",
       "      <td>13270219</td>\n",
       "      <td>22:34509709:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.017356</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>12510.0</td>\n",
       "      <td>126.11</td>\n",
       "      <td>0.001541</td>\n",
       "      <td>0.006316</td>\n",
       "      <td>0.244006</td>\n",
       "      <td>0.807227</td>\n",
       "      <td>22</td>\n",
       "      <td>34509709</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>951.497</td>\n",
       "      <td>952.420</td>\n",
       "      <td>1143.736</td>\n",
       "      <td>1185.559</td>\n",
       "      <td>1266.645</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85683</th>\n",
       "      <td>85683</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100284</td>\n",
       "      <td>ENSG00000100281</td>\n",
       "      <td>ENSG00000175329</td>\n",
       "      <td>ENSG00000133424</td>\n",
       "      <td>13270220</td>\n",
       "      <td>22:34509836:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.006344</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>4572.53</td>\n",
       "      <td>17.8461</td>\n",
       "      <td>-0.009873</td>\n",
       "      <td>0.010883</td>\n",
       "      <td>-0.907196</td>\n",
       "      <td>0.364304</td>\n",
       "      <td>22</td>\n",
       "      <td>34509836</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>951.624</td>\n",
       "      <td>952.293</td>\n",
       "      <td>1143.609</td>\n",
       "      <td>1185.432</td>\n",
       "      <td>1266.518</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        level_0    GeneSymbol.f5  ... s_het_w.f4 s_het_w.f5\n",
       "35779     35779  ENSG00000100014  ...   0.002741   0.000696\n",
       "76336     76336  ENSG00000185666  ...   0.000086   0.000214\n",
       "107911   107911  ENSG00000100146  ...   0.001480   0.002077\n",
       "62361     62361  ENSG00000100330  ...   0.000199   0.001071\n",
       "65901     65901  ENSG00000100036  ...   0.000014   0.000137\n",
       "...         ...              ...  ...        ...        ...\n",
       "85681     85681  ENSG00000100292  ...   0.000007   0.000004\n",
       "85685     85685  ENSG00000100292  ...   0.000007   0.000004\n",
       "85684     85684  ENSG00000100292  ...   0.000007   0.000004\n",
       "85682     85682  ENSG00000100292  ...   0.000007   0.000004\n",
       "85683     85683  ENSG00000100292  ...   0.000007   0.000004\n",
       "\n",
       "[179766 rows x 32 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_het_w.sort_values(by = \"dist.f1\",ascending=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e68796f0-c8cf-49e3-8f39-e12e786c34c2",
   "metadata": {},
   "source": [
    "## Prepare the annotation file containing our features for the LDSC program "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6523dfd1-d531-4acd-8860-e609ec77fd36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>s_het.f1</th>\n",
       "      <th>s_het.f2</th>\n",
       "      <th>s_het.f3</th>\n",
       "      <th>s_het.f4</th>\n",
       "      <th>s_het.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16488635</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16488702</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16495833</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16498458</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>22</td>\n",
       "      <td>51229717</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>22</td>\n",
       "      <td>51229805</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179763</th>\n",
       "      <td>22</td>\n",
       "      <td>51231220</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179764</th>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179765</th>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR        BP              SNP  ...  s_het.f3  s_het.f4  s_het.f5\n",
       "0       22  16464274  22:16464274:A:C  ...  0.000249  0.008926  0.118445\n",
       "1       22  16488635  22:16488635:C:A  ...  0.000249  0.008926  0.118445\n",
       "2       22  16488702  22:16488702:G:C  ...  0.000249  0.008926  0.118445\n",
       "3       22  16495833  22:16495833:C:A  ...  0.000249  0.008926  0.118445\n",
       "4       22  16498458  22:16498458:G:A  ...  0.000249  0.008926  0.118445\n",
       "...     ..       ...              ...  ...       ...       ...       ...\n",
       "179761  22  51229717  22:51229717:A:T  ...  0.248066  0.001543  0.706036\n",
       "179762  22  51229805  22:51229805:T:C  ...  0.248066  0.001543  0.706036\n",
       "179763  22  51231220  22:51231220:A:G  ...  0.248066  0.001543  0.706036\n",
       "179764  22  51237063  22:51237063:T:C  ...  0.248066  0.001543  0.706036\n",
       "179765  22  51237712  22:51237712:G:A  ...  0.248066  0.001543  0.706036\n",
       "\n",
       "[179766 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#format the dataframe's column names, columns to retain, and match the SNPs present in the bim file\n",
    "s_het_w = s_het_w.reset_index().rename(columns={'chr': 'CHR','pos': 'BP','variant': 'SNP'})\n",
    "#annot_df = s_het_w[['CHR', 'BP', 'SNP', 's_het_w.f1', 's_het_w.f2', 's_het_w.f3', 's_het_w.f4', 's_het_w.f5']]\n",
    "annot_df = s_het_w[['CHR', 'BP', 'SNP', 's_het.f1', 's_het.f2', 's_het.f3', 's_het.f4', 's_het.f5']]\n",
    "annot_df = annot_df.sort_values(['CHR', 'BP'])\n",
    "annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "aba2a2f5-4c82-4534-8a48-b154312d8adf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>s_het.f1</th>\n",
       "      <th>s_het.f2</th>\n",
       "      <th>s_het.f3</th>\n",
       "      <th>s_het.f4</th>\n",
       "      <th>s_het.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16488635</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16488702</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16495833</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16498458</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.118445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>22</td>\n",
       "      <td>51229717</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>22</td>\n",
       "      <td>51229805</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179763</th>\n",
       "      <td>22</td>\n",
       "      <td>51231220</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179764</th>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179765</th>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.706036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR        BP              SNP  ...  s_het.f3  s_het.f4  s_het.f5\n",
       "0       22  16464274  22:16464274:A:C  ...  0.000249  0.008926  0.118445\n",
       "1       22  16488635  22:16488635:C:A  ...  0.000249  0.008926  0.118445\n",
       "2       22  16488702  22:16488702:G:C  ...  0.000249  0.008926  0.118445\n",
       "3       22  16495833  22:16495833:C:A  ...  0.000249  0.008926  0.118445\n",
       "4       22  16498458  22:16498458:G:A  ...  0.000249  0.008926  0.118445\n",
       "...     ..       ...              ...  ...       ...       ...       ...\n",
       "179761  22  51229717  22:51229717:A:T  ...  0.248066  0.001543  0.706036\n",
       "179762  22  51229805  22:51229805:T:C  ...  0.248066  0.001543  0.706036\n",
       "179763  22  51231220  22:51231220:A:G  ...  0.248066  0.001543  0.706036\n",
       "179764  22  51237063  22:51237063:T:C  ...  0.248066  0.001543  0.706036\n",
       "179765  22  51237712  22:51237712:G:A  ...  0.248066  0.001543  0.706036\n",
       "\n",
       "[179766 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "67b4efa1-6c17-4ec6-9171-cb4d4340dbd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#column names for the bim file according to the plink documentation site\n",
    "names = [\"CHR\",\"SNP\",\"POS\",\"BP\",\"A1\",\"A2\"]\n",
    "os.chdir(\"/Users/tanvibansal/Documents/GitHub/ldsc\")\n",
    "bim = pd.read_csv(\"22.bim\",sep=\"\\t\",header=None,names=names)\n",
    "\n",
    "#join the annotated df onto the bim file and sort in the order of the bim file exactly to prep for ldsc. we should have the same exact # of rows \n",
    "annot_df_filt = bim.set_index([\"CHR\",\"BP\"]).join(annot_df.set_index([\"CHR\",\"BP\"]),how=\"left\",rsuffix=\".a\").drop_duplicates(subset=[\"SNP\"])\n",
    "annot_df_sort = annot_df_filt.reset_index().set_index(\"SNP\").loc[bim.SNP.values]\n",
    "annot_df_sort[\"CM\"] = 0.0\n",
    "\n",
    "#select the cols of interest and write the file out as .annot to use for ldsc\n",
    "#annot_df_prep = annot_df_sort.reset_index()[[\"CHR\",\"BP\",\"SNP\", \"CM\", \"s_het_w.f1\", \"s_het_w.f2\", \"s_het_w.f3\", \"s_het_w.f4\", \"s_het_w.f5\"]]\n",
    "annot_df_prep = annot_df_sort.reset_index()[[\"CHR\",\"BP\",\"SNP\", \"CM\", \"s_het.f1\", \"s_het.f2\", \"s_het.f3\", \"s_het.f4\", \"s_het.f5\"]]\n",
    "annot_df_prep = annot_df_prep.dropna()\n",
    "annot_df_prep.to_csv(\"1kg_eur/s_het_unweighted_ch22.annot\",sep=\"\\t\",index=False)\n",
    "annot_df_prep\n",
    "\n",
    "unique_snps = annot_df_prep['SNP'].unique()  # Get unique SNPs to avoid duplicates\n",
    "\n",
    "# Save the SNPs to a text file\n",
    "with open('1kg_eur/s_het_22/filtered_snps.txt', 'w') as f:\n",
    "    for snp in unique_snps:\n",
    "        f.write(f\"{snp}\\n\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0aa6a86-99cc-48cc-9b43-5771a5dc5b98",
   "metadata": {},
   "source": [
    "#### we need to run the following 2 commands to get the ldsc computation to run as intended \n",
    "\n",
    "1. Make the .bed/.bim files with our filtered SNPs\n",
    "\n",
    "\n",
    "plink --bfile /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/22 --extract /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22/filtered_snps.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22\n",
    "\n",
    "2. Compute the ld scores with our inputs specified\n",
    "\n",
    "python ldsc.py --l2 --bfile /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22_uw/s_het_22_uw --ld-wind-cm 1 --annot /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22_uw/s_het_uw_ch22.annot --out 1kg_eur/s_het_22_uw/s_het_22_uw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "40e26bac-c33f-4d5f-afb9-db9b48b13578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip \n",
    "with gzip.open('1kg_eur/s_het_22.l2.ldscore.gz', 'rb') as f: \n",
    "    file_content = pd.read_csv(f,sep=\"\\t\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "2fd480fe-ac65-4e60-92b6-24b625ad242d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>s_het_w.f1L2</th>\n",
       "      <th>s_het_w.f2L2</th>\n",
       "      <th>s_het_w.f3L2</th>\n",
       "      <th>s_het_w.f4L2</th>\n",
       "      <th>s_het_w.f5L2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7276</th>\n",
       "      <td>22</td>\n",
       "      <td>rs12484974</td>\n",
       "      <td>31436494</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7552</th>\n",
       "      <td>22</td>\n",
       "      <td>rs131243</td>\n",
       "      <td>32124778</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7305</th>\n",
       "      <td>22</td>\n",
       "      <td>rs917208</td>\n",
       "      <td>31493125</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459</th>\n",
       "      <td>22</td>\n",
       "      <td>rs929104</td>\n",
       "      <td>38208511</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11460</th>\n",
       "      <td>22</td>\n",
       "      <td>rs2285179</td>\n",
       "      <td>38211954</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.057</td>\n",
       "      <td>0.065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7659</th>\n",
       "      <td>22</td>\n",
       "      <td>rs976915</td>\n",
       "      <td>32428445</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>22</td>\n",
       "      <td>rs130256</td>\n",
       "      <td>32418552</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7657</th>\n",
       "      <td>22</td>\n",
       "      <td>rs17744994</td>\n",
       "      <td>32413527</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7654</th>\n",
       "      <td>22</td>\n",
       "      <td>rs4821014</td>\n",
       "      <td>32392612</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8110</th>\n",
       "      <td>22</td>\n",
       "      <td>rs2051569</td>\n",
       "      <td>33025180</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18768 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR         SNP        BP  ...  s_het_w.f3L2  s_het_w.f4L2  s_het_w.f5L2\n",
       "7276    22  rs12484974  31436494  ...         0.003         0.002         0.004\n",
       "7552    22    rs131243  32124778  ...         0.021         0.020         0.016\n",
       "7305    22    rs917208  31493125  ...         0.014         0.010         0.009\n",
       "11459   22    rs929104  38208511  ...         0.005         0.005         0.015\n",
       "11460   22   rs2285179  38211954  ...         0.041         0.057         0.065\n",
       "...    ...         ...       ...  ...           ...           ...           ...\n",
       "7659    22    rs976915  32428445  ...         0.008         0.005         0.006\n",
       "7658    22    rs130256  32418552  ...         0.009         0.007         0.011\n",
       "7657    22  rs17744994  32413527  ...         0.009         0.006         0.007\n",
       "7654    22   rs4821014  32392612  ...         0.016         0.016         0.015\n",
       "8110    22   rs2051569  33025180  ...         0.002         0.001         0.001\n",
       "\n",
       "[18768 rows x 8 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_content.sort_values(by=\"s_het_w.f1L2\",ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "dc2b867f-6d18-422b-9f0a-cc440a0c507a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>L2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7464</th>\n",
       "      <td>22</td>\n",
       "      <td>rs12484974</td>\n",
       "      <td>31436494</td>\n",
       "      <td>21.367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHR         SNP        BP      L2\n",
       "7464   22  rs12484974  31436494  21.367"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gzip \n",
    "with gzip.open('22.l2.ldscore.gz', 'rb') as f: \n",
    "    v_file_content = pd.read_csv(f,sep=\"\\t\")\n",
    "v_file_content.loc[v_file_content.SNP == \"rs12484974\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "29ad9f30-e03f-4485-a11f-2aaa7d711109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>s_het_w.f1</th>\n",
       "      <th>s_het_w.f2</th>\n",
       "      <th>s_het_w.f3</th>\n",
       "      <th>s_het_w.f4</th>\n",
       "      <th>s_het_w.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8108</th>\n",
       "      <td>22</td>\n",
       "      <td>32809832</td>\n",
       "      <td>rs9609538</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000960</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11817</th>\n",
       "      <td>22</td>\n",
       "      <td>38480659</td>\n",
       "      <td>rs2267366</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>0.001480</td>\n",
       "      <td>0.002077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12935</th>\n",
       "      <td>22</td>\n",
       "      <td>41347346</td>\n",
       "      <td>rs34654213</td>\n",
       "      <td>0.0</td>\n",
       "      <td>113.538600</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.001116</td>\n",
       "      <td>0.002587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15610</th>\n",
       "      <td>22</td>\n",
       "      <td>45898117</td>\n",
       "      <td>rs3810633</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.266300</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000295</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6895</th>\n",
       "      <td>22</td>\n",
       "      <td>30184599</td>\n",
       "      <td>rs140147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.282600</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9373</th>\n",
       "      <td>22</td>\n",
       "      <td>34514810</td>\n",
       "      <td>rs243001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9372</th>\n",
       "      <td>22</td>\n",
       "      <td>34513777</td>\n",
       "      <td>rs130662</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9371</th>\n",
       "      <td>22</td>\n",
       "      <td>34512123</td>\n",
       "      <td>rs5999208</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9370</th>\n",
       "      <td>22</td>\n",
       "      <td>34510915</td>\n",
       "      <td>rs243000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9369</th>\n",
       "      <td>22</td>\n",
       "      <td>34510633</td>\n",
       "      <td>rs135200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18768 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR        BP         SNP  ...  s_het_w.f3  s_het_w.f4  s_het_w.f5\n",
       "8108    22  32809832   rs9609538  ...    0.000053    0.000086    0.000214\n",
       "11817   22  38480659   rs2267366  ...    0.000336    0.001480    0.002077\n",
       "12935   22  41347346  rs34654213  ...    0.000032    0.001116    0.002587\n",
       "15610   22  45898117   rs3810633  ...    0.000295    0.000101    0.000070\n",
       "6895    22  30184599    rs140147  ...    0.000074    0.000199    0.001071\n",
       "...    ...       ...         ...  ...         ...         ...         ...\n",
       "9373    22  34514810    rs243001  ...    0.000074    0.000007    0.000004\n",
       "9372    22  34513777    rs130662  ...    0.000074    0.000007    0.000004\n",
       "9371    22  34512123   rs5999208  ...    0.000074    0.000007    0.000004\n",
       "9370    22  34510915    rs243000  ...    0.000074    0.000007    0.000004\n",
       "9369    22  34510633    rs135200  ...    0.000074    0.000007    0.000004\n",
       "\n",
       "[18768 rows x 9 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df_prep.loc[annot_df_prep.SNP ==\"rs12484974\"]\n",
    "annot_df_prep.sort_values(by = \"s_het_w.f1\",ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29aea24e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot histograms for each difference column\n",
    "def plot_histograms(df, columns):\n",
    "    plt.figure(figsize=(14, 10))\n",
    "    \n",
    "    for i, col in enumerate(columns):\n",
    "        plt.subplot(2, 3, i + 1)  # Create a subplot for each column\n",
    "        df[col].hist(bins=50, color='skyblue', edgecolor='black')\n",
    "        plt.title(f'Distribution of {col}')\n",
    "        plt.xlabel('Value')\n",
    "        plt.ylabel('Frequency')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "# Define the columns to plot\n",
    "columns_to_plot = ['diff_f1', 'diff_f2', 'diff_f3', 'diff_b1', 'diff_b2', 'diff_b3']\n",
    "# Call the function to plot histograms\n",
    "plot_histograms(f_b_final_with_differences, columns_to_plot)\n",
    "\n",
    "columns_to_plot = [\"f1_dist\",\"f2_dist\",\"f3_dist\",\"f4_dist\",\"f5_dist\"]\n",
    "plot_histograms(nearest_genes_sample_df, columns_to_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e4837c-ffa1-491f-989b-a6cd2611bd33",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_het_info = pd.read_excel('s_het_info.xlsx',sheet_name=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4f1fa143",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to perform the merge for all forward and backward GeneSymbols and rename the 'post_mean' columns\n",
    "def merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges=3):\n",
    "    # Rename 'ensg' column in s_het_info to match the GeneSymbol columns in f_b_final_with_differences\n",
    "    s_het_info = s_het_info.rename(columns={'ensg': 'GeneSymbol'})\n",
    "\n",
    "    # Iterate over both forward and backward gene symbols\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Forward merge for each GeneSymbol_f\n",
    "        f_b_final_with_differences = pd.merge(\n",
    "            f_b_final_with_differences, \n",
    "            s_het_info[['GeneSymbol', 'post_mean']], \n",
    "            left_on=f'GeneSymbol_f{i}', \n",
    "            right_on='GeneSymbol', \n",
    "            how='left',\n",
    "            suffixes=('', f'_f{i}')\n",
    "        )\n",
    "        \n",
    "        # Rename the post_mean column to indicate the forward direction\n",
    "        f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "            columns={'post_mean': f's_het_post_f{i}'}\n",
    "        )\n",
    "        \n",
    "        # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "        f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "        # Backward merge for each GeneSymbol_b\n",
    "        f_b_final_with_differences = pd.merge(\n",
    "            f_b_final_with_differences, \n",
    "            s_het_info[['GeneSymbol', 'post_mean']], \n",
    "            left_on=f'GeneSymbol_b{i}', \n",
    "            right_on='GeneSymbol', \n",
    "            how='left',\n",
    "            suffixes=('', f'_b{i}')\n",
    "        )\n",
    "        \n",
    "        # Rename the post_mean column to indicate the backward direction\n",
    "        f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "            columns={'post_mean': f's_het_post_b{i}'}\n",
    "        )\n",
    "        \n",
    "        # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "        f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "    return f_b_final_with_differences\n",
    "\n",
    "# Example usage:\n",
    "f_b_final_merged_s_het = merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db752cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f7cc0427",
   "metadata": {},
   "outputs": [],
   "source": [
    "# f_b_final_merged_s_het[['diff_b1']] = abs(f_b_final_merged_s_het[['diff_b1']])\n",
    "# f_b_final_merged_s_het[['diff_b2']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# f_b_final_merged_s_het[['diff_b3']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# Loop to apply the absolute value for all diff_b columns\n",
    "def apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3):\n",
    "    for i in range(1, num_merges + 1):\n",
    "        f_b_final_merged_s_het[f'diff_b{i}'] = f_b_final_merged_s_het[f'diff_b{i}'].abs()\n",
    "\n",
    "    return f_b_final_merged_s_het\n",
    "\n",
    "# Example usage:\n",
    "f_b_final_merged_s_het = apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4c6da80",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3566073f",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to keep only the distance and s_het measures from the dataframe\n",
    "def keep_dist_and_s_het_measures(df, num_merges=3):\n",
    "    # List to store column names for distance and s_het measures\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the diff and s_het_post column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'diff_f{i}')\n",
    "        columns_to_keep.append(f'diff_b{i}')\n",
    "        columns_to_keep.append(f's_het_post_f{i}')\n",
    "        columns_to_keep.append(f's_het_post_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "filtered_f_b_final = keep_dist_and_s_het_measures(f_b_final_merged_s_het, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9536b18f",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8efcecb5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to compute s_het weighted by 1/distance per gene-SNP pair\n",
    "def compute_weighted_s_het(df, num_merges=3):\n",
    "    # Loop to compute the weighted s_het for both forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Calculate inverse distance for forward and backward directions\n",
    "        df[f'inv_dist_f{i}'] = 1 / df[f'diff_f{i}']\n",
    "        df[f'inv_dist_b{i}'] = 1 / df[f'diff_b{i}']\n",
    "        \n",
    "        # Compute weighted s_het by multiplying s_het by inverse distance\n",
    "        df[f'weighted_s_het_f{i}'] = df[f'inv_dist_f{i}'] * df[f's_het_post_f{i}']\n",
    "        df[f'weighted_s_het_b{i}'] = df[f'inv_dist_b{i}'] * df[f's_het_post_b{i}']\n",
    "\n",
    "    return df\n",
    "\n",
    "# Example usage:\n",
    "weighted_f_b_final = compute_weighted_s_het(filtered_f_b_final, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b3df79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0b3796b6",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to keep only the weighted s_het values from the dataframe\n",
    "def keep_weighted_s_het(df, num_merges=3):\n",
    "    # List to store column names for weighted s_het values\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the weighted_s_het column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'weighted_s_het_f{i}')\n",
    "        columns_to_keep.append(f'weighted_s_het_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "weighted_s_het_only = keep_weighted_s_het(weighted_f_b_final, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f06a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_s_het_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381a5b0e-240d-4840-8232-2655389eb38c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#oliver's results \n",
    "weighted_s_het_only.iloc[np.argwhere(np.isin(weighted_s_het_only.variant, m[\"variant.mut\"].values)).flatten()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8f8edd-ed90-4661-bd3a-a067dec7af2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (Spyder)",
   "language": "python3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
