{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e5608d8",
   "metadata": {},
   "source": [
    "# Import Gene Annotation File"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fb9c920",
   "metadata": {},
   "source": [
    "https://github.com/hakha-most/gwas_eqtl/blob/master/gene_annotations/genes.protein_coding.v39.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b64193a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "\n",
    "# Local path to GTF file\n",
    "gtf_file_path = '/Users/natasharecoder/Documents/NYU/Term3/Capstone/Data/genes.protein_coding.v39.gtf'\n",
    "\n",
    "# Load the GTF file into a pandas DataFrame and use the first row as headers\n",
    "gtf_df = pd.read_csv(gtf_file_path, sep='\\t', header=0)\n",
    "\n",
    "# Display the first few rows of the dataframe\n",
    "gtf_df.sort_values('start')\n",
    "gtf_df.head()\n",
    "\n",
    "#how many closest genes we want to find\n",
    "n_closest = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "db08fb30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18332, 10)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68c645d",
   "metadata": {},
   "source": [
    "#### convert chr to int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a7fd4032",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df['chr'] = gtf_df['chr'].str[3:].astype(int)\n",
    "gtf_df['start'] = gtf_df['start'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "62061659",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>max</th>\n",
       "      <th>min</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>249200395</td>\n",
       "      <td>65419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>242836139</td>\n",
       "      <td>38814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>197687071</td>\n",
       "      <td>238446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>190945522</td>\n",
       "      <td>53180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>180794269</td>\n",
       "      <td>92283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>170884383</td>\n",
       "      <td>291630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>158820866</td>\n",
       "      <td>330136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>146277851</td>\n",
       "      <td>116086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>140772234</td>\n",
       "      <td>14475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>135437399</td>\n",
       "      <td>92828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>134248398</td>\n",
       "      <td>167784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>133794898</td>\n",
       "      <td>175933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>115079957</td>\n",
       "      <td>19747912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>105992920</td>\n",
       "      <td>19377594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>102462245</td>\n",
       "      <td>20737094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>90122974</td>\n",
       "      <td>96407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>81037485</td>\n",
       "      <td>5810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>77915115</td>\n",
       "      <td>47221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>59073298</td>\n",
       "      <td>107104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>62887048</td>\n",
       "      <td>67891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>48055507</td>\n",
       "      <td>14982098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>51205929</td>\n",
       "      <td>17071641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        max       min\n",
       "0     1  249200395     65419\n",
       "1     2  242836139     38814\n",
       "2     3  197687071    238446\n",
       "3     4  190945522     53180\n",
       "4     5  180794269     92283\n",
       "5     6  170884383    291630\n",
       "6     7  158820866    330136\n",
       "7     8  146277851    116086\n",
       "8     9  140772234     14475\n",
       "9    10  135437399     92828\n",
       "10   11  134248398    167784\n",
       "11   12  133794898    175933\n",
       "12   13  115079957  19747912\n",
       "13   14  105992920  19377594\n",
       "14   15  102462245  20737094\n",
       "15   16   90122974     96407\n",
       "16   17   81037485      5810\n",
       "17   18   77915115     47221\n",
       "18   19   59073298    107104\n",
       "19   20   62887048     67891\n",
       "20   21   48055507  14982098\n",
       "21   22   51205929  17071641"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.groupby('chr')['start'].agg({'min', 'max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e19dcc",
   "metadata": {},
   "source": [
    "Positions reset for each chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77808ff",
   "metadata": {},
   "source": [
    "### File Description\n",
    "\n",
    "This file contains **gene annotations** for **protein-coding genes** and provides information about the **location of genes on the genome**. Below is a description of each column:\n",
    "\n",
    "1. **chr**: The chromosome where the gene is located (e.g., `chr1`, `chr2`, etc.).\n",
    "2. **start**: The start position of the gene on the chromosome (in base pairs).\n",
    "3. **end**: The end position of the gene on the chromosome (in base pairs).\n",
    "4. **strand**: Indicates the strand on which the gene is located (`+` for forward strand, `-` for reverse strand).\n",
    "5. **GeneSymbol**: The symbol or name of the gene (e.g., `OR4F5`, `SAMD11`), typically assigned by organizations like HGNC (HUGO Gene Nomenclature Committee).\n",
    "6. **cons**: This column indicates the type of gene. In this file, all genes are classified as **protein_coding**.\n",
    "7. **gene**: The Ensembl gene ID, a unique identifier for the gene (e.g., `ENSG00000186092`).\n",
    "8. **hgnc_id**: The unique identifier for the gene assigned by the **HGNC** (HUGO Gene Nomenclature Committee).\n",
    "9. **tss**: The **transcription start site**, the position where transcription of the gene starts on the chromosome.\n",
    "10. **tes**: The **transcription end site**, the position where transcription of the gene ends on the chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e70a8e",
   "metadata": {},
   "source": [
    "# Import GWAS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7500998",
   "metadata": {},
   "source": [
    " https://docs.google.com/spreadsheets/d/1kvPoupSzsSFBNSztMzl04xMoSC3Kcx3CrjVf4yBmESU/edit?gid=178908679#gid=178908679 (row 7217)\n",
    " \n",
    " 50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "072a09d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Path to GWAS file\n",
    "gwas_file_path = '/Users/natasharecoder/Documents/NYU/Term3/Capstone/Data/50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz'\n",
    "\n",
    "# Open and read the compressed .bgz file using gzip\n",
    "with gzip.open(gwas_file_path, 'rt') as f:\n",
    "    # Load the file into a pandas DataFrame\n",
    "    gwas_df = pd.read_csv(f, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ab8c875a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.440760e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003474</td>\n",
       "      <td>18.049900</td>\n",
       "      <td>178.468000</td>\n",
       "      <td>0.101138</td>\n",
       "      <td>0.919441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.761760e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>-0.087536</td>\n",
       "      <td>-0.041345</td>\n",
       "      <td>0.349577</td>\n",
       "      <td>-0.118272</td>\n",
       "      <td>0.905852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.881580e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>135.620000</td>\n",
       "      <td>-2.079970</td>\n",
       "      <td>-0.046994</td>\n",
       "      <td>0.062695</td>\n",
       "      <td>-0.749571</td>\n",
       "      <td>0.453514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.674710e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>-0.105659</td>\n",
       "      <td>-0.042056</td>\n",
       "      <td>0.349588</td>\n",
       "      <td>-0.120302</td>\n",
       "      <td>0.904244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106010e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>79718.500000</td>\n",
       "      <td>101.503000</td>\n",
       "      <td>0.000797</td>\n",
       "      <td>0.002899</td>\n",
       "      <td>0.274788</td>\n",
       "      <td>0.783479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>2.454380e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>176906.000000</td>\n",
       "      <td>-787.373000</td>\n",
       "      <td>-0.001657</td>\n",
       "      <td>0.001588</td>\n",
       "      <td>-1.044020</td>\n",
       "      <td>0.296477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.297000e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>165562.000000</td>\n",
       "      <td>-734.113000</td>\n",
       "      <td>-0.001654</td>\n",
       "      <td>0.001657</td>\n",
       "      <td>-0.998454</td>\n",
       "      <td>0.318060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.393990e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>172553.000000</td>\n",
       "      <td>-480.000000</td>\n",
       "      <td>-0.001689</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>-1.011490</td>\n",
       "      <td>0.311783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>2.458510e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>177204.000000</td>\n",
       "      <td>-706.111000</td>\n",
       "      <td>-0.001520</td>\n",
       "      <td>0.001587</td>\n",
       "      <td>-0.957369</td>\n",
       "      <td>0.338381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.955520e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>14094.900000</td>\n",
       "      <td>-265.942000</td>\n",
       "      <td>-0.013297</td>\n",
       "      <td>0.005446</td>\n",
       "      <td>-2.441660</td>\n",
       "      <td>0.014620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    variant minor_allele      minor_AF  \\\n",
       "0               1:15791:C:T            T  5.440760e-09   \n",
       "1               1:69487:G:A            A  5.761760e-06   \n",
       "2               1:69569:T:C            C  1.881580e-04   \n",
       "3              1:139853:C:T            T  5.674710e-06   \n",
       "4             1:692794:CA:C            C  1.106010e-01   \n",
       "...                     ...          ...           ...   \n",
       "13791462    X:154929412:C:T            T  2.454380e-01   \n",
       "13791463   X:154929637:CT:C            C  2.297000e-01   \n",
       "13791464  X:154929952:CAA:C            C  2.393990e-01   \n",
       "13791465    X:154930230:A:G            G  2.458510e-01   \n",
       "13791466    X:154930487:T:A            A  1.955520e-02   \n",
       "\n",
       "          low_confidence_variant  n_complete_samples             AC  \\\n",
       "0                           True              360388       0.003922   \n",
       "1                           True              360388       4.152940   \n",
       "2                           True              360388     135.620000   \n",
       "3                           True              360388       4.090200   \n",
       "4                          False              360388   79718.500000   \n",
       "...                          ...                 ...            ...   \n",
       "13791462                   False              360388  176906.000000   \n",
       "13791463                   False              360388  165562.000000   \n",
       "13791464                   False              360388  172553.000000   \n",
       "13791465                   False              360388  177204.000000   \n",
       "13791466                   False              360388   14094.900000   \n",
       "\n",
       "                 ytx       beta          se     tstat      pval  \n",
       "0           0.003474  18.049900  178.468000  0.101138  0.919441  \n",
       "1          -0.087536  -0.041345    0.349577 -0.118272  0.905852  \n",
       "2          -2.079970  -0.046994    0.062695 -0.749571  0.453514  \n",
       "3          -0.105659  -0.042056    0.349588 -0.120302  0.904244  \n",
       "4         101.503000   0.000797    0.002899  0.274788  0.783479  \n",
       "...              ...        ...         ...       ...       ...  \n",
       "13791462 -787.373000  -0.001657    0.001588 -1.044020  0.296477  \n",
       "13791463 -734.113000  -0.001654    0.001657 -0.998454  0.318060  \n",
       "13791464 -480.000000  -0.001689    0.001670 -1.011490  0.311783  \n",
       "13791465 -706.111000  -0.001520    0.001587 -0.957369  0.338381  \n",
       "13791466 -265.942000  -0.013297    0.005446 -2.441660  0.014620  \n",
       "\n",
       "[13791467 rows x 11 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first few rows to inspect the structure of the GWAS file\n",
    "gwas_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1418c7f4",
   "metadata": {},
   "source": [
    "### GWAS Summary Statistics Dataset Description\n",
    "\n",
    "1. **variant**: The unique identifier for each SNP (Single Nucleotide Polymorphism). This can include information like chromosome, position, reference allele, and alternative allele (e.g., `1:12345:A:G`).\n",
    "   - **Example**: `1:10583:T:G`\n",
    "\n",
    "2. **minor_allele**: The allele that is less frequent in the population (minor allele) for this particular SNP.\n",
    "   - **Example**: `G`\n",
    "\n",
    "3. **minor_AF**: The **minor allele frequency** (AF), which represents the frequency of the minor allele in the population. It ranges from 0 to 1.\n",
    "   - **Example**: `0.35` (35% of individuals carry the minor allele)\n",
    "\n",
    "4. **low_confidence_variant**: A flag indicating whether the variant has **low confidence** due to imputation quality or other uncertainties. Values may be `TRUE` or `FALSE`.\n",
    "   - **Example**: `FALSE`\n",
    "\n",
    "5. **n_complete_samples**: The number of samples for which complete genotype data is available for this variant.\n",
    "   - **Example**: `300,000`\n",
    "\n",
    "6. **AC**: The **allele count** of the minor allele, i.e., the number of times the minor allele appears in the study population (across all samples).\n",
    "   - **Example**: `50000`\n",
    "\n",
    "7. **ytx**: Likely a placeholder for a phenotype-related statistic; depending on the dataset, this could represent something like the trait mean or effect size (its exact meaning depends on the specific analysis).\n",
    "\n",
    "8. **beta**: The **effect size** of the SNP on the trait being studied (in this case, likely height). It represents the change in the trait per additional copy of the minor allele.\n",
    "   - **Example**: `0.05` (the trait increases by 0.05 units for each additional copy of the minor allele)\n",
    "\n",
    "9. **se**: The **standard error** of the effect size (beta), indicating the precision of the estimated effect.\n",
    "   - **Example**: `0.01`\n",
    "\n",
    "10. **tstat**: The **t-statistic** for the beta estimate, which is the ratio of the beta estimate to its standard error.\n",
    "    - **Example**: `5.0` (higher values indicate more significant associations)\n",
    "\n",
    "11. **pval**: The **p-value** of the association between the SNP and the trait. This indicates the significance of the result, with smaller p-values suggesting stronger evidence that the SNP is associated with the trait.\n",
    "    - **Example**: `1.2e-6` (a very small p-value, indicating strong evidence of association)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "aae97970",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13791467, 11)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac3adf9",
   "metadata": {},
   "source": [
    "# Find Variant Chromosome and Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ca4213d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.440760e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003474</td>\n",
       "      <td>18.0499</td>\n",
       "      <td>178.468</td>\n",
       "      <td>0.101138</td>\n",
       "      <td>0.919441</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       variant minor_allele      minor_AF  low_confidence_variant  \\\n",
       "0  1:15791:C:T            T  5.440760e-09                    True   \n",
       "\n",
       "   n_complete_samples        AC       ytx     beta       se     tstat  \\\n",
       "0              360388  0.003922  0.003474  18.0499  178.468  0.101138   \n",
       "\n",
       "       pval chr    pos ref alt  \n",
       "0  0.919441   1  15791   C   T  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df[['chr', 'pos', 'ref', 'alt']] = gwas_df['variant'].str.split(':', expand=True)\n",
    "gwas_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7c596e85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/natasharecoder/anaconda3/envs/sherlock2/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/natasharecoder/anaconda3/envs/sherlock2/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "gwas_df = gwas_df[gwas_df['chr'] != 'X']\n",
    "gwas_df['chr'] = gwas_df['chr'].astype(int)\n",
    "gwas_df['pos'] = gwas_df['pos'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "330ab4d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>max</th>\n",
       "      <th>min</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>249230914</td>\n",
       "      <td>15791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>243082521</td>\n",
       "      <td>10181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>197900375</td>\n",
       "      <td>60197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>190922257</td>\n",
       "      <td>13012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>180736061</td>\n",
       "      <td>11882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>170934414</td>\n",
       "      <td>202076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>159128550</td>\n",
       "      <td>27916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>146303560</td>\n",
       "      <td>34440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>141104957</td>\n",
       "      <td>40997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>135503523</td>\n",
       "      <td>61334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>134946451</td>\n",
       "      <td>131128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>133840247</td>\n",
       "      <td>163752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>115109852</td>\n",
       "      <td>19022393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>107289436</td>\n",
       "      <td>19185163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>102461014</td>\n",
       "      <td>20001226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>90238127</td>\n",
       "      <td>60455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>81109537</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>78017073</td>\n",
       "      <td>11603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>59117442</td>\n",
       "      <td>229783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>62963628</td>\n",
       "      <td>61098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>48119634</td>\n",
       "      <td>9574185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>16464274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        max       min\n",
       "0     1  249230914     15791\n",
       "1     2  243082521     10181\n",
       "2     3  197900375     60197\n",
       "3     4  190922257     13012\n",
       "4     5  180736061     11882\n",
       "5     6  170934414    202076\n",
       "6     7  159128550     27916\n",
       "7     8  146303560     34440\n",
       "8     9  141104957     40997\n",
       "9    10  135503523     61334\n",
       "10   11  134946451    131128\n",
       "11   12  133840247    163752\n",
       "12   13  115109852  19022393\n",
       "13   14  107289436  19185163\n",
       "14   15  102461014  20001226\n",
       "15   16   90238127     60455\n",
       "16   17   81109537       302\n",
       "17   18   78017073     11603\n",
       "18   19   59117442    229783\n",
       "19   20   62963628     61098\n",
       "20   21   48119634   9574185\n",
       "21   22   51237712  16464274"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.groupby('chr')['pos'].agg({'min','max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764da1f1",
   "metadata": {},
   "source": [
    "It appears that position resets on each chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae3e953",
   "metadata": {},
   "source": [
    "# Join to Find 5 Nearest Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4f400cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#here we are performing how far the start of our mutation is from the start of the \n",
    "#following genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "5c7b3145",
   "metadata": {},
   "outputs": [],
   "source": [
    "#allow exact matches is false\n",
    "#so we are not allowing there to be a mutation that is the same as teh start of a gene\n",
    "#why\n",
    "#is it to avoid div by 0?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "32b65ecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#so this finds the three closest forward genes\n",
    "#what if all the closest are fwd vs all the closest are backward\n",
    "#should we not find teh 5 fwd and 5 bck and then compare?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "802efe59",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def repeated_forward_merge(gwas_df, gtf_df, num_merges= n_closest):\n",
    "    # Initial sorting of key columns for the first merge\n",
    "    gwas_sorted = gwas_df[['variant', 'pval', 'chr', 'pos']].sort_values('pos')\n",
    "    gtf_sorted = gtf_df[['chr', 'start', 'GeneSymbol', 'gene']].sort_values('start')\n",
    "\n",
    "    # Ensure both 'start' columns are of the same type (convert to float to handle NaNs)\n",
    "    gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "    gwas_sorted['pos'] = gwas_sorted['pos'].astype(float)\n",
    "\n",
    "    # First merge_asof\n",
    "    merged_df = pd.merge_asof(\n",
    "        gwas_sorted, \n",
    "        gtf_sorted, \n",
    "        by='chr', \n",
    "        left_on='pos', \n",
    "        right_on='start', \n",
    "        direction='forward',\n",
    "        suffixes=('', '_f1')  # First set of suffixes\n",
    "    )\n",
    "\n",
    "    # Replace NaN values in 'start' column and rename for subsequent merges\n",
    "    merged_df['start'] = merged_df['start'].fillna(np.inf)\n",
    "    merged_df = merged_df.rename(columns={'start': 'start_f1', 'GeneSymbol': 'GeneSymbol_f1', 'gene': 'gene_f1'})\n",
    "    merged_df = merged_df.sort_values('start_f1')\n",
    "\n",
    "    # Iterative forward merges\n",
    "    for i in range(2, num_merges + 1):\n",
    "        # Ensure both columns are of the same type\n",
    "        merged_df[f'start_f{i-1}'] = merged_df[f'start_f{i-1}'].astype(float)\n",
    "        gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "\n",
    "        # Perform the next forward merge\n",
    "        merged_df = pd.merge_asof(\n",
    "            merged_df, \n",
    "            gtf_sorted, \n",
    "            by='chr', \n",
    "            left_on=f'start_f{i-1}', \n",
    "            right_on='start', \n",
    "            direction='forward',\n",
    "            allow_exact_matches=False,\n",
    "            suffixes=(f'_f{i-1}', f'_f{i}')\n",
    "        )\n",
    "\n",
    "        # Replace NaN values in the new 'start' column and rename for next iteration\n",
    "        merged_df['start'] = merged_df['start'].fillna(np.inf)\n",
    "        merged_df = merged_df.rename(columns={\n",
    "            'start': f'start_f{i}', \n",
    "            'GeneSymbol': f'GeneSymbol_f{i}', \n",
    "            'gene': f'gene_f{i}'\n",
    "        })\n",
    "        merged_df = merged_df.sort_values(f'start_f{i}')\n",
    "\n",
    "    # After the last merge, replace all np.inf values back to NaN\n",
    "    for i in range(1, num_merges + 1):\n",
    "        merged_df[f'start_f{i}'] = merged_df[f'start_f{i}'].replace(np.inf, np.nan)\n",
    "    \n",
    "    return merged_df\n",
    "\n",
    "f_final = repeated_forward_merge(gwas_df, gtf_df, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "ed00ee17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>start_f1</th>\n",
       "      <th>GeneSymbol_f1</th>\n",
       "      <th>gene_f1</th>\n",
       "      <th>start_f2</th>\n",
       "      <th>GeneSymbol_f2</th>\n",
       "      <th>gene_f2</th>\n",
       "      <th>start_f3</th>\n",
       "      <th>GeneSymbol_f3</th>\n",
       "      <th>gene_f3</th>\n",
       "      <th>start_f4</th>\n",
       "      <th>GeneSymbol_f4</th>\n",
       "      <th>gene_f4</th>\n",
       "      <th>start_f5</th>\n",
       "      <th>GeneSymbol_f5</th>\n",
       "      <th>gene_f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16:94919:G:A</td>\n",
       "      <td>0.557584</td>\n",
       "      <td>16</td>\n",
       "      <td>94919.0</td>\n",
       "      <td>96407.0</td>\n",
       "      <td>ENSG00000161980</td>\n",
       "      <td>POLR3K</td>\n",
       "      <td>103828.0</td>\n",
       "      <td>ENSG00000161981</td>\n",
       "      <td>SNRNP25</td>\n",
       "      <td>108058.0</td>\n",
       "      <td>ENSG00000007384</td>\n",
       "      <td>RHBDF1</td>\n",
       "      <td>127006.0</td>\n",
       "      <td>ENSG00000103152</td>\n",
       "      <td>MPG</td>\n",
       "      <td>134270.0</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>NPRL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>16:88529:C:T</td>\n",
       "      <td>0.610190</td>\n",
       "      <td>16</td>\n",
       "      <td>88529.0</td>\n",
       "      <td>96407.0</td>\n",
       "      <td>ENSG00000161980</td>\n",
       "      <td>POLR3K</td>\n",
       "      <td>103828.0</td>\n",
       "      <td>ENSG00000161981</td>\n",
       "      <td>SNRNP25</td>\n",
       "      <td>108058.0</td>\n",
       "      <td>ENSG00000007384</td>\n",
       "      <td>RHBDF1</td>\n",
       "      <td>127006.0</td>\n",
       "      <td>ENSG00000103152</td>\n",
       "      <td>MPG</td>\n",
       "      <td>134270.0</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>NPRL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>16:94585:AGATCAGGAGGCT:A</td>\n",
       "      <td>0.641873</td>\n",
       "      <td>16</td>\n",
       "      <td>94585.0</td>\n",
       "      <td>96407.0</td>\n",
       "      <td>ENSG00000161980</td>\n",
       "      <td>POLR3K</td>\n",
       "      <td>103828.0</td>\n",
       "      <td>ENSG00000161981</td>\n",
       "      <td>SNRNP25</td>\n",
       "      <td>108058.0</td>\n",
       "      <td>ENSG00000007384</td>\n",
       "      <td>RHBDF1</td>\n",
       "      <td>127006.0</td>\n",
       "      <td>ENSG00000103152</td>\n",
       "      <td>MPG</td>\n",
       "      <td>134270.0</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>NPRL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>16:83887:C:A</td>\n",
       "      <td>0.955567</td>\n",
       "      <td>16</td>\n",
       "      <td>83887.0</td>\n",
       "      <td>96407.0</td>\n",
       "      <td>ENSG00000161980</td>\n",
       "      <td>POLR3K</td>\n",
       "      <td>103828.0</td>\n",
       "      <td>ENSG00000161981</td>\n",
       "      <td>SNRNP25</td>\n",
       "      <td>108058.0</td>\n",
       "      <td>ENSG00000007384</td>\n",
       "      <td>RHBDF1</td>\n",
       "      <td>127006.0</td>\n",
       "      <td>ENSG00000103152</td>\n",
       "      <td>MPG</td>\n",
       "      <td>134270.0</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>NPRL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>16:89077:G:A</td>\n",
       "      <td>0.477961</td>\n",
       "      <td>16</td>\n",
       "      <td>89077.0</td>\n",
       "      <td>96407.0</td>\n",
       "      <td>ENSG00000161980</td>\n",
       "      <td>POLR3K</td>\n",
       "      <td>103828.0</td>\n",
       "      <td>ENSG00000161981</td>\n",
       "      <td>SNRNP25</td>\n",
       "      <td>108058.0</td>\n",
       "      <td>ENSG00000007384</td>\n",
       "      <td>RHBDF1</td>\n",
       "      <td>127006.0</td>\n",
       "      <td>ENSG00000103152</td>\n",
       "      <td>MPG</td>\n",
       "      <td>134270.0</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>NPRL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13324773</th>\n",
       "      <td>4:189124592:T:C</td>\n",
       "      <td>0.146931</td>\n",
       "      <td>4</td>\n",
       "      <td>189124592.0</td>\n",
       "      <td>190862025.0</td>\n",
       "      <td>ENSG00000109536</td>\n",
       "      <td>FRG1</td>\n",
       "      <td>190945522.0</td>\n",
       "      <td>ENSG00000205097</td>\n",
       "      <td>FRG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13324774</th>\n",
       "      <td>4:190843675:C:T</td>\n",
       "      <td>0.550784</td>\n",
       "      <td>4</td>\n",
       "      <td>190843675.0</td>\n",
       "      <td>190862025.0</td>\n",
       "      <td>ENSG00000109536</td>\n",
       "      <td>FRG1</td>\n",
       "      <td>190945522.0</td>\n",
       "      <td>ENSG00000205097</td>\n",
       "      <td>FRG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13324775</th>\n",
       "      <td>4:189124184:GT:G</td>\n",
       "      <td>0.708610</td>\n",
       "      <td>4</td>\n",
       "      <td>189124184.0</td>\n",
       "      <td>190862025.0</td>\n",
       "      <td>ENSG00000109536</td>\n",
       "      <td>FRG1</td>\n",
       "      <td>190945522.0</td>\n",
       "      <td>ENSG00000205097</td>\n",
       "      <td>FRG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13324777</th>\n",
       "      <td>4:189109193:T:C</td>\n",
       "      <td>0.162214</td>\n",
       "      <td>4</td>\n",
       "      <td>189109193.0</td>\n",
       "      <td>190862025.0</td>\n",
       "      <td>ENSG00000109536</td>\n",
       "      <td>FRG1</td>\n",
       "      <td>190945522.0</td>\n",
       "      <td>ENSG00000205097</td>\n",
       "      <td>FRG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>0.459274</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           variant      pval  chr          pos     start_f1  \\\n",
       "0                     16:94919:G:A  0.557584   16      94919.0      96407.0   \n",
       "72                    16:88529:C:T  0.610190   16      88529.0      96407.0   \n",
       "71        16:94585:AGATCAGGAGGCT:A  0.641873   16      94585.0      96407.0   \n",
       "70                    16:83887:C:A  0.955567   16      83887.0      96407.0   \n",
       "69                    16:89077:G:A  0.477961   16      89077.0      96407.0   \n",
       "...                            ...       ...  ...          ...          ...   \n",
       "13324773           4:189124592:T:C  0.146931    4  189124592.0  190862025.0   \n",
       "13324774           4:190843675:C:T  0.550784    4  190843675.0  190862025.0   \n",
       "13324775          4:189124184:GT:G  0.708610    4  189124184.0  190862025.0   \n",
       "13324777           4:189109193:T:C  0.162214    4  189109193.0  190862025.0   \n",
       "13364302           1:249230914:C:T  0.459274    1  249230914.0          NaN   \n",
       "\n",
       "            GeneSymbol_f1 gene_f1     start_f2    GeneSymbol_f2  gene_f2  \\\n",
       "0         ENSG00000161980  POLR3K     103828.0  ENSG00000161981  SNRNP25   \n",
       "72        ENSG00000161980  POLR3K     103828.0  ENSG00000161981  SNRNP25   \n",
       "71        ENSG00000161980  POLR3K     103828.0  ENSG00000161981  SNRNP25   \n",
       "70        ENSG00000161980  POLR3K     103828.0  ENSG00000161981  SNRNP25   \n",
       "69        ENSG00000161980  POLR3K     103828.0  ENSG00000161981  SNRNP25   \n",
       "...                   ...     ...          ...              ...      ...   \n",
       "13324773  ENSG00000109536    FRG1  190945522.0  ENSG00000205097     FRG2   \n",
       "13324774  ENSG00000109536    FRG1  190945522.0  ENSG00000205097     FRG2   \n",
       "13324775  ENSG00000109536    FRG1  190945522.0  ENSG00000205097     FRG2   \n",
       "13324777  ENSG00000109536    FRG1  190945522.0  ENSG00000205097     FRG2   \n",
       "13364302              NaN     NaN          NaN              NaN      NaN   \n",
       "\n",
       "          start_f3    GeneSymbol_f3 gene_f3  start_f4    GeneSymbol_f4  \\\n",
       "0         108058.0  ENSG00000007384  RHBDF1  127006.0  ENSG00000103152   \n",
       "72        108058.0  ENSG00000007384  RHBDF1  127006.0  ENSG00000103152   \n",
       "71        108058.0  ENSG00000007384  RHBDF1  127006.0  ENSG00000103152   \n",
       "70        108058.0  ENSG00000007384  RHBDF1  127006.0  ENSG00000103152   \n",
       "69        108058.0  ENSG00000007384  RHBDF1  127006.0  ENSG00000103152   \n",
       "...            ...              ...     ...       ...              ...   \n",
       "13324773       NaN              NaN     NaN       NaN              NaN   \n",
       "13324774       NaN              NaN     NaN       NaN              NaN   \n",
       "13324775       NaN              NaN     NaN       NaN              NaN   \n",
       "13324777       NaN              NaN     NaN       NaN              NaN   \n",
       "13364302       NaN              NaN     NaN       NaN              NaN   \n",
       "\n",
       "         gene_f4  start_f5    GeneSymbol_f5 gene_f5  \n",
       "0            MPG  134270.0  ENSG00000103148   NPRL3  \n",
       "72           MPG  134270.0  ENSG00000103148   NPRL3  \n",
       "71           MPG  134270.0  ENSG00000103148   NPRL3  \n",
       "70           MPG  134270.0  ENSG00000103148   NPRL3  \n",
       "69           MPG  134270.0  ENSG00000103148   NPRL3  \n",
       "...          ...       ...              ...     ...  \n",
       "13324773     NaN       NaN              NaN     NaN  \n",
       "13324774     NaN       NaN              NaN     NaN  \n",
       "13324775     NaN       NaN              NaN     NaN  \n",
       "13324777     NaN       NaN              NaN     NaN  \n",
       "13364302     NaN       NaN              NaN     NaN  \n",
       "\n",
       "[13364303 rows x 19 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "2a8cef2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjusted function to ensure all negative infinity values are replaced with NaN in the final result\n",
    "def repeated_backward_merge_with_nan(gwas_df, gtf_df, num_merges=n_closest):\n",
    "    # Initial sorting of key columns for the first merge\n",
    "    gwas_sorted = gwas_df.sort_values('pos')\n",
    "    gtf_sorted = gtf_df[['chr', 'start', 'GeneSymbol', 'gene']].sort_values('start')\n",
    "\n",
    "    # Ensure both 'start' columns are of the same type (convert to float to handle NaNs)\n",
    "    gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "    gwas_sorted['pos'] = gwas_sorted['pos'].astype(float)\n",
    "\n",
    "    # First backward merge_asof\n",
    "    merged_df = pd.merge_asof(\n",
    "        gwas_sorted, \n",
    "        gtf_sorted, \n",
    "        by='chr', \n",
    "        left_on='pos', \n",
    "        right_on='start', \n",
    "        direction='backward',\n",
    "        suffixes=('', '_b1')  # First set of suffixes\n",
    "    )\n",
    "\n",
    "    # Replace NaN values in 'start' column and rename for subsequent merges\n",
    "    merged_df['start'] = merged_df['start'].fillna(-np.inf)\n",
    "    merged_df = merged_df.rename(columns={'start': 'start_b1', 'GeneSymbol': 'GeneSymbol_b1', 'gene': 'gene_b1'})\n",
    "    merged_df = merged_df.sort_values('start_b1')\n",
    "\n",
    "    # Iterative backward merges\n",
    "    for i in range(2, num_merges + 1):\n",
    "        # Ensure both columns are of the same type\n",
    "        merged_df[f'start_b{i-1}'] = merged_df[f'start_b{i-1}'].astype(float)\n",
    "        gtf_sorted['start'] = gtf_sorted['start'].astype(float)\n",
    "\n",
    "        # Perform the next backward merge\n",
    "        merged_df = pd.merge_asof(\n",
    "            merged_df, \n",
    "            gtf_sorted, \n",
    "            by='chr', \n",
    "            left_on=f'start_b{i-1}', \n",
    "            right_on='start', \n",
    "            direction='backward',\n",
    "            allow_exact_matches=False,\n",
    "            suffixes=(f'_b{i-1}', f'_b{i}')\n",
    "        )\n",
    "\n",
    "        # Replace NaN values in the new 'start' column and rename for next iteration\n",
    "        merged_df['start'] = merged_df['start'].fillna(-np.inf)\n",
    "        merged_df = merged_df.rename(columns={\n",
    "            'start': f'start_b{i}', \n",
    "            'GeneSymbol': f'GeneSymbol_b{i}', \n",
    "            'gene': f'gene_b{i}'\n",
    "        })\n",
    "        merged_df = merged_df.sort_values(f'start_b{i}')\n",
    "\n",
    "    # After the last merge, replace all -np.inf values back to NaN\n",
    "    for i in range(1, num_merges + 1):\n",
    "        merged_df[f'start_b{i}'] = merged_df[f'start_b{i}'].replace(-np.inf, np.nan)\n",
    "    \n",
    "    return merged_df\n",
    "\n",
    "# Example usage:\n",
    "f_b_final = repeated_backward_merge_with_nan(f_final, gtf_df, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a343494d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>start_f1</th>\n",
       "      <th>GeneSymbol_f1</th>\n",
       "      <th>gene_f1</th>\n",
       "      <th>start_f2</th>\n",
       "      <th>GeneSymbol_f2</th>\n",
       "      <th>gene_f2</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_b2</th>\n",
       "      <th>start_b3</th>\n",
       "      <th>GeneSymbol_b3</th>\n",
       "      <th>gene_b3</th>\n",
       "      <th>start_b4</th>\n",
       "      <th>GeneSymbol_b4</th>\n",
       "      <th>gene_b4</th>\n",
       "      <th>start_b5</th>\n",
       "      <th>GeneSymbol_b5</th>\n",
       "      <th>gene_b5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17:302:T:TA</td>\n",
       "      <td>0.281579</td>\n",
       "      <td>17</td>\n",
       "      <td>302.0</td>\n",
       "      <td>5810.0</td>\n",
       "      <td>ENSG00000272636</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>62180.0</td>\n",
       "      <td>ENSG00000181031</td>\n",
       "      <td>RPH3AL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50036</th>\n",
       "      <td>21:15357076:C:A</td>\n",
       "      <td>0.573263</td>\n",
       "      <td>21</td>\n",
       "      <td>15357076.0</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50035</th>\n",
       "      <td>21:15357104:C:G</td>\n",
       "      <td>0.572347</td>\n",
       "      <td>21</td>\n",
       "      <td>15357104.0</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50034</th>\n",
       "      <td>21:15366358:A:G</td>\n",
       "      <td>0.789949</td>\n",
       "      <td>21</td>\n",
       "      <td>15366358.0</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50033</th>\n",
       "      <td>21:15212919:G:C</td>\n",
       "      <td>0.094328</td>\n",
       "      <td>21</td>\n",
       "      <td>15212919.0</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364242</th>\n",
       "      <td>1:249205823:G:A</td>\n",
       "      <td>0.918853</td>\n",
       "      <td>1</td>\n",
       "      <td>249205823.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249132409.0</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>ZNF672</td>\n",
       "      <td>249104645.0</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>OR14I1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364241</th>\n",
       "      <td>1:249205806:T:C</td>\n",
       "      <td>0.989577</td>\n",
       "      <td>1</td>\n",
       "      <td>249205806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249132409.0</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>ZNF672</td>\n",
       "      <td>249104645.0</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>OR14I1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364240</th>\n",
       "      <td>1:249205774:T:C</td>\n",
       "      <td>0.767464</td>\n",
       "      <td>1</td>\n",
       "      <td>249205774.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249132409.0</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>ZNF672</td>\n",
       "      <td>249104645.0</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>OR14I1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364248</th>\n",
       "      <td>1:249205246:T:A</td>\n",
       "      <td>0.937160</td>\n",
       "      <td>1</td>\n",
       "      <td>249205246.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249132409.0</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>ZNF672</td>\n",
       "      <td>249104645.0</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>OR14I1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>0.459274</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249132409.0</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>ZNF672</td>\n",
       "      <td>249104645.0</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>OR14I1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant      pval  chr          pos    start_f1  \\\n",
       "0             17:302:T:TA  0.281579   17        302.0      5810.0   \n",
       "50036     21:15357076:C:A  0.573263   21   15357076.0  15481134.0   \n",
       "50035     21:15357104:C:G  0.572347   21   15357104.0  15481134.0   \n",
       "50034     21:15366358:A:G  0.789949   21   15366358.0  15481134.0   \n",
       "50033     21:15212919:G:C  0.094328   21   15212919.0  15481134.0   \n",
       "...                   ...       ...  ...          ...         ...   \n",
       "13364242  1:249205823:G:A  0.918853    1  249205823.0         NaN   \n",
       "13364241  1:249205806:T:C  0.989577    1  249205806.0         NaN   \n",
       "13364240  1:249205774:T:C  0.767464    1  249205774.0         NaN   \n",
       "13364248  1:249205246:T:A  0.937160    1  249205246.0         NaN   \n",
       "13364302  1:249230914:C:T  0.459274    1  249230914.0         NaN   \n",
       "\n",
       "            GeneSymbol_f1 gene_f1    start_f2    GeneSymbol_f2 gene_f2  ...  \\\n",
       "0         ENSG00000272636   DOC2B     62180.0  ENSG00000181031  RPH3AL  ...   \n",
       "50036     ENSG00000188992    LIPI  15588451.0  ENSG00000185272   RBM11  ...   \n",
       "50035     ENSG00000188992    LIPI  15588451.0  ENSG00000185272   RBM11  ...   \n",
       "50034     ENSG00000188992    LIPI  15588451.0  ENSG00000185272   RBM11  ...   \n",
       "50033     ENSG00000188992    LIPI  15588451.0  ENSG00000185272   RBM11  ...   \n",
       "...                   ...     ...         ...              ...     ...  ...   \n",
       "13364242              NaN     NaN         NaN              NaN     NaN  ...   \n",
       "13364241              NaN     NaN         NaN              NaN     NaN  ...   \n",
       "13364240              NaN     NaN         NaN              NaN     NaN  ...   \n",
       "13364248              NaN     NaN         NaN              NaN     NaN  ...   \n",
       "13364302              NaN     NaN         NaN              NaN     NaN  ...   \n",
       "\n",
       "          gene_b2     start_b3    GeneSymbol_b3  gene_b3     start_b4  \\\n",
       "0             NaN          NaN              NaN      NaN          NaN   \n",
       "50036         NaN          NaN              NaN      NaN          NaN   \n",
       "50035         NaN          NaN              NaN      NaN          NaN   \n",
       "50034         NaN          NaN              NaN      NaN          NaN   \n",
       "50033         NaN          NaN              NaN      NaN          NaN   \n",
       "...           ...          ...              ...      ...          ...   \n",
       "13364242   ZNF692  249132409.0  ENSG00000171161   ZNF672  249104645.0   \n",
       "13364241   ZNF692  249132409.0  ENSG00000171161   ZNF672  249104645.0   \n",
       "13364240   ZNF692  249132409.0  ENSG00000171161   ZNF672  249104645.0   \n",
       "13364248   ZNF692  249132409.0  ENSG00000171161   ZNF672  249104645.0   \n",
       "13364302   ZNF692  249132409.0  ENSG00000171161   ZNF672  249104645.0   \n",
       "\n",
       "            GeneSymbol_b4  gene_b4     start_b5    GeneSymbol_b5  gene_b5  \n",
       "0                     NaN      NaN          NaN              NaN      NaN  \n",
       "50036                 NaN      NaN          NaN              NaN      NaN  \n",
       "50035                 NaN      NaN          NaN              NaN      NaN  \n",
       "50034                 NaN      NaN          NaN              NaN      NaN  \n",
       "50033                 NaN      NaN          NaN              NaN      NaN  \n",
       "...                   ...      ...          ...              ...      ...  \n",
       "13364242  ENSG00000175137  SH3BP5L  248844623.0  ENSG00000189181   OR14I1  \n",
       "13364241  ENSG00000175137  SH3BP5L  248844623.0  ENSG00000189181   OR14I1  \n",
       "13364240  ENSG00000175137  SH3BP5L  248844623.0  ENSG00000189181   OR14I1  \n",
       "13364248  ENSG00000175137  SH3BP5L  248844623.0  ENSG00000189181   OR14I1  \n",
       "13364302  ENSG00000175137  SH3BP5L  248844623.0  ENSG00000189181   OR14I1  \n",
       "\n",
       "[13364303 rows x 34 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "083587df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to calculate differences between pos and all forward and backward start columns\n",
    "def calculate_differences(df, num_merges=n_closest):\n",
    "    # Calculate the difference between pos and start_f1, start_f2, ..., start_fx for forward merges\n",
    "    for i in range(1, num_merges + 1):\n",
    "        df[f'diff_f{i}'] = np.abs(df[f'start_f{i}'] - df['pos'])\n",
    "    \n",
    "    # Calculate the difference between pos and start_b1, start_b2, ..., start_bx for backward merges\n",
    "    for i in range(1, num_merges + 1):\n",
    "        df[f'diff_b{i}'] = np.abs(df[f'start_b{i}'] - df['pos'])\n",
    "    \n",
    "    return df\n",
    "\n",
    "f_b_final_with_differences = calculate_differences(f_b_final, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "74a04d23",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "f_b_final_with_differences = f_b_final_with_differences.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c0667913",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "diff_f1      54818.0\n",
       "diff_f2     154693.0\n",
       "diff_f3     216769.0\n",
       "diff_f4     270874.0\n",
       "diff_f5     281997.0\n",
       "diff_b1     154192.0\n",
       "diff_b2     247232.0\n",
       "diff_b3    3718470.0\n",
       "diff_b4    4518184.0\n",
       "diff_b5    4589502.0\n",
       "Name: 364242, dtype: object"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_b_final_with_differences.iloc[364242].loc[['diff_f1', 'diff_f2', 'diff_f3', 'diff_f4', 'diff_f5', 'diff_b1', 'diff_b2', 'diff_b3', 'diff_b4', 'diff_b5']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "cb3b90fb",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "diff_f1    4.507356e+05\n",
       "diff_f2    8.421126e+05\n",
       "diff_f3    1.120898e+06\n",
       "diff_b1   -4.557157e+05\n",
       "diff_b2   -8.418453e+05\n",
       "diff_b3   -1.118083e+06\n",
       "dtype: float64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_b_final_with_differences[['diff_f1', 'diff_f2', 'diff_f3', 'diff_b1', 'diff_b2', 'diff_b3']].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6db6e042",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vectorized_reorder_rename_columns(df):\n",
    "    # Get all diff columns (assuming 5 forward and 5 backward)\n",
    "    diff_columns = [col for col in df.columns if col.startswith('diff_')]\n",
    "\n",
    "    # Create a DataFrame of just the diff values\n",
    "    diff_df = df[diff_columns]\n",
    "    \n",
    "    # Get the indices of the 5 smallest diffs for each row\n",
    "    diff_order = diff_df.values.argsort(axis=1)[:, :n_closest]\n",
    "    \n",
    "    # Function to reorder and rename a set of columns\n",
    "    def reorder_and_rename(prefix):\n",
    "        cols = [col.replace('diff_', prefix) for col in diff_columns]\n",
    "        values = df[cols].values\n",
    "        reordered = np.take_along_axis(values, diff_order, axis=1)\n",
    "        return pd.DataFrame(reordered, columns=[f'{prefix[:-1]}_{i}' for i in range(1, n_closest+1)], index=df.index)\n",
    "\n",
    "    # Reorder and rename for each set of columns\n",
    "    new_dfs = [reorder_and_rename(prefix) for prefix in ['diff_', 'gene_', 'start_', 'GeneSymbol_']]\n",
    "    \n",
    "    # Create direction columns\n",
    "    ordered_diff_columns = np.array(diff_columns)[diff_order]\n",
    "    #something wrong here\n",
    "    #directions = np.where(np.char.find(ordered_diff_columns.astype(str), 'f') != -1, 'f', 'b')\n",
    "    directions = np.where(\n",
    "    np.char.find(ordered_diff_columns.astype(str), 'diff_f') != -1,  # Check for 'diff_f' in the column names\n",
    "    'f',  # Set direction as 'f'\n",
    "    'b'   # Otherwise, set direction as 'b'\n",
    ")\n",
    "    direction_df = pd.DataFrame(directions, columns=[f'direction_{i}' for i in range(1, n_closest+1)], index=df.index)\n",
    "    new_dfs.append(direction_df)\n",
    "    \n",
    "    # Combine all new DataFrames\n",
    "    result_df = pd.concat(new_dfs, axis=1)\n",
    "    \n",
    "    # Add any other columns that weren't part of the reordering\n",
    "    other_columns = [col for col in df.columns if not any(col.startswith(prefix) for prefix in ['diff_', 'gene_', 'start_', 'GeneSymbol_'])]\n",
    "    result_df = pd.concat([result_df, df[other_columns]], axis=1)\n",
    "    \n",
    "    # Reorder columns to group related information together\n",
    "    column_order = (\n",
    "        [f'{prefix}_{i}' for i in range(1, n_closest+1) for prefix in ['diff', 'gene', 'start', 'GeneSymbol', 'direction']] +\n",
    "        other_columns\n",
    "    )\n",
    "    \n",
    "    return result_df[column_order]\n",
    "\n",
    "f_b_final_with_differences = vectorized_reorder_rename_columns(f_b_final_with_differences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "a210892f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diff_1</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>start_1</th>\n",
       "      <th>GeneSymbol_1</th>\n",
       "      <th>direction_1</th>\n",
       "      <th>diff_2</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>GeneSymbol_2</th>\n",
       "      <th>direction_2</th>\n",
       "      <th>...</th>\n",
       "      <th>direction_4</th>\n",
       "      <th>diff_5</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>start_5</th>\n",
       "      <th>GeneSymbol_5</th>\n",
       "      <th>direction_5</th>\n",
       "      <th>variant</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5508.0</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>5810.0</td>\n",
       "      <td>ENSG00000272636</td>\n",
       "      <td>f</td>\n",
       "      <td>61878.0</td>\n",
       "      <td>RPH3AL</td>\n",
       "      <td>62180.0</td>\n",
       "      <td>ENSG00000181031</td>\n",
       "      <td>f</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>411441.0</td>\n",
       "      <td>VPS53</td>\n",
       "      <td>411743.0</td>\n",
       "      <td>ENSG00000141252</td>\n",
       "      <td>f</td>\n",
       "      <td>17:302:T:TA</td>\n",
       "      <td>0.281579</td>\n",
       "      <td>17</td>\n",
       "      <td>302.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>124058.0</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>f</td>\n",
       "      <td>231375.0</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>f</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>500473.0</td>\n",
       "      <td>SAMSN1</td>\n",
       "      <td>15857549.0</td>\n",
       "      <td>ENSG00000155307</td>\n",
       "      <td>f</td>\n",
       "      <td>21:15357076:C:A</td>\n",
       "      <td>0.573263</td>\n",
       "      <td>21</td>\n",
       "      <td>15357076.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>124030.0</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>f</td>\n",
       "      <td>231347.0</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>f</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>500445.0</td>\n",
       "      <td>SAMSN1</td>\n",
       "      <td>15857549.0</td>\n",
       "      <td>ENSG00000155307</td>\n",
       "      <td>f</td>\n",
       "      <td>21:15357104:C:G</td>\n",
       "      <td>0.572347</td>\n",
       "      <td>21</td>\n",
       "      <td>15357104.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>114776.0</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>f</td>\n",
       "      <td>222093.0</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>15588451.0</td>\n",
       "      <td>ENSG00000185272</td>\n",
       "      <td>f</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>491191.0</td>\n",
       "      <td>SAMSN1</td>\n",
       "      <td>15857549.0</td>\n",
       "      <td>ENSG00000155307</td>\n",
       "      <td>f</td>\n",
       "      <td>21:15366358:A:G</td>\n",
       "      <td>0.789949</td>\n",
       "      <td>21</td>\n",
       "      <td>15366358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230821.0</td>\n",
       "      <td>POTED</td>\n",
       "      <td>14982098.0</td>\n",
       "      <td>ENSG00000166351</td>\n",
       "      <td>b</td>\n",
       "      <td>268215.0</td>\n",
       "      <td>LIPI</td>\n",
       "      <td>15481134.0</td>\n",
       "      <td>ENSG00000188992</td>\n",
       "      <td>f</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>644630.0</td>\n",
       "      <td>SAMSN1</td>\n",
       "      <td>15857549.0</td>\n",
       "      <td>ENSG00000155307</td>\n",
       "      <td>f</td>\n",
       "      <td>21:15212919:G:C</td>\n",
       "      <td>0.094328</td>\n",
       "      <td>21</td>\n",
       "      <td>15212919.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364298</th>\n",
       "      <td>5428.0</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>249200395.0</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>b</td>\n",
       "      <td>61618.0</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249144205.0</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>b</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>361200.0</td>\n",
       "      <td>OR14I1</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>b</td>\n",
       "      <td>1:249205823:G:A</td>\n",
       "      <td>0.918853</td>\n",
       "      <td>1</td>\n",
       "      <td>249205823.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364299</th>\n",
       "      <td>5411.0</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>249200395.0</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>b</td>\n",
       "      <td>61601.0</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249144205.0</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>b</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>361183.0</td>\n",
       "      <td>OR14I1</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>b</td>\n",
       "      <td>1:249205806:T:C</td>\n",
       "      <td>0.989577</td>\n",
       "      <td>1</td>\n",
       "      <td>249205806.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364300</th>\n",
       "      <td>5379.0</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>249200395.0</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>b</td>\n",
       "      <td>61569.0</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249144205.0</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>b</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>361151.0</td>\n",
       "      <td>OR14I1</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>b</td>\n",
       "      <td>1:249205774:T:C</td>\n",
       "      <td>0.767464</td>\n",
       "      <td>1</td>\n",
       "      <td>249205774.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364301</th>\n",
       "      <td>4851.0</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>249200395.0</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>b</td>\n",
       "      <td>61041.0</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249144205.0</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>b</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>360623.0</td>\n",
       "      <td>OR14I1</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>b</td>\n",
       "      <td>1:249205246:T:A</td>\n",
       "      <td>0.937160</td>\n",
       "      <td>1</td>\n",
       "      <td>249205246.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>30519.0</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>249200395.0</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>b</td>\n",
       "      <td>86709.0</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>249144205.0</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>b</td>\n",
       "      <td>...</td>\n",
       "      <td>b</td>\n",
       "      <td>386291.0</td>\n",
       "      <td>OR14I1</td>\n",
       "      <td>248844623.0</td>\n",
       "      <td>ENSG00000189181</td>\n",
       "      <td>b</td>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>0.459274</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            diff_1 gene_1      start_1     GeneSymbol_1 direction_1    diff_2  \\\n",
       "0           5508.0  DOC2B       5810.0  ENSG00000272636           f   61878.0   \n",
       "1         124058.0   LIPI   15481134.0  ENSG00000188992           f  231375.0   \n",
       "2         124030.0   LIPI   15481134.0  ENSG00000188992           f  231347.0   \n",
       "3         114776.0   LIPI   15481134.0  ENSG00000188992           f  222093.0   \n",
       "4         230821.0  POTED   14982098.0  ENSG00000166351           b  268215.0   \n",
       "...            ...    ...          ...              ...         ...       ...   \n",
       "13364298    5428.0  PGBD2  249200395.0  ENSG00000185220           b   61618.0   \n",
       "13364299    5411.0  PGBD2  249200395.0  ENSG00000185220           b   61601.0   \n",
       "13364300    5379.0  PGBD2  249200395.0  ENSG00000185220           b   61569.0   \n",
       "13364301    4851.0  PGBD2  249200395.0  ENSG00000185220           b   61041.0   \n",
       "13364302   30519.0  PGBD2  249200395.0  ENSG00000185220           b   86709.0   \n",
       "\n",
       "          gene_2      start_2     GeneSymbol_2 direction_2  ...  direction_4  \\\n",
       "0         RPH3AL      62180.0  ENSG00000181031           f  ...            f   \n",
       "1          RBM11   15588451.0  ENSG00000185272           f  ...            f   \n",
       "2          RBM11   15588451.0  ENSG00000185272           f  ...            f   \n",
       "3          RBM11   15588451.0  ENSG00000185272           f  ...            b   \n",
       "4           LIPI   15481134.0  ENSG00000188992           f  ...            f   \n",
       "...          ...          ...              ...         ...  ...          ...   \n",
       "13364298  ZNF692  249144205.0  ENSG00000171163           b  ...            b   \n",
       "13364299  ZNF692  249144205.0  ENSG00000171163           b  ...            b   \n",
       "13364300  ZNF692  249144205.0  ENSG00000171163           b  ...            b   \n",
       "13364301  ZNF692  249144205.0  ENSG00000171163           b  ...            b   \n",
       "13364302  ZNF692  249144205.0  ENSG00000171163           b  ...            b   \n",
       "\n",
       "            diff_5  gene_5      start_5     GeneSymbol_5  direction_5  \\\n",
       "0         411441.0   VPS53     411743.0  ENSG00000141252            f   \n",
       "1         500473.0  SAMSN1   15857549.0  ENSG00000155307            f   \n",
       "2         500445.0  SAMSN1   15857549.0  ENSG00000155307            f   \n",
       "3         491191.0  SAMSN1   15857549.0  ENSG00000155307            f   \n",
       "4         644630.0  SAMSN1   15857549.0  ENSG00000155307            f   \n",
       "...            ...     ...          ...              ...          ...   \n",
       "13364298  361200.0  OR14I1  248844623.0  ENSG00000189181            b   \n",
       "13364299  361183.0  OR14I1  248844623.0  ENSG00000189181            b   \n",
       "13364300  361151.0  OR14I1  248844623.0  ENSG00000189181            b   \n",
       "13364301  360623.0  OR14I1  248844623.0  ENSG00000189181            b   \n",
       "13364302  386291.0  OR14I1  248844623.0  ENSG00000189181            b   \n",
       "\n",
       "                  variant      pval chr          pos  \n",
       "0             17:302:T:TA  0.281579  17        302.0  \n",
       "1         21:15357076:C:A  0.573263  21   15357076.0  \n",
       "2         21:15357104:C:G  0.572347  21   15357104.0  \n",
       "3         21:15366358:A:G  0.789949  21   15366358.0  \n",
       "4         21:15212919:G:C  0.094328  21   15212919.0  \n",
       "...                   ...       ...  ..          ...  \n",
       "13364298  1:249205823:G:A  0.918853   1  249205823.0  \n",
       "13364299  1:249205806:T:C  0.989577   1  249205806.0  \n",
       "13364300  1:249205774:T:C  0.767464   1  249205774.0  \n",
       "13364301  1:249205246:T:A  0.937160   1  249205246.0  \n",
       "13364302  1:249230914:C:T  0.459274   1  249230914.0  \n",
       "\n",
       "[13364303 rows x 29 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_b_final_with_differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "29aea24e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'diff_1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/envs/sherlock2/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3360\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3361\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sherlock2/lib/python3.7/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sherlock2/lib/python3.7/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'diff_1'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/jk/c8gg6y6d0q7c296yj5mrfzlh0000gn/T/ipykernel_92837/383488426.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;31m# Call the function to plot histograms\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0mplot_histograms\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf_b_final_with_differences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns_to_plot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/jk/c8gg6y6d0q7c296yj5mrfzlh0000gn/T/ipykernel_92837/383488426.py\u001b[0m in \u001b[0;36mplot_histograms\u001b[0;34m(df, columns)\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Create a subplot for each column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m         \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'skyblue'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0medgecolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'black'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Distribution of {col}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxlabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Value'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sherlock2/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3456\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3457\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3458\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3459\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sherlock2/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3361\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3363\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3365\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'diff_1'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAGPCAYAAABbFLtAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAZ00lEQVR4nO3df2yV5d3H8c9pS0+RrccIWgrUWhxolYijDZWyxuiwBgyGxIUaFgoMExt1/OhwUruIEJNGF8lEaf3VSkyqawAh/NEh548Nyo/9oGuNsU00wmzR1qY1nhZ1Rcr1/MHTk+fYgr1Pzyn0+b5fyfnjXLvuc65r3d7evXvn1ueccwIAmJFwpRcAABhbhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCM8Rz+I0eOaOnSpZo2bZp8Pp/279//o8ccPnxYOTk5SklJ0cyZM/Xqq69Gs1YAQAx4Dv8333yjuXPn6pVXXhnR/NOnT2vJkiUqKChQU1OTnn76aa1bt0579+71vFgAwOj5RvOQNp/Pp3379mnZsmWXnPPUU0/pwIEDam1tDY+VlJTogw8+0IkTJ6L9agBAlJLi/QUnTpxQYWFhxNj999+v6upqff/995owYcKQY/r7+9Xf3x9+f+HCBX311VeaPHmyfD5fvJcMAFcF55z6+vo0bdo0JSTE7k+ycQ9/Z2en0tLSIsbS0tJ0/vx5dXd3Kz09fcgxFRUV2rp1a7yXBgDjQnt7u2bMmBGzz4t7+CUNOUsfvLp0qbP3srIylZaWht+HQiHdeOONam9vV2pqavwWCgBXkd7eXmVkZOinP/1pTD837uGfOnWqOjs7I8a6urqUlJSkyZMnD3uM3++X3+8fMp6amkr4AZgT60vccb+Pf8GCBQoGgxFjhw4dUm5u7rDX9wEA8eU5/GfPnlVzc7Oam5slXbxds7m5WW1tbZIuXqYpLi4Ozy8pKdFnn32m0tJStba2qqamRtXV1dq0aVNsdgAA8MTzpZ6TJ0/qnnvuCb8fvBa/atUq7dq1Sx0dHeF/CEhSVlaW6uvrtXHjRu3cuVPTpk3Tjh079NBDD8Vg+QAAr0Z1H/9Y6e3tVSAQUCgU4ho/ADPi1T6e1QMAxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxUYW/srJSWVlZSklJUU5OjhoaGi47v7a2VnPnztU111yj9PR0rVmzRj09PVEtGAAwOp7DX1dXpw0bNqi8vFxNTU0qKCjQ4sWL1dbWNuz8o0ePqri4WGvXrtVHH32k3bt361//+pceeeSRUS8eAOCd5/Bv375da9eu1SOPPKLs7Gz96U9/UkZGhqqqqoad//e//1033XST1q1bp6ysLP3iF7/Qo48+qpMnT4568QAA7zyF/9y5c2psbFRhYWHEeGFhoY4fPz7sMfn5+Tpz5ozq6+vlnNOXX36pPXv26IEHHrjk9/T396u3tzfiBQCIDU/h7+7u1sDAgNLS0iLG09LS1NnZOewx+fn5qq2tVVFRkZKTkzV16lRde+21evnlly/5PRUVFQoEAuFXRkaGl2UCAC4jqj/u+ny+iPfOuSFjg1paWrRu3To988wzamxs1MGDB3X69GmVlJRc8vPLysoUCoXCr/b29miWCQAYRpKXyVOmTFFiYuKQs/uurq4hvwUMqqio0MKFC/Xkk09Kku644w5NmjRJBQUFeu6555Senj7kGL/fL7/f72VpAIAR8nTGn5ycrJycHAWDwYjxYDCo/Pz8YY/59ttvlZAQ+TWJiYmSLv6mAAAYW54v9ZSWlurNN99UTU2NWltbtXHjRrW1tYUv3ZSVlam4uDg8f+nSpXrvvfdUVVWlU6dO6dixY1q3bp3mz5+vadOmxW4nAIAR8XSpR5KKiorU09Ojbdu2qaOjQ3PmzFF9fb0yMzMlSR0dHRH39K9evVp9fX165ZVX9Lvf/U7XXnut7r33Xj3//POx2wUAYMR8bhxcb+nt7VUgEFAoFFJqauqVXg4AjIl4tY9n9QCAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGBMVOGvrKxUVlaWUlJSlJOTo4aGhsvO7+/vV3l5uTIzM+X3+3XzzTerpqYmqgUDAEYnyesBdXV12rBhgyorK7Vw4UK99tprWrx4sVpaWnTjjTcOe8zy5cv15Zdfqrq6Wj/72c/U1dWl8+fPj3rxAADvfM455+WAvLw8zZs3T1VVVeGx7OxsLVu2TBUVFUPmHzx4UA8//LBOnTql6667LqpF9vb2KhAIKBQKKTU1NarPAIDxJl7t83Sp59y5c2psbFRhYWHEeGFhoY4fPz7sMQcOHFBubq5eeOEFTZ8+XbNnz9amTZv03XffXfJ7+vv71dvbG/ECAMSGp0s93d3dGhgYUFpaWsR4WlqaOjs7hz3m1KlTOnr0qFJSUrRv3z51d3frscce01dffXXJ6/wVFRXaunWrl6UBAEYoqj/u+ny+iPfOuSFjgy5cuCCfz6fa2lrNnz9fS5Ys0fbt27Vr165LnvWXlZUpFAqFX+3t7dEsEwAwDE9n/FOmTFFiYuKQs/uurq4hvwUMSk9P1/Tp0xUIBMJj2dnZcs7pzJkzmjVr1pBj/H6//H6/l6UBAEbI0xl/cnKycnJyFAwGI8aDwaDy8/OHPWbhwoX64osvdPbs2fDYxx9/rISEBM2YMSOKJQMARsPzpZ7S0lK9+eabqqmpUWtrqzZu3Ki2tjaVlJRIuniZpri4ODx/xYoVmjx5stasWaOWlhYdOXJETz75pH7zm99o4sSJsdsJAGBEPN/HX1RUpJ6eHm3btk0dHR2aM2eO6uvrlZmZKUnq6OhQW1tbeP5PfvITBYNB/fa3v1Vubq4mT56s5cuX67nnnovdLgAAI+b5Pv4rgfv4AVh0VdzHDwAY/wg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcCYqMJfWVmprKwspaSkKCcnRw0NDSM67tixY0pKStKdd94ZzdcCAGLAc/jr6uq0YcMGlZeXq6mpSQUFBVq8eLHa2toue1woFFJxcbF++ctfRr1YAMDo+ZxzzssBeXl5mjdvnqqqqsJj2dnZWrZsmSoqKi553MMPP6xZs2YpMTFR+/fvV3Nz84i/s7e3V4FAQKFQSKmpqV6WCwDjVrza5+mM/9y5c2psbFRhYWHEeGFhoY4fP37J49566y19+umn2rJly4i+p7+/X729vREvAEBseAp/d3e3BgYGlJaWFjGelpamzs7OYY/55JNPtHnzZtXW1iopKWlE31NRUaFAIBB+ZWRkeFkmAOAyovrjrs/ni3jvnBsyJkkDAwNasWKFtm7dqtmzZ4/488vKyhQKhcKv9vb2aJYJABjGyE7B/9eUKVOUmJg45Oy+q6tryG8BktTX16eTJ0+qqalJTzzxhCTpwoULcs4pKSlJhw4d0r333jvkOL/fL7/f72VpAIAR8nTGn5ycrJycHAWDwYjxYDCo/Pz8IfNTU1P14Ycfqrm5OfwqKSnRLbfcoubmZuXl5Y1u9QAAzzyd8UtSaWmpVq5cqdzcXC1YsECvv/662traVFJSIuniZZrPP/9cb7/9thISEjRnzpyI42+44QalpKQMGQcAjA3P4S8qKlJPT4+2bdumjo4OzZkzR/X19crMzJQkdXR0/Og9/QCAK8fzffxXAvfxA7DoqriPHwAw/hF+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxUYW/srJSWVlZSklJUU5OjhoaGi4597333tN9992n66+/XqmpqVqwYIHef//9qBcMABgdz+Gvq6vThg0bVF5erqamJhUUFGjx4sVqa2sbdv6RI0d03333qb6+Xo2Njbrnnnu0dOlSNTU1jXrxAADvfM455+WAvLw8zZs3T1VVVeGx7OxsLVu2TBUVFSP6jNtvv11FRUV65plnRjS/t7dXgUBAoVBIqampXpYLAONWvNrn6Yz/3LlzamxsVGFhYcR4YWGhjh8/PqLPuHDhgvr6+nTdddddck5/f796e3sjXgCA2PAU/u7ubg0MDCgtLS1iPC0tTZ2dnSP6jBdffFHffPONli9ffsk5FRUVCgQC4VdGRoaXZQIALiOqP+76fL6I9865IWPDeffdd/Xss8+qrq5ON9xwwyXnlZWVKRQKhV/t7e3RLBMAMIwkL5OnTJmixMTEIWf3XV1dQ34L+KG6ujqtXbtWu3fv1qJFiy471+/3y+/3e1kaAGCEPJ3xJycnKycnR8FgMGI8GAwqPz//kse9++67Wr16td555x098MAD0a0UABATns74Jam0tFQrV65Ubm6uFixYoNdff11tbW0qKSmRdPEyzeeff663335b0sXoFxcX66WXXtJdd90V/m1h4sSJCgQCMdwKAGAkPIe/qKhIPT092rZtmzo6OjRnzhzV19crMzNTktTR0RFxT/9rr72m8+fP6/HHH9fjjz8eHl+1apV27do1+h0AADzxfB//lcB9/AAsuiru4wcAjH+EHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgDOEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAGMIPAMYQfgAwhvADgDGEHwCMIfwAYAzhBwBjCD8AGEP4AcAYwg8AxhB+ADCG8AOAMYQfAIwh/ABgTFThr6ysVFZWllJSUpSTk6OGhobLzj98+LBycnKUkpKimTNn6tVXX41qsQCA0fMc/rq6Om3YsEHl5eVqampSQUGBFi9erLa2tmHnnz59WkuWLFFBQYGampr09NNPa926ddq7d++oFw8A8M7nnHNeDsjLy9O8efNUVVUVHsvOztayZctUUVExZP5TTz2lAwcOqLW1NTxWUlKiDz74QCdOnBjRd/b29ioQCCgUCik1NdXLcgFg3IpX+5K8TD537pwaGxu1efPmiPHCwkIdP3582GNOnDihwsLCiLH7779f1dXV+v777zVhwoQhx/T396u/vz/8PhQKSbr4XwIAWDHYPI/n5z/KU/i7u7s1MDCgtLS0iPG0tDR1dnYOe0xnZ+ew88+fP6/u7m6lp6cPOaaiokJbt24dMp6RkeFluQDw/0JPT48CgUDMPs9T+Af5fL6I9865IWM/Nn+48UFlZWUqLS0Nv//666+VmZmptra2mG5+vOjt7VVGRoba29vNXeqyvHeJ/VvffygU0o033qjrrrsupp/rKfxTpkxRYmLikLP7rq6uIWf1g6ZOnTrs/KSkJE2ePHnYY/x+v/x+/5DxQCBg8oc/KDU11ez+Le9dYv/W95+QENs77z19WnJysnJychQMBiPGg8Gg8vPzhz1mwYIFQ+YfOnRIubm5w17fBwDEl+d/jJSWlurNN99UTU2NWltbtXHjRrW1tamkpETSxcs0xcXF4fklJSX67LPPVFpaqtbWVtXU1Ki6ulqbNm2K3S4AACPm+Rp/UVGRenp6tG3bNnV0dGjOnDmqr69XZmamJKmjoyPinv6srCzV19dr48aN2rlzp6ZNm6YdO3booYceGvF3+v1+bdmyZdjLPxZY3r/lvUvsn/3HZ/+e7+MHAIxvPKsHAIwh/ABgDOEHAGMIPwAYc9WE3/Kjnr3s/b333tN9992n66+/XqmpqVqwYIHef//9MVxt7Hn92Q86duyYkpKSdOedd8Z3gXHmdf/9/f0qLy9XZmam/H6/br75ZtXU1IzRamPP6/5ra2s1d+5cXXPNNUpPT9eaNWvU09MzRquNnSNHjmjp0qWaNm2afD6f9u/f/6PHxKx77irw5z//2U2YMMG98cYbrqWlxa1fv95NmjTJffbZZ8POP3XqlLvmmmvc+vXrXUtLi3vjjTfchAkT3J49e8Z45aPnde/r1693zz//vPvnP//pPv74Y1dWVuYmTJjg/v3vf4/xymPD6/4Hff31127mzJmusLDQzZ07d2wWGwfR7P/BBx90eXl5LhgMutOnT7t//OMf7tixY2O46tjxuv+GhgaXkJDgXnrpJXfq1CnX0NDgbr/9drds2bIxXvno1dfXu/Lycrd3714nye3bt++y82PZvasi/PPnz3clJSURY7feeqvbvHnzsPN///vfu1tvvTVi7NFHH3V33XVX3NYYL173PpzbbrvNbd26NdZLGxPR7r+oqMj94Q9/cFu2bBnX4fe6/7/85S8uEAi4np6esVhe3Hnd/x//+Ec3c+bMiLEdO3a4GTNmxG2NY2Ek4Y9l9674pZ7BRz3/8NHN0Tzq+eTJk/r+++/jttZYi2bvP3ThwgX19fXF/CFOYyHa/b/11lv69NNPtWXLlngvMa6i2f+BAweUm5urF154QdOnT9fs2bO1adMmfffdd2Ox5JiKZv/5+fk6c+aM6uvr5ZzTl19+qT179uiBBx4YiyVfUbHsXlRP54ylsXrU89Uomr3/0IsvvqhvvvlGy5cvj8cS4yqa/X/yySfavHmzGhoalJR0xf/nOyrR7P/UqVM6evSoUlJStG/fPnV3d+uxxx7TV199Ne6u80ez//z8fNXW1qqoqEj//e9/df78eT344IN6+eWXx2LJV1Qsu3fFz/gHxftRz1czr3sf9O677+rZZ59VXV2dbrjhhngtL+5Guv+BgQGtWLFCW7du1ezZs8dqeXHn5ed/4cIF+Xw+1dbWav78+VqyZIm2b9+uXbt2jcuzfsnb/ltaWrRu3To988wzamxs1MGDB3X69Onws8L+v4tV9674KdNYPer5ahTN3gfV1dVp7dq12r17txYtWhTPZcaN1/339fXp5MmTampq0hNPPCHpYgidc0pKStKhQ4d07733jsnaYyGan396erqmT58e8e+lyM7OlnNOZ86c0axZs+K65liKZv8VFRVauHChnnzySUnSHXfcoUmTJqmgoEDPPffcuPltPxqx7N4VP+O3/KjnaPYuXTzTX716td55551xfW3T6/5TU1P14Ycfqrm5OfwqKSnRLbfcoubmZuXl5Y3V0mMimp//woUL9cUXX+js2bPhsY8//lgJCQmaMWNGXNcba9Hs/9tvvx3ybPrExERJsf/XE15tYto9z38OjoPBW7qqq6tdS0uL27Bhg5s0aZL7z3/+45xzbvPmzW7lypXh+YO3NW3cuNG1tLS46urqcX8750j3/s4777ikpCS3c+dO19HREX59/fXXV2oLo+J1/z803u/q8br/vr4+N2PGDPerX/3KffTRR+7w4cNu1qxZ7pFHHrlSWxgVr/t/6623XFJSkqusrHSffvqpO3r0qMvNzXXz58+/UluIWl9fn2tqanJNTU1Oktu+fbtramoK38oaz+5dFeF3zrmdO3e6zMxMl5yc7ObNm+cOHz4c/s9WrVrl7r777oj5f/vb39zPf/5zl5yc7G666SZXVVU1xiuOHS97v/vuu52kIa9Vq1aN/cJjxOvP/v8a7+F3zvv+W1tb3aJFi9zEiRPdjBkzXGlpqfv222/HeNWx43X/O3bscLfddpubOHGiS09Pd7/+9a/dmTNnxnjVo/fXv/71sv9fjmf3eCwzABhzxa/xAwDGFuEHAGMIPwAYQ/gBwBjCDwDGEH4AMIbwA4AxhB8AjCH8AGAM4QcAYwg/ABhD+AHAmP8BFwQZFvbUfmsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1400x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot histograms for each difference column\n",
    "def plot_histograms(df, columns):\n",
    "    plt.figure(figsize=(14, 10))\n",
    "    \n",
    "    for i, col in enumerate(columns):\n",
    "        plt.subplot(2, 3, i + 1)  # Create a subplot for each column\n",
    "        df[col].hist(bins=50, color='skyblue', edgecolor='black')\n",
    "        plt.title(f'Distribution of {col}')\n",
    "        plt.xlabel('Value')\n",
    "        plt.ylabel('Frequency')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "# Define the columns to plot\n",
    "columns_to_plot = ['diff_1', 'diff_2', 'diff_3', 'diff_4', 'diff_5']\n",
    "\n",
    "# Call the function to plot histograms\n",
    "plot_histograms(f_b_final_with_differences, columns_to_plot)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e7bc34d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "s_het_info = pd.read_excel('/Users/natasharecoder/Documents/NYU/Term3/Capstone/Data/s_het_info.xlsx', sheet_name = 'Supplementary Table 1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "3d62e2ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>hgnc</th>\n",
       "      <th>chrom</th>\n",
       "      <th>obs_lof</th>\n",
       "      <th>exp_lof</th>\n",
       "      <th>prior_mean</th>\n",
       "      <th>post_mean</th>\n",
       "      <th>post_lower_95</th>\n",
       "      <th>post_upper_95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198488</td>\n",
       "      <td>HGNC:24141</td>\n",
       "      <td>chr11</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.9777</td>\n",
       "      <td>0.001997</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000164363</td>\n",
       "      <td>HGNC:26441</td>\n",
       "      <td>chr5</td>\n",
       "      <td>31.0</td>\n",
       "      <td>28.5500</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000159337</td>\n",
       "      <td>HGNC:30038</td>\n",
       "      <td>chr15</td>\n",
       "      <td>28.0</td>\n",
       "      <td>41.8400</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000177558</td>\n",
       "      <td>HGNC:26366</td>\n",
       "      <td>chr19</td>\n",
       "      <td>12.0</td>\n",
       "      <td>9.9101</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000062524</td>\n",
       "      <td>HGNC:6721</td>\n",
       "      <td>chr15</td>\n",
       "      <td>40.0</td>\n",
       "      <td>42.6820</td>\n",
       "      <td>0.002000</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19066</th>\n",
       "      <td>ENSG00000108298</td>\n",
       "      <td>HGNC:10312</td>\n",
       "      <td>chr17</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.2410</td>\n",
       "      <td>0.349655</td>\n",
       "      <td>0.777394</td>\n",
       "      <td>0.188250</td>\n",
       "      <td>0.999947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19067</th>\n",
       "      <td>ENSG00000142937</td>\n",
       "      <td>HGNC:10441</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.8460</td>\n",
       "      <td>0.367146</td>\n",
       "      <td>0.781929</td>\n",
       "      <td>0.176357</td>\n",
       "      <td>0.999976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19068</th>\n",
       "      <td>ENSG00000231500</td>\n",
       "      <td>HGNC:10401</td>\n",
       "      <td>chr6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.4500</td>\n",
       "      <td>0.407376</td>\n",
       "      <td>0.823303</td>\n",
       "      <td>0.190498</td>\n",
       "      <td>0.999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19069</th>\n",
       "      <td>ENSG00000108055</td>\n",
       "      <td>HGNC:2468</td>\n",
       "      <td>chr10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>79.4900</td>\n",
       "      <td>0.376332</td>\n",
       "      <td>0.841343</td>\n",
       "      <td>0.460300</td>\n",
       "      <td>0.998543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19070</th>\n",
       "      <td>ENSG00000118058</td>\n",
       "      <td>HGNC:7132</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5.0</td>\n",
       "      <td>160.6800</td>\n",
       "      <td>0.458713</td>\n",
       "      <td>0.880224</td>\n",
       "      <td>0.574254</td>\n",
       "      <td>0.998856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19071 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ensg        hgnc  chrom  obs_lof   exp_lof  prior_mean  \\\n",
       "0      ENSG00000198488  HGNC:24141  chr11     12.0    8.9777    0.001997   \n",
       "1      ENSG00000164363  HGNC:26441   chr5     31.0   28.5500    0.000973   \n",
       "2      ENSG00000159337  HGNC:30038  chr15     28.0   41.8400    0.000856   \n",
       "3      ENSG00000177558  HGNC:26366  chr19     12.0    9.9101    0.001040   \n",
       "4      ENSG00000062524   HGNC:6721  chr15     40.0   42.6820    0.002000   \n",
       "...                ...         ...    ...      ...       ...         ...   \n",
       "19066  ENSG00000108298  HGNC:10312  chr17      0.0   12.2410    0.349655   \n",
       "19067  ENSG00000142937  HGNC:10441   chr1      0.0   13.8460    0.367146   \n",
       "19068  ENSG00000231500  HGNC:10401   chr6      0.0   11.4500    0.407376   \n",
       "19069  ENSG00000108055   HGNC:2468  chr10      0.0   79.4900    0.376332   \n",
       "19070  ENSG00000118058   HGNC:7132  chr11      5.0  160.6800    0.458713   \n",
       "\n",
       "       post_mean  post_lower_95  post_upper_95  \n",
       "0       0.000065       0.000007       0.000178  \n",
       "1       0.000161       0.000026       0.000442  \n",
       "2       0.000170       0.000019       0.000533  \n",
       "3       0.000175       0.000025       0.000487  \n",
       "4       0.000180       0.000012       0.000589  \n",
       "...          ...            ...            ...  \n",
       "19066   0.777394       0.188250       0.999947  \n",
       "19067   0.781929       0.176357       0.999976  \n",
       "19068   0.823303       0.190498       0.999999  \n",
       "19069   0.841343       0.460300       0.998543  \n",
       "19070   0.880224       0.574254       0.998856  \n",
       "\n",
       "[19071 rows x 9 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_het_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "4f1fa143",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to perform the merge for all forward and backward GeneSymbols and rename the 'post_mean' columns\n",
    "def merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges):\n",
    "    # Rename 'ensg' column in s_het_info to match the GeneSymbol columns in f_b_final_with_differences\n",
    "    s_het_info = s_het_info.rename(columns={'ensg': 'GeneSymbol'})\n",
    "\n",
    "    # Iterate over both forward and backward gene symbols\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Forward merge for each GeneSymbol_f\n",
    "        f_b_final_with_differences = pd.merge(\n",
    "            f_b_final_with_differences, \n",
    "            s_het_info[['GeneSymbol', 'post_mean']], \n",
    "            left_on=f'GeneSymbol_{i}', \n",
    "            right_on='GeneSymbol', \n",
    "            how='left',\n",
    "            suffixes=('', f'_{i}')\n",
    "        )\n",
    "        \n",
    "        # Rename the post_mean column to indicate the forward direction\n",
    "        f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "            columns={'post_mean': f's_het_post_{i}'}\n",
    "        )\n",
    "        \n",
    "        # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "        f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "#         # Backward merge for each GeneSymbol_b\n",
    "#         f_b_final_with_differences = pd.merge(\n",
    "#             f_b_final_with_differences, \n",
    "#             s_het_info[['GeneSymbol', 'post_mean']], \n",
    "#             left_on=f'GeneSymbol_b{i}', \n",
    "#             right_on='GeneSymbol', \n",
    "#             how='left',\n",
    "#             suffixes=('', f'_b{i}')\n",
    "#         )\n",
    "        \n",
    "#         # Rename the post_mean column to indicate the backward direction\n",
    "#         f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "#             columns={'post_mean': f's_het_post_b{i}'}\n",
    "#         )\n",
    "        \n",
    "#         # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "#         f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "    return f_b_final_with_differences\n",
    "\n",
    "# Example usage:\n",
    "f_b_final_merged_s_het = merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a068a9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db752cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f7cc0427",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dont need anymore\n",
    "\n",
    "# # f_b_final_merged_s_het[['diff_b1']] = abs(f_b_final_merged_s_het[['diff_b1']])\n",
    "# # f_b_final_merged_s_het[['diff_b2']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# # f_b_final_merged_s_het[['diff_b3']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# # Loop to apply the absolute value for all diff_b columns\n",
    "# def apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3):\n",
    "#     for i in range(1, num_merges + 1):\n",
    "#         f_b_final_merged_s_het[f'diff_b{i}'] = f_b_final_merged_s_het[f'diff_b{i}'].abs()\n",
    "\n",
    "#     return f_b_final_merged_s_het\n",
    "\n",
    "# # Example usage:\n",
    "# f_b_final_merged_s_het = apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "3566073f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to keep only the distance and s_het measures from the dataframe\n",
    "def keep_dist_and_s_het_measures(df, num_merges):\n",
    "    # List to store column names for distance and s_het measures\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the diff and s_het_post column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'diff_{i}')\n",
    "#         columns_to_keep.append(f'diff_b{i}')\n",
    "        columns_to_keep.append(f's_het_post_{i}')\n",
    "#         columns_to_keep.append(f's_het_post_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "filtered_f_b_final = keep_dist_and_s_het_measures(f_b_final_merged_s_het, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9536b18f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>diff_f1</th>\n",
       "      <th>diff_b1</th>\n",
       "      <th>s_het_post_f1</th>\n",
       "      <th>s_het_post_b1</th>\n",
       "      <th>diff_f2</th>\n",
       "      <th>diff_b2</th>\n",
       "      <th>s_het_post_f2</th>\n",
       "      <th>s_het_post_b2</th>\n",
       "      <th>diff_f3</th>\n",
       "      <th>diff_b3</th>\n",
       "      <th>s_het_post_f3</th>\n",
       "      <th>s_het_post_b3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17:302:T:TA</td>\n",
       "      <td>17</td>\n",
       "      <td>302.0</td>\n",
       "      <td>5508.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.014912</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61878.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>259814.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001599</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16:105325:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>105325.0</td>\n",
       "      <td>2733.0</td>\n",
       "      <td>1497.0</td>\n",
       "      <td>0.015023</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>21681.0</td>\n",
       "      <td>8918.0</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>0.004248</td>\n",
       "      <td>28945.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16:105444:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>105444.0</td>\n",
       "      <td>2614.0</td>\n",
       "      <td>1616.0</td>\n",
       "      <td>0.015023</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>21562.0</td>\n",
       "      <td>9037.0</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>0.004248</td>\n",
       "      <td>28826.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16:105752:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>105752.0</td>\n",
       "      <td>2306.0</td>\n",
       "      <td>1924.0</td>\n",
       "      <td>0.015023</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>21254.0</td>\n",
       "      <td>9345.0</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>0.004248</td>\n",
       "      <td>28518.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16:107211:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>107211.0</td>\n",
       "      <td>847.0</td>\n",
       "      <td>3383.0</td>\n",
       "      <td>0.015023</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>19795.0</td>\n",
       "      <td>10804.0</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>0.004248</td>\n",
       "      <td>27059.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364298</th>\n",
       "      <td>1:249205823:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249205823.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5428.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61618.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73414.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364299</th>\n",
       "      <td>1:249205806:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5411.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61601.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73397.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364300</th>\n",
       "      <td>1:249205774:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205774.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61569.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73365.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364301</th>\n",
       "      <td>1:249206890:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249206890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6495.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62685.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74481.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30519.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86709.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98505.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant  chr          pos  diff_f1  diff_b1  s_het_post_f1  \\\n",
       "0             17:302:T:TA   17        302.0   5508.0      NaN       0.014912   \n",
       "1           16:105325:G:A   16     105325.0   2733.0   1497.0       0.015023   \n",
       "2           16:105444:C:T   16     105444.0   2614.0   1616.0       0.015023   \n",
       "3           16:105752:C:T   16     105752.0   2306.0   1924.0       0.015023   \n",
       "4           16:107211:G:A   16     107211.0    847.0   3383.0       0.015023   \n",
       "...                   ...  ...          ...      ...      ...            ...   \n",
       "13364298  1:249205823:G:A    1  249205823.0      NaN   5428.0            NaN   \n",
       "13364299  1:249205806:T:C    1  249205806.0      NaN   5411.0            NaN   \n",
       "13364300  1:249205774:T:C    1  249205774.0      NaN   5379.0            NaN   \n",
       "13364301  1:249206890:T:A    1  249206890.0      NaN   6495.0            NaN   \n",
       "13364302  1:249230914:C:T    1  249230914.0      NaN  30519.0            NaN   \n",
       "\n",
       "          s_het_post_b1  diff_f2  diff_b2  s_het_post_f2  s_het_post_b2  \\\n",
       "0                   NaN  61878.0      NaN       0.001230            NaN   \n",
       "1              0.005491  21681.0   8918.0       0.000742       0.004248   \n",
       "2              0.005491  21562.0   9037.0       0.000742       0.004248   \n",
       "3              0.005491  21254.0   9345.0       0.000742       0.004248   \n",
       "4              0.005491  19795.0  10804.0       0.000742       0.004248   \n",
       "...                 ...      ...      ...            ...            ...   \n",
       "13364298       0.001772      NaN  61618.0            NaN       0.006796   \n",
       "13364299       0.001772      NaN  61601.0            NaN       0.006796   \n",
       "13364300       0.001772      NaN  61569.0            NaN       0.006796   \n",
       "13364301       0.001772      NaN  62685.0            NaN       0.006796   \n",
       "13364302       0.001772      NaN  86709.0            NaN       0.006796   \n",
       "\n",
       "           diff_f3  diff_b3  s_het_post_f3  s_het_post_b3  \n",
       "0         259814.0      NaN       0.001599            NaN  \n",
       "1          28945.0      NaN       0.020996            NaN  \n",
       "2          28826.0      NaN       0.020996            NaN  \n",
       "3          28518.0      NaN       0.020996            NaN  \n",
       "4          27059.0      NaN       0.020996            NaN  \n",
       "...            ...      ...            ...            ...  \n",
       "13364298       NaN  73414.0            NaN       0.003478  \n",
       "13364299       NaN  73397.0            NaN       0.003478  \n",
       "13364300       NaN  73365.0            NaN       0.003478  \n",
       "13364301       NaN  74481.0            NaN       0.003478  \n",
       "13364302       NaN  98505.0            NaN       0.003478  \n",
       "\n",
       "[13364303 rows x 15 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "8efcecb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/natasharecoder/anaconda3/envs/sherlock2/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/natasharecoder/anaconda3/envs/sherlock2/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n"
     ]
    }
   ],
   "source": [
    "# Function to compute s_het weighted by 1/distance per gene-SNP pair\n",
    "def compute_weighted_s_het(df, num_merges):\n",
    "    # Loop to compute the weighted s_het for both forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Calculate inverse distance for forward and backward directions\n",
    "        df[f'inv_dist_{i}'] = 1 / df[f'diff_{i}']\n",
    "#         df[f'inv_dist_b{i}'] = 1 / df[f'diff_b{i}']\n",
    "        \n",
    "        # Compute weighted s_het by multiplying s_het by inverse distance\n",
    "        df[f'weighted_s_het_{i}'] = df[f'inv_dist_{i}'] * df[f's_het_post_{i}']\n",
    "#         df[f'weighted_s_het_b{i}'] = df[f'inv_dist_b{i}'] * df[f's_het_post_b{i}']\n",
    "\n",
    "    return df\n",
    "\n",
    "# Example usage:\n",
    "weighted_f_b_final = compute_weighted_s_het(filtered_f_b_final, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "8b3df79a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>diff_1</th>\n",
       "      <th>s_het_post_1</th>\n",
       "      <th>diff_2</th>\n",
       "      <th>s_het_post_2</th>\n",
       "      <th>diff_3</th>\n",
       "      <th>s_het_post_3</th>\n",
       "      <th>diff_4</th>\n",
       "      <th>...</th>\n",
       "      <th>inv_dist_1</th>\n",
       "      <th>weighted_s_het_1</th>\n",
       "      <th>inv_dist_2</th>\n",
       "      <th>weighted_s_het_2</th>\n",
       "      <th>inv_dist_3</th>\n",
       "      <th>weighted_s_het_3</th>\n",
       "      <th>inv_dist_4</th>\n",
       "      <th>weighted_s_het_4</th>\n",
       "      <th>inv_dist_5</th>\n",
       "      <th>weighted_s_het_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17:302:T:TA</td>\n",
       "      <td>17</td>\n",
       "      <td>302.0</td>\n",
       "      <td>5508.0</td>\n",
       "      <td>0.014912</td>\n",
       "      <td>61878.0</td>\n",
       "      <td>0.001230</td>\n",
       "      <td>259814.0</td>\n",
       "      <td>0.001599</td>\n",
       "      <td>289467.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>2.707407e-06</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>1.987007e-08</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>6.153787e-09</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>7.181060e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>5.191534e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21:15357076:C:A</td>\n",
       "      <td>21</td>\n",
       "      <td>15357076.0</td>\n",
       "      <td>124058.0</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>231375.0</td>\n",
       "      <td>0.002477</td>\n",
       "      <td>374978.0</td>\n",
       "      <td>0.011208</td>\n",
       "      <td>386360.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>7.873092e-09</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1.070699e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.988869e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>5.548608e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.307121e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21:15357104:C:G</td>\n",
       "      <td>21</td>\n",
       "      <td>15357104.0</td>\n",
       "      <td>124030.0</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>231347.0</td>\n",
       "      <td>0.002477</td>\n",
       "      <td>375006.0</td>\n",
       "      <td>0.011208</td>\n",
       "      <td>386332.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>7.874869e-09</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1.070829e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.988646e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>5.549010e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.307195e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21:15366358:A:G</td>\n",
       "      <td>21</td>\n",
       "      <td>15366358.0</td>\n",
       "      <td>114776.0</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>222093.0</td>\n",
       "      <td>0.002477</td>\n",
       "      <td>377078.0</td>\n",
       "      <td>0.021438</td>\n",
       "      <td>384260.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>8.509793e-09</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>1.115447e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>5.685190e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.916671e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.331822e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21:15212919:G:C</td>\n",
       "      <td>21</td>\n",
       "      <td>15212919.0</td>\n",
       "      <td>230821.0</td>\n",
       "      <td>0.011208</td>\n",
       "      <td>268215.0</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>375532.0</td>\n",
       "      <td>0.002477</td>\n",
       "      <td>530517.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>4.855537e-08</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>3.641556e-09</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>6.596855e-09</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>4.040888e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.014813e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364298</th>\n",
       "      <td>1:249205823:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249205823.0</td>\n",
       "      <td>5428.0</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>61618.0</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>73414.0</td>\n",
       "      <td>0.003478</td>\n",
       "      <td>101178.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>3.264536e-07</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>1.102965e-07</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>4.737216e-08</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>4.227965e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>4.262569e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364299</th>\n",
       "      <td>1:249205806:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205806.0</td>\n",
       "      <td>5411.0</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>61601.0</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>73397.0</td>\n",
       "      <td>0.003478</td>\n",
       "      <td>101161.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>3.274792e-07</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>1.103269e-07</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>4.738314e-08</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>4.228675e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>4.262770e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364300</th>\n",
       "      <td>1:249205774:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205774.0</td>\n",
       "      <td>5379.0</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>61569.0</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>73365.0</td>\n",
       "      <td>0.003478</td>\n",
       "      <td>101129.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>3.294274e-07</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>1.103843e-07</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>4.740380e-08</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>4.230013e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>4.263148e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364301</th>\n",
       "      <td>1:249205246:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249205246.0</td>\n",
       "      <td>4851.0</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>61041.0</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>72837.0</td>\n",
       "      <td>0.003478</td>\n",
       "      <td>100601.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000206</td>\n",
       "      <td>3.652834e-07</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>1.113391e-07</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>4.774744e-08</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>4.252214e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>4.269389e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "      <td>30519.0</td>\n",
       "      <td>0.001772</td>\n",
       "      <td>86709.0</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>98505.0</td>\n",
       "      <td>0.003478</td>\n",
       "      <td>126269.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>5.806186e-08</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>7.837998e-08</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>3.530562e-08</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>3.387823e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>3.985700e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant  chr          pos    diff_1  s_het_post_1    diff_2  \\\n",
       "0             17:302:T:TA   17        302.0    5508.0      0.014912   61878.0   \n",
       "1         21:15357076:C:A   21   15357076.0  124058.0      0.000977  231375.0   \n",
       "2         21:15357104:C:G   21   15357104.0  124030.0      0.000977  231347.0   \n",
       "3         21:15366358:A:G   21   15366358.0  114776.0      0.000977  222093.0   \n",
       "4         21:15212919:G:C   21   15212919.0  230821.0      0.011208  268215.0   \n",
       "...                   ...  ...          ...       ...           ...       ...   \n",
       "13364298  1:249205823:G:A    1  249205823.0    5428.0      0.001772   61618.0   \n",
       "13364299  1:249205806:T:C    1  249205806.0    5411.0      0.001772   61601.0   \n",
       "13364300  1:249205774:T:C    1  249205774.0    5379.0      0.001772   61569.0   \n",
       "13364301  1:249205246:T:A    1  249205246.0    4851.0      0.001772   61041.0   \n",
       "13364302  1:249230914:C:T    1  249230914.0   30519.0      0.001772   86709.0   \n",
       "\n",
       "          s_het_post_2    diff_3  s_het_post_3    diff_4  ...  inv_dist_1  \\\n",
       "0             0.001230  259814.0      0.001599  289467.0  ...    0.000182   \n",
       "1             0.002477  374978.0      0.011208  386360.0  ...    0.000008   \n",
       "2             0.002477  375006.0      0.011208  386332.0  ...    0.000008   \n",
       "3             0.002477  377078.0      0.021438  384260.0  ...    0.000009   \n",
       "4             0.000977  375532.0      0.002477  530517.0  ...    0.000004   \n",
       "...                ...       ...           ...       ...  ...         ...   \n",
       "13364298      0.006796   73414.0      0.003478  101178.0  ...    0.000184   \n",
       "13364299      0.006796   73397.0      0.003478  101161.0  ...    0.000185   \n",
       "13364300      0.006796   73365.0      0.003478  101129.0  ...    0.000186   \n",
       "13364301      0.006796   72837.0      0.003478  100601.0  ...    0.000206   \n",
       "13364302      0.006796   98505.0      0.003478  126269.0  ...    0.000033   \n",
       "\n",
       "          weighted_s_het_1  inv_dist_2  weighted_s_het_2  inv_dist_3  \\\n",
       "0             2.707407e-06    0.000016      1.987007e-08    0.000004   \n",
       "1             7.873092e-09    0.000004      1.070699e-08    0.000003   \n",
       "2             7.874869e-09    0.000004      1.070829e-08    0.000003   \n",
       "3             8.509793e-09    0.000005      1.115447e-08    0.000003   \n",
       "4             4.855537e-08    0.000004      3.641556e-09    0.000003   \n",
       "...                    ...         ...               ...         ...   \n",
       "13364298      3.264536e-07    0.000016      1.102965e-07    0.000014   \n",
       "13364299      3.274792e-07    0.000016      1.103269e-07    0.000014   \n",
       "13364300      3.294274e-07    0.000016      1.103843e-07    0.000014   \n",
       "13364301      3.652834e-07    0.000016      1.113391e-07    0.000014   \n",
       "13364302      5.806186e-08    0.000012      7.837998e-08    0.000010   \n",
       "\n",
       "          weighted_s_het_3  inv_dist_4  weighted_s_het_4  inv_dist_5  \\\n",
       "0             6.153787e-09    0.000003      7.181060e-08    0.000002   \n",
       "1             2.988869e-08    0.000003      5.548608e-08    0.000002   \n",
       "2             2.988646e-08    0.000003      5.549010e-08    0.000002   \n",
       "3             5.685190e-08    0.000003      2.916671e-08    0.000002   \n",
       "4             6.596855e-09    0.000002      4.040888e-08    0.000002   \n",
       "...                    ...         ...               ...         ...   \n",
       "13364298      4.737216e-08    0.000010      4.227965e-07    0.000003   \n",
       "13364299      4.738314e-08    0.000010      4.228675e-07    0.000003   \n",
       "13364300      4.740380e-08    0.000010      4.230013e-07    0.000003   \n",
       "13364301      4.774744e-08    0.000010      4.252214e-07    0.000003   \n",
       "13364302      3.530562e-08    0.000008      3.387823e-07    0.000003   \n",
       "\n",
       "          weighted_s_het_5  \n",
       "0             5.191534e-08  \n",
       "1             1.307121e-08  \n",
       "2             1.307195e-08  \n",
       "3             1.331822e-08  \n",
       "4             1.014813e-08  \n",
       "...                    ...  \n",
       "13364298      4.262569e-09  \n",
       "13364299      4.262770e-09  \n",
       "13364300      4.263148e-09  \n",
       "13364301      4.269389e-09  \n",
       "13364302      3.985700e-09  \n",
       "\n",
       "[13364303 rows x 23 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weighted_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "0b3796b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to keep only the weighted s_het values from the dataframe\n",
    "def keep_weighted_s_het(df, num_merges):\n",
    "    # List to store column names for weighted s_het values\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the weighted_s_het column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'weighted_s_het_{i}')\n",
    "#         columns_to_keep.append(f'weighted_s_het_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "weighted_s_het_only = keep_weighted_s_het(weighted_f_b_final, num_merges=n_closest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a8f06a41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>weighted_s_het_1</th>\n",
       "      <th>weighted_s_het_2</th>\n",
       "      <th>weighted_s_het_3</th>\n",
       "      <th>weighted_s_het_4</th>\n",
       "      <th>weighted_s_het_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17:302:T:TA</td>\n",
       "      <td>17</td>\n",
       "      <td>302.0</td>\n",
       "      <td>2.707407e-06</td>\n",
       "      <td>1.987007e-08</td>\n",
       "      <td>6.153787e-09</td>\n",
       "      <td>7.181060e-08</td>\n",
       "      <td>5.191534e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21:15357076:C:A</td>\n",
       "      <td>21</td>\n",
       "      <td>15357076.0</td>\n",
       "      <td>7.873092e-09</td>\n",
       "      <td>1.070699e-08</td>\n",
       "      <td>2.988869e-08</td>\n",
       "      <td>5.548608e-08</td>\n",
       "      <td>1.307121e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21:15357104:C:G</td>\n",
       "      <td>21</td>\n",
       "      <td>15357104.0</td>\n",
       "      <td>7.874869e-09</td>\n",
       "      <td>1.070829e-08</td>\n",
       "      <td>2.988646e-08</td>\n",
       "      <td>5.549010e-08</td>\n",
       "      <td>1.307195e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21:15366358:A:G</td>\n",
       "      <td>21</td>\n",
       "      <td>15366358.0</td>\n",
       "      <td>8.509793e-09</td>\n",
       "      <td>1.115447e-08</td>\n",
       "      <td>5.685190e-08</td>\n",
       "      <td>2.916671e-08</td>\n",
       "      <td>1.331822e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21:15212919:G:C</td>\n",
       "      <td>21</td>\n",
       "      <td>15212919.0</td>\n",
       "      <td>4.855537e-08</td>\n",
       "      <td>3.641556e-09</td>\n",
       "      <td>6.596855e-09</td>\n",
       "      <td>4.040888e-08</td>\n",
       "      <td>1.014813e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364298</th>\n",
       "      <td>1:249205823:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249205823.0</td>\n",
       "      <td>3.264536e-07</td>\n",
       "      <td>1.102965e-07</td>\n",
       "      <td>4.737216e-08</td>\n",
       "      <td>4.227965e-07</td>\n",
       "      <td>4.262569e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364299</th>\n",
       "      <td>1:249205806:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205806.0</td>\n",
       "      <td>3.274792e-07</td>\n",
       "      <td>1.103269e-07</td>\n",
       "      <td>4.738314e-08</td>\n",
       "      <td>4.228675e-07</td>\n",
       "      <td>4.262770e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364300</th>\n",
       "      <td>1:249205774:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>249205774.0</td>\n",
       "      <td>3.294274e-07</td>\n",
       "      <td>1.103843e-07</td>\n",
       "      <td>4.740380e-08</td>\n",
       "      <td>4.230013e-07</td>\n",
       "      <td>4.263148e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364301</th>\n",
       "      <td>1:249205246:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>249205246.0</td>\n",
       "      <td>3.652834e-07</td>\n",
       "      <td>1.113391e-07</td>\n",
       "      <td>4.774744e-08</td>\n",
       "      <td>4.252214e-07</td>\n",
       "      <td>4.269389e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364302</th>\n",
       "      <td>1:249230914:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>249230914.0</td>\n",
       "      <td>5.806186e-08</td>\n",
       "      <td>7.837998e-08</td>\n",
       "      <td>3.530562e-08</td>\n",
       "      <td>3.387823e-07</td>\n",
       "      <td>3.985700e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13364303 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant  chr          pos  weighted_s_het_1  \\\n",
       "0             17:302:T:TA   17        302.0      2.707407e-06   \n",
       "1         21:15357076:C:A   21   15357076.0      7.873092e-09   \n",
       "2         21:15357104:C:G   21   15357104.0      7.874869e-09   \n",
       "3         21:15366358:A:G   21   15366358.0      8.509793e-09   \n",
       "4         21:15212919:G:C   21   15212919.0      4.855537e-08   \n",
       "...                   ...  ...          ...               ...   \n",
       "13364298  1:249205823:G:A    1  249205823.0      3.264536e-07   \n",
       "13364299  1:249205806:T:C    1  249205806.0      3.274792e-07   \n",
       "13364300  1:249205774:T:C    1  249205774.0      3.294274e-07   \n",
       "13364301  1:249205246:T:A    1  249205246.0      3.652834e-07   \n",
       "13364302  1:249230914:C:T    1  249230914.0      5.806186e-08   \n",
       "\n",
       "          weighted_s_het_2  weighted_s_het_3  weighted_s_het_4  \\\n",
       "0             1.987007e-08      6.153787e-09      7.181060e-08   \n",
       "1             1.070699e-08      2.988869e-08      5.548608e-08   \n",
       "2             1.070829e-08      2.988646e-08      5.549010e-08   \n",
       "3             1.115447e-08      5.685190e-08      2.916671e-08   \n",
       "4             3.641556e-09      6.596855e-09      4.040888e-08   \n",
       "...                    ...               ...               ...   \n",
       "13364298      1.102965e-07      4.737216e-08      4.227965e-07   \n",
       "13364299      1.103269e-07      4.738314e-08      4.228675e-07   \n",
       "13364300      1.103843e-07      4.740380e-08      4.230013e-07   \n",
       "13364301      1.113391e-07      4.774744e-08      4.252214e-07   \n",
       "13364302      7.837998e-08      3.530562e-08      3.387823e-07   \n",
       "\n",
       "          weighted_s_het_5  \n",
       "0             5.191534e-08  \n",
       "1             1.307121e-08  \n",
       "2             1.307195e-08  \n",
       "3             1.331822e-08  \n",
       "4             1.014813e-08  \n",
       "...                    ...  \n",
       "13364298      4.262569e-09  \n",
       "13364299      4.262770e-09  \n",
       "13364300      4.263148e-09  \n",
       "13364301      4.269389e-09  \n",
       "13364302      3.985700e-09  \n",
       "\n",
       "[13364303 rows x 8 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weighted_s_het_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b6c662",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
