{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e5608d8",
   "metadata": {},
   "source": [
    "# Import Gene Annotation File"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fb9c920",
   "metadata": {},
   "source": [
    "https://github.com/hakha-most/gwas_eqtl/blob/master/gene_annotations/genes.protein_coding.v39.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b64193a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>cons</th>\n",
       "      <th>gene</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621096</td>\n",
       "      <td>622034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>622034</td>\n",
       "      <td>621096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859303</td>\n",
       "      <td>879955</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC:28706</td>\n",
       "      <td>859303</td>\n",
       "      <td>879955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879583</td>\n",
       "      <td>894689</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>HGNC:24517</td>\n",
       "      <td>894689</td>\n",
       "      <td>879583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>895964</td>\n",
       "      <td>901099</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>HGNC:24023</td>\n",
       "      <td>895964</td>\n",
       "      <td>901099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end strand  ...    gene     hgnc_id     tss     tes\n",
       "0  chr1   65419   71585      +  ...   OR4F5  HGNC:14825   65419   71585\n",
       "1  chr1  621096  622034      -  ...  OR4F16  HGNC:15079  622034  621096\n",
       "2  chr1  859303  879955      +  ...  SAMD11  HGNC:28706  859303  879955\n",
       "3  chr1  879583  894689      -  ...   NOC2L  HGNC:24517  894689  879583\n",
       "4  chr1  895964  901099      +  ...  KLHL17  HGNC:24023  895964  901099\n",
       "\n",
       "[5 rows x 10 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "import os \n",
    "os.chdir(\"/Users/tanvibansal/Documents/GitHub/Capstone\")\n",
    "# Local path to GTF file\n",
    "gtf_file_path = 'genes.protein_coding.v39.gtf'\n",
    "\n",
    "# Load the GTF file into a pandas DataFrame and use the first row as headers\n",
    "gtf_df = pd.read_csv(gtf_file_path, sep='\\t', header=0)\n",
    "\n",
    "# Display the first few rows of the dataframe\n",
    "gtf_df.sort_values('start')\n",
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68c645d",
   "metadata": {},
   "source": [
    "#### convert chr to int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7fd4032",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>max</th>\n",
       "      <th>min</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>249200395</td>\n",
       "      <td>65419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>242836139</td>\n",
       "      <td>38814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>197687071</td>\n",
       "      <td>238446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>190945522</td>\n",
       "      <td>53180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>180794269</td>\n",
       "      <td>92283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>170884383</td>\n",
       "      <td>291630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>158820866</td>\n",
       "      <td>330136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>146277851</td>\n",
       "      <td>116086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>140772234</td>\n",
       "      <td>14475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>135437399</td>\n",
       "      <td>92828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>134248398</td>\n",
       "      <td>167784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>133794898</td>\n",
       "      <td>175933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>115079957</td>\n",
       "      <td>19747912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>105992920</td>\n",
       "      <td>19377594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>102462245</td>\n",
       "      <td>20737094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>90122974</td>\n",
       "      <td>96407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>81037485</td>\n",
       "      <td>5810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>77915115</td>\n",
       "      <td>47221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>59073298</td>\n",
       "      <td>107104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>62887048</td>\n",
       "      <td>67891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>48055507</td>\n",
       "      <td>14982098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>51205929</td>\n",
       "      <td>17071641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        max       min\n",
       "0     1  249200395     65419\n",
       "1     2  242836139     38814\n",
       "2     3  197687071    238446\n",
       "3     4  190945522     53180\n",
       "4     5  180794269     92283\n",
       "5     6  170884383    291630\n",
       "6     7  158820866    330136\n",
       "7     8  146277851    116086\n",
       "8     9  140772234     14475\n",
       "9    10  135437399     92828\n",
       "10   11  134248398    167784\n",
       "11   12  133794898    175933\n",
       "12   13  115079957  19747912\n",
       "13   14  105992920  19377594\n",
       "14   15  102462245  20737094\n",
       "15   16   90122974     96407\n",
       "16   17   81037485      5810\n",
       "17   18   77915115     47221\n",
       "18   19   59073298    107104\n",
       "19   20   62887048     67891\n",
       "20   21   48055507  14982098\n",
       "21   22   51205929  17071641"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df['chr'] = gtf_df['chr'].str[3:].astype(int)\n",
    "gtf_df['start'] = gtf_df['start'].astype(int)\n",
    "gtf_df.groupby('chr')['start'].agg({'min', 'max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e19dcc",
   "metadata": {},
   "source": [
    "Positions reset for each chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77808ff",
   "metadata": {},
   "source": [
    "### File Description\n",
    "\n",
    "This file contains **gene annotations** for **protein-coding genes** and provides information about the **location of genes on the genome**. Below is a description of each column:\n",
    "\n",
    "1. **chr**: The chromosome where the gene is located (e.g., `chr1`, `chr2`, etc.).\n",
    "2. **start**: The start position of the gene on the chromosome (in base pairs).\n",
    "3. **end**: The end position of the gene on the chromosome (in base pairs).\n",
    "4. **strand**: Indicates the strand on which the gene is located (`+` for forward strand, `-` for reverse strand).\n",
    "5. **GeneSymbol**: The symbol or name of the gene (e.g., `OR4F5`, `SAMD11`), typically assigned by organizations like HGNC (HUGO Gene Nomenclature Committee).\n",
    "6. **cons**: This column indicates the type of gene. In this file, all genes are classified as **protein_coding**.\n",
    "7. **gene**: The Ensembl gene ID, a unique identifier for the gene (e.g., `ENSG00000186092`).\n",
    "8. **hgnc_id**: The unique identifier for the gene assigned by the **HGNC** (HUGO Gene Nomenclature Committee).\n",
    "9. **tss**: The **transcription start site**, the position where transcription of the gene starts on the chromosome.\n",
    "10. **tes**: The **transcription end site**, the position where transcription of the gene ends on the chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e70a8e",
   "metadata": {},
   "source": [
    "# Import GWAS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7500998",
   "metadata": {},
   "source": [
    " https://docs.google.com/spreadsheets/d/1kvPoupSzsSFBNSztMzl04xMoSC3Kcx3CrjVf4yBmESU/edit?gid=178908679#gid=178908679 (row 7217)\n",
    " \n",
    " 50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "072a09d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to GWAS file\n",
    "gwas_file_path = '50_irnt.gwas.imputed_v3.both_sexes.tsv.bgz'\n",
    "import gzip \n",
    "# Open and read the compressed .bgz file using gzip\n",
    "with gzip.open(gwas_file_path, 'rt') as f:\n",
    "    # Load the file into a pandas DataFrame\n",
    "    gwas_df = pd.read_csv(f, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ab8c875a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.440760e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.003474</td>\n",
       "      <td>18.049900</td>\n",
       "      <td>178.468000</td>\n",
       "      <td>0.101138</td>\n",
       "      <td>0.919441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.761760e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>-0.087536</td>\n",
       "      <td>-0.041345</td>\n",
       "      <td>0.349577</td>\n",
       "      <td>-0.118272</td>\n",
       "      <td>0.905852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.881580e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>135.620000</td>\n",
       "      <td>-2.079970</td>\n",
       "      <td>-0.046994</td>\n",
       "      <td>0.062695</td>\n",
       "      <td>-0.749571</td>\n",
       "      <td>0.453514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.674710e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>360388</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>-0.105659</td>\n",
       "      <td>-0.042056</td>\n",
       "      <td>0.349588</td>\n",
       "      <td>-0.120302</td>\n",
       "      <td>0.904244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.106010e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>79718.500000</td>\n",
       "      <td>101.503000</td>\n",
       "      <td>0.000797</td>\n",
       "      <td>0.002899</td>\n",
       "      <td>0.274788</td>\n",
       "      <td>0.783479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>2.454380e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>176906.000000</td>\n",
       "      <td>-787.373000</td>\n",
       "      <td>-0.001657</td>\n",
       "      <td>0.001588</td>\n",
       "      <td>-1.044020</td>\n",
       "      <td>0.296477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.297000e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>165562.000000</td>\n",
       "      <td>-734.113000</td>\n",
       "      <td>-0.001654</td>\n",
       "      <td>0.001657</td>\n",
       "      <td>-0.998454</td>\n",
       "      <td>0.318060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.393990e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>172553.000000</td>\n",
       "      <td>-480.000000</td>\n",
       "      <td>-0.001689</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>-1.011490</td>\n",
       "      <td>0.311783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>2.458510e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>177204.000000</td>\n",
       "      <td>-706.111000</td>\n",
       "      <td>-0.001520</td>\n",
       "      <td>0.001587</td>\n",
       "      <td>-0.957369</td>\n",
       "      <td>0.338381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.955520e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>14094.900000</td>\n",
       "      <td>-265.942000</td>\n",
       "      <td>-0.013297</td>\n",
       "      <td>0.005446</td>\n",
       "      <td>-2.441660</td>\n",
       "      <td>0.014620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    variant minor_allele  ...     tstat      pval\n",
       "0               1:15791:C:T            T  ...  0.101138  0.919441\n",
       "1               1:69487:G:A            A  ... -0.118272  0.905852\n",
       "2               1:69569:T:C            C  ... -0.749571  0.453514\n",
       "3              1:139853:C:T            T  ... -0.120302  0.904244\n",
       "4             1:692794:CA:C            C  ...  0.274788  0.783479\n",
       "...                     ...          ...  ...       ...       ...\n",
       "13791462    X:154929412:C:T            T  ... -1.044020  0.296477\n",
       "13791463   X:154929637:CT:C            C  ... -0.998454  0.318060\n",
       "13791464  X:154929952:CAA:C            C  ... -1.011490  0.311783\n",
       "13791465    X:154930230:A:G            G  ... -0.957369  0.338381\n",
       "13791466    X:154930487:T:A            A  ... -2.441660  0.014620\n",
       "\n",
       "[13791467 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first few rows to inspect the structure of the GWAS file\n",
    "gwas_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1418c7f4",
   "metadata": {},
   "source": [
    "### GWAS Summary Statistics Dataset Description\n",
    "\n",
    "1. **variant**: The unique identifier for each SNP (Single Nucleotide Polymorphism). This can include information like chromosome, position, reference allele, and alternative allele (e.g., `1:12345:A:G`).\n",
    "   - **Example**: `1:10583:T:G`\n",
    "\n",
    "2. **minor_allele**: The allele that is less frequent in the population (minor allele) for this particular SNP.\n",
    "   - **Example**: `G`\n",
    "\n",
    "3. **minor_AF**: The **minor allele frequency** (AF), which represents the frequency of the minor allele in the population. It ranges from 0 to 1.\n",
    "   - **Example**: `0.35` (35% of individuals carry the minor allele)\n",
    "\n",
    "4. **low_confidence_variant**: A flag indicating whether the variant has **low confidence** due to imputation quality or other uncertainties. Values may be `TRUE` or `FALSE`.\n",
    "   - **Example**: `FALSE`\n",
    "\n",
    "5. **n_complete_samples**: The number of samples for which complete genotype data is available for this variant.\n",
    "   - **Example**: `300,000`\n",
    "\n",
    "6. **AC**: The **allele count** of the minor allele, i.e., the number of times the minor allele appears in the study population (across all samples).\n",
    "   - **Example**: `50000`\n",
    "\n",
    "7. **ytx**: Likely a placeholder for a phenotype-related statistic; depending on the dataset, this could represent something like the trait mean or effect size (its exact meaning depends on the specific analysis).\n",
    "\n",
    "8. **beta**: The **effect size** of the SNP on the trait being studied (in this case, likely height). It represents the change in the trait per additional copy of the minor allele.\n",
    "   - **Example**: `0.05` (the trait increases by 0.05 units for each additional copy of the minor allele)\n",
    "\n",
    "9. **se**: The **standard error** of the effect size (beta), indicating the precision of the estimated effect.\n",
    "   - **Example**: `0.01`\n",
    "\n",
    "10. **tstat**: The **t-statistic** for the beta estimate, which is the ratio of the beta estimate to its standard error.\n",
    "    - **Example**: `5.0` (higher values indicate more significant associations)\n",
    "\n",
    "11. **pval**: The **p-value** of the association between the SNP and the trait. This indicates the significance of the result, with smaller p-values suggesting stronger evidence that the SNP is associated with the trait.\n",
    "    - **Example**: `1.2e-6` (a very small p-value, indicating strong evidence of association)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac3adf9",
   "metadata": {},
   "source": [
    "# Find Variant Chromosome and Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca4213d3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/z0/1hhlpnfn2mz3vk7t57zdcy0w0000gn/T/ipykernel_46541/2974836220.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_df['chr'] = gwas_df['chr'].astype(int)\n",
      "/var/folders/z0/1hhlpnfn2mz3vk7t57zdcy0w0000gn/T/ipykernel_46541/2974836220.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_df['pos'] = gwas_df['pos'].astype(int)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>max</th>\n",
       "      <th>min</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>249230914</td>\n",
       "      <td>15791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>243082521</td>\n",
       "      <td>10181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>197900375</td>\n",
       "      <td>60197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>190922257</td>\n",
       "      <td>13012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>180736061</td>\n",
       "      <td>11882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>170934414</td>\n",
       "      <td>202076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>159128550</td>\n",
       "      <td>27916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>146303560</td>\n",
       "      <td>34440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>141104957</td>\n",
       "      <td>40997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>135503523</td>\n",
       "      <td>61334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>134946451</td>\n",
       "      <td>131128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>133840247</td>\n",
       "      <td>163752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>115109852</td>\n",
       "      <td>19022393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>107289436</td>\n",
       "      <td>19185163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>102461014</td>\n",
       "      <td>20001226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>90238127</td>\n",
       "      <td>60455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>81109537</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>78017073</td>\n",
       "      <td>11603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>59117442</td>\n",
       "      <td>229783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>62963628</td>\n",
       "      <td>61098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>48119634</td>\n",
       "      <td>9574185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>16464274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        max       min\n",
       "0     1  249230914     15791\n",
       "1     2  243082521     10181\n",
       "2     3  197900375     60197\n",
       "3     4  190922257     13012\n",
       "4     5  180736061     11882\n",
       "5     6  170934414    202076\n",
       "6     7  159128550     27916\n",
       "7     8  146303560     34440\n",
       "8     9  141104957     40997\n",
       "9    10  135503523     61334\n",
       "10   11  134946451    131128\n",
       "11   12  133840247    163752\n",
       "12   13  115109852  19022393\n",
       "13   14  107289436  19185163\n",
       "14   15  102461014  20001226\n",
       "15   16   90238127     60455\n",
       "16   17   81109537       302\n",
       "17   18   78017073     11603\n",
       "18   19   59117442    229783\n",
       "19   20   62963628     61098\n",
       "20   21   48119634   9574185\n",
       "21   22   51237712  16464274"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df[['chr', 'pos', 'ref', 'alt']] = gwas_df['variant'].str.split(':', expand=True)\n",
    "gwas_df = gwas_df.loc[gwas_df['chr'] != 'X']\n",
    "gwas_df['chr'] = gwas_df['chr'].astype(int)\n",
    "gwas_df['pos'] = gwas_df['pos'].astype(int)\n",
    "\n",
    "gwas_df.groupby('chr')['pos'].agg({'min','max'}).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764da1f1",
   "metadata": {},
   "source": [
    "It appears that position resets on each chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae3e953",
   "metadata": {},
   "source": [
    "# Join to Find 5 Nearest Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87110b9f-045c-4dad-b627-9c95d1524505",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandarallel import pandarallel\n",
    "import os\n",
    "import gc\n",
    "os.chdir(\"/Users/tanvibansal/Documents/GitHub/Capstone\")\n",
    "#grab a small sample of the gwas dataframe containing mutation information \n",
    "gwas_df.head()\n",
    "gtf_df.head()\n",
    "gwas_sample = gwas_df.iloc[np.random.choice(range(len(gwas_df)),size=100000,replace=False)] \n",
    "\n",
    "#drop the genes that are not present in the s_het file as these may cause na's that will break the ldsc algorithm in the next step\n",
    "s_het_info = pd.read_excel('s_het_info.xlsx',sheet_name=1).rename(columns={\"ensg\":\"GeneSymbol\"}).set_index(\"GeneSymbol\")\n",
    "gtf_df_tidy = s_het_info.join(gtf_df.set_index(\"GeneSymbol\"),how=\"inner\")[[\"chr\",\"start\",\"end\",\"gene\"]]\n",
    "\n",
    "#write a function to find the 5 nearest genes by absolute difference in mutation position and gene start position\n",
    "def nearest_gene_finder(m):\n",
    "    #get the absolute l1 distance \n",
    "    dists = np.abs(gtf_df_tidy.loc[gtf_df_tidy.chr == m.chr,\"start\"] - m.pos)\n",
    "    top_5 = dists.sort_values()[0:5]\n",
    "    out = pd.DataFrame(m).T\n",
    "    out[[\"GeneSymbol.f1\",\"GeneSymbol.f2\",\"GeneSymbol.f3\",\"GeneSymbol.f4\",\"GeneSymbol.f5\"]] = top_5.index\n",
    "    out[[\"dist.f1\",\"dist.f2\",\"dist.f3\",\"dist.f4\",\"dist.f5\"]] = top_5.values/1000 #convert distance to kilobases by dividing through 1000\n",
    "    return(out)\n",
    "\n",
    "#create a sample of the gwas dataframe with only mutations on chr 22 for this stage\n",
    "gwas_sample_22 = gwas_df.loc[gwas_df.chr ==22]\n",
    "\n",
    "#apply the function over each row of the sample mutation df \n",
    "pandarallel.initialize(progress_bar=True)\n",
    "mutn_5_nearest_genes_sample = gwas_sample_22.parallel_apply(lambda m: nearest_gene_finder(m),axis=1)\n",
    "gc.collect()\n",
    "mutn_5_nearest_genes_sample_df.loc[mutn_5_nearest_genes_sample_df['dist.f1'] == 0,\"dist.f1\"] = 1/1000\n",
    "mutn_5_nearest_genes_sample_df = pd.concat(list(mutn_5_nearest_genes_sample))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fe557883-379c-4252-b4d2-e5d996259a5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol.f5</th>\n",
       "      <th>GeneSymbol.f4</th>\n",
       "      <th>GeneSymbol.f3</th>\n",
       "      <th>GeneSymbol.f2</th>\n",
       "      <th>GeneSymbol.f1</th>\n",
       "      <th>index</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>dist.f1</th>\n",
       "      <th>dist.f2</th>\n",
       "      <th>dist.f3</th>\n",
       "      <th>dist.f4</th>\n",
       "      <th>dist.f5</th>\n",
       "      <th>s_het.f1</th>\n",
       "      <th>s_het_w.f1</th>\n",
       "      <th>s_het.f2</th>\n",
       "      <th>s_het_w.f2</th>\n",
       "      <th>s_het.f3</th>\n",
       "      <th>s_het_w.f3</th>\n",
       "      <th>s_het.f4</th>\n",
       "      <th>s_het_w.f4</th>\n",
       "      <th>s_het.f5</th>\n",
       "      <th>s_het_w.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184537</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.084046</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>60578.3</td>\n",
       "      <td>-262.512</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.003183</td>\n",
       "      <td>1.07204</td>\n",
       "      <td>0.283702</td>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>607.367</td>\n",
       "      <td>800.028</td>\n",
       "      <td>978.552</td>\n",
       "      <td>1101.570</td>\n",
       "      <td>1132.913</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.546518e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184538</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.080987</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>58373.6</td>\n",
       "      <td>-220.884</td>\n",
       "      <td>0.003806</td>\n",
       "      <td>0.003111</td>\n",
       "      <td>1.22341</td>\n",
       "      <td>0.221175</td>\n",
       "      <td>22</td>\n",
       "      <td>16488635</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>583.006</td>\n",
       "      <td>775.667</td>\n",
       "      <td>954.191</td>\n",
       "      <td>1077.209</td>\n",
       "      <td>1108.552</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.611532e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184539</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.080078</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>57718.1</td>\n",
       "      <td>-250.356</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>0.003121</td>\n",
       "      <td>1.19791</td>\n",
       "      <td>0.230954</td>\n",
       "      <td>22</td>\n",
       "      <td>16488702</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>582.939</td>\n",
       "      <td>775.600</td>\n",
       "      <td>954.124</td>\n",
       "      <td>1077.142</td>\n",
       "      <td>1108.485</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.611715e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184540</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078102</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>56294.0</td>\n",
       "      <td>-249.417</td>\n",
       "      <td>0.004095</td>\n",
       "      <td>0.003069</td>\n",
       "      <td>1.33402</td>\n",
       "      <td>0.182199</td>\n",
       "      <td>22</td>\n",
       "      <td>16495833</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>575.808</td>\n",
       "      <td>768.469</td>\n",
       "      <td>946.993</td>\n",
       "      <td>1070.011</td>\n",
       "      <td>1101.354</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.631382e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000215568</td>\n",
       "      <td>ENSG00000172967</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "      <td>13184541</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.07444</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>53654.4</td>\n",
       "      <td>-203.467</td>\n",
       "      <td>0.004117</td>\n",
       "      <td>0.003224</td>\n",
       "      <td>1.27711</td>\n",
       "      <td>0.201566</td>\n",
       "      <td>22</td>\n",
       "      <td>16498458</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>573.183</td>\n",
       "      <td>765.844</td>\n",
       "      <td>944.368</td>\n",
       "      <td>1067.386</td>\n",
       "      <td>1098.729</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>2.638696e-07</td>\n",
       "      <td>0.008926</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.118445</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364298</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.29612</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>213436.0</td>\n",
       "      <td>-437.663</td>\n",
       "      <td>-0.002442</td>\n",
       "      <td>0.002004</td>\n",
       "      <td>-1.21861</td>\n",
       "      <td>0.222992</td>\n",
       "      <td>22</td>\n",
       "      <td>51229717</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>23.788</td>\n",
       "      <td>53.093</td>\n",
       "      <td>116.874</td>\n",
       "      <td>168.535</td>\n",
       "      <td>190.495</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.122508e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364299</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.072999</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>52615.7</td>\n",
       "      <td>-172.528</td>\n",
       "      <td>0.001794</td>\n",
       "      <td>0.003184</td>\n",
       "      <td>0.563533</td>\n",
       "      <td>0.573073</td>\n",
       "      <td>22</td>\n",
       "      <td>51229805</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>23.876</td>\n",
       "      <td>53.181</td>\n",
       "      <td>116.962</td>\n",
       "      <td>168.623</td>\n",
       "      <td>190.583</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.120911e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179763</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364300</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.053676</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>38688.2</td>\n",
       "      <td>-20.4974</td>\n",
       "      <td>0.002084</td>\n",
       "      <td>0.003916</td>\n",
       "      <td>0.532177</td>\n",
       "      <td>0.594604</td>\n",
       "      <td>22</td>\n",
       "      <td>51231220</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25.291</td>\n",
       "      <td>54.596</td>\n",
       "      <td>118.377</td>\n",
       "      <td>170.038</td>\n",
       "      <td>191.998</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>2.095559e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179764</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364301</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.299014</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>215522.0</td>\n",
       "      <td>-260.204</td>\n",
       "      <td>-0.001741</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>-0.895286</td>\n",
       "      <td>0.370635</td>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>31.134</td>\n",
       "      <td>60.439</td>\n",
       "      <td>124.220</td>\n",
       "      <td>175.881</td>\n",
       "      <td>197.841</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>1.996989e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179765</th>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>13364302</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.05515</td>\n",
       "      <td>False</td>\n",
       "      <td>360388</td>\n",
       "      <td>39750.5</td>\n",
       "      <td>-23.8471</td>\n",
       "      <td>0.001338</td>\n",
       "      <td>0.003885</td>\n",
       "      <td>0.344285</td>\n",
       "      <td>0.730632</td>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31.783</td>\n",
       "      <td>61.088</td>\n",
       "      <td>124.869</td>\n",
       "      <td>176.530</td>\n",
       "      <td>198.490</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.008019</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.248066</td>\n",
       "      <td>1.986610e-03</td>\n",
       "      <td>0.001543</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.706036</td>\n",
       "      <td>0.003557</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneSymbol.f5    GeneSymbol.f4  ...  s_het.f5 s_het_w.f5\n",
       "0       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000105\n",
       "1       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000107\n",
       "2       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000107\n",
       "3       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000108\n",
       "4       ENSG00000183307  ENSG00000177663  ...  0.118445   0.000108\n",
       "...                 ...              ...  ...       ...        ...\n",
       "179761  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003706\n",
       "179762  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003705\n",
       "179763  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003677\n",
       "179764  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003569\n",
       "179765  ENSG00000008735  ENSG00000100299  ...  0.706036   0.003557\n",
       "\n",
       "[179766 rows x 36 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now join the s_het values onto each of our 5 nearest genes and compute the distance weighted s_het \n",
    "s_het_w = mutn_5_nearest_genes_sample_df.copy(deep=True)\n",
    "for i in range(1,6):\n",
    "    s_het_temp = s_het_info[[\"post_mean\"]].reset_index().rename(columns = {\"GeneSymbol\":\"GeneSymbol.f%s\"%(i)}).set_index(\"GeneSymbol.f%s\"%(i))\n",
    "    s_het_w = s_het_w.reset_index().set_index(\"GeneSymbol.f%s\"%(i)).join(s_het_temp)\n",
    "    s_het_w[\"s_het.f%s\"%(i)] = s_het_w[\"post_mean\"]\n",
    "    s_het_w[\"s_het_w.f%s\"%(i)] = s_het_w[\"post_mean\"]/s_het_w[\"dist.f%s\"%(i)]\n",
    "    s_het_w = s_het_w.drop(columns=\"post_mean\")\n",
    "s_het_w = s_het_w.reset_index() \n",
    "s_het_w"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e68796f0-c8cf-49e3-8f39-e12e786c34c2",
   "metadata": {},
   "source": [
    "## Prepare the annotation file containing our features for the LDSC program "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6523dfd1-d531-4acd-8860-e609ec77fd36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>s_het_w.f1</th>\n",
       "      <th>s_het_w.f2</th>\n",
       "      <th>s_het_w.f3</th>\n",
       "      <th>s_het_w.f4</th>\n",
       "      <th>s_het_w.f5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16464274</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.546518e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16488635</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.611532e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16488702</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.611715e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16495833</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.631382e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16498458</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.638696e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179761</th>\n",
       "      <td>22</td>\n",
       "      <td>51229717</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>2.122508e-03</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.003706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179762</th>\n",
       "      <td>22</td>\n",
       "      <td>51229805</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>2.120911e-03</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.003705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179763</th>\n",
       "      <td>22</td>\n",
       "      <td>51231220</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>2.095559e-03</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.003677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179764</th>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>1.996989e-03</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.003569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179765</th>\n",
       "      <td>22</td>\n",
       "      <td>51237712</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>1.986610e-03</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.003557</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179766 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR        BP              SNP  ...    s_het_w.f3  s_het_w.f4  s_het_w.f5\n",
       "0       22  16464274  22:16464274:A:C  ...  2.546518e-07    0.000008    0.000105\n",
       "1       22  16488635  22:16488635:C:A  ...  2.611532e-07    0.000008    0.000107\n",
       "2       22  16488702  22:16488702:G:C  ...  2.611715e-07    0.000008    0.000107\n",
       "3       22  16495833  22:16495833:C:A  ...  2.631382e-07    0.000008    0.000108\n",
       "4       22  16498458  22:16498458:G:A  ...  2.638696e-07    0.000008    0.000108\n",
       "...     ..       ...              ...  ...           ...         ...         ...\n",
       "179761  22  51229717  22:51229717:A:T  ...  2.122508e-03    0.000009    0.003706\n",
       "179762  22  51229805  22:51229805:T:C  ...  2.120911e-03    0.000009    0.003705\n",
       "179763  22  51231220  22:51231220:A:G  ...  2.095559e-03    0.000009    0.003677\n",
       "179764  22  51237063  22:51237063:T:C  ...  1.996989e-03    0.000009    0.003569\n",
       "179765  22  51237712  22:51237712:G:A  ...  1.986610e-03    0.000009    0.003557\n",
       "\n",
       "[179766 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#format the dataframe's column names, columns to retain, and match the SNPs present in the bim file\n",
    "s_het_w = s_het_w.reset_index().rename(columns={'chr': 'CHR','pos': 'BP','variant': 'SNP'})\n",
    "annot_df = s_het_w[['CHR', 'BP', 'SNP', 's_het_w.f1', 's_het_w.f2', 's_het_w.f3', 's_het_w.f4', 's_het_w.f5']]\n",
    "#annot_df = s_het_w[['CHR', 'BP', 'SNP', 's_het.f1', 's_het.f2', 's_het.f3', 's_het.f4', 's_het.f5']]\n",
    "annot_df = annot_df.sort_values(['CHR', 'BP'])\n",
    "annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "67b4efa1-6c17-4ec6-9171-cb4d4340dbd4",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'annot_df' is not defined",
     "output_type": "error",
     "traceback": [
      "Traceback \u001b[0;36m(most recent call last)\u001b[0m:\n",
      "\u001b[0;36m  Cell \u001b[0;32mIn[238], line 8\u001b[0;36m\n\u001b[0;31m    annot_df_filt = bim.set_index([\"CHR\",\"BP\"]).join(annot_df.set_index([\"CHR\",\"BP\"]),how=\"left\",rsuffix=\".a\").drop_duplicates(subset=[\"SNP\"])\u001b[0;36m\n",
      "\u001b[0;31mNameError\u001b[0m\u001b[0;31m:\u001b[0m name 'annot_df' is not defined\n"
     ]
    }
   ],
   "source": [
    "#column names for the bim file according to the plink documentation site\n",
    "names = [\"CHR\",\"SNP\",\"POS\",\"BP\",\"A1\",\"A2\"]\n",
    "os.chdir(\"/Users/tanvibansal/Documents/GitHub/ldsc\")\n",
    "bim = pd.read_csv(\"22.bim\",sep=\"\\t\",header=None,names=names)\n",
    "\n",
    "#join the annotated df onto the bim file and sort in the order of the bim file exactly to prep for ldsc. we should have the same exact # of rows \n",
    "annot_df_filt = bim.set_index([\"CHR\",\"BP\"]).join(annot_df.set_index([\"CHR\",\"BP\"]),how=\"left\",rsuffix=\".a\").drop_duplicates(subset=[\"SNP\"])\n",
    "annot_df_sort = annot_df_filt.reset_index().set_index(\"SNP\").loc[bim.SNP.values]\n",
    "annot_df_sort[\"CM\"] = 0.0\n",
    "\n",
    "#select the cols of interest and write the file out as .annot to use for ldsc\n",
    "#annot_df_prep = annot_df_sort.reset_index()[[\"CHR\",\"BP\",\"SNP\", \"CM\", \"s_het_w.f1\", \"s_het_w.f2\", \"s_het_w.f3\", \"s_het_w.f4\", \"s_het_w.f5\"]]\n",
    "annot_df_prep = annot_df_sort.reset_index()[[\"CHR\",\"BP\",\"SNP\", \"CM\", \"s_het.f1\", \"s_het.f2\", \"s_het.f3\", \"s_het.f4\", \"s_het.f5\"]]\n",
    "annot_df_prep = annot_df_prep.dropna()\n",
    "annot_df_prep.to_csv(\"1kg_eur/s_het_unweighted_ch22.annot\",sep=\"\\t\",index=False)\n",
    "annot_df_prep\n",
    "\n",
    "#unique_snps = annot_df_prep['SNP'].unique()  # Get unique SNPs to avoid duplicates\n",
    "\n",
    "# Save the SNPs to a text file\n",
    "#with open('1kg_eur/s_het_22/filtered_snps.txt', 'w') as f:\n",
    " #   for snp in unique_snps:\n",
    "  #      f.write(f\"{snp}\\n\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0aa6a86-99cc-48cc-9b43-5771a5dc5b98",
   "metadata": {},
   "source": [
    "#### we need to run the following 2 commands to get the ldsc computation to run as intended \n",
    "\n",
    "1. Make the .bed/.bim files with our filtered SNPs\n",
    "\n",
    "\n",
    "plink --bfile /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/22 --extract /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22/filtered_snps.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22\n",
    "\n",
    "2. Compute the ld scores with our inputs specified\n",
    "\n",
    "python ldsc.py --l2 --bfile /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22_uw/s_het_22_uw --ld-wind-cm 1 --annot /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22_uw/s_het_uw_ch22.annot --out 1kg_eur/s_het_22_uw/s_het_22_uw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "40e26bac-c33f-4d5f-afb9-db9b48b13578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip \n",
    "with gzip.open('1kg_eur/s_het_22.l2.ldscore.gz', 'rb') as f: \n",
    "    file_content = pd.read_csv(f,sep=\"\\t\") \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232b6ced-8287-482f-9f47-85d089ed9932",
   "metadata": {},
   "source": [
    "## Generate plink triplet files needed for LD score computation for all chromosomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f81daf-6644-45aa-8450-f86b9329510b",
   "metadata": {},
   "source": [
    "#### 1. Convert and filter 10000 genomes data in .vcf format to plink triplet files for europeans"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e81b6a1-a571-4369-9df6-d6da03702628",
   "metadata": {},
   "source": [
    "##### 1.1 Extract IDs for european individuals in the 1000 genomes data set "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "b3ab03ee-d64e-4fc5-a607-cdf183fe623c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|      | Family ID   | Individual ID   |\n",
      "|-----:|:------------|:----------------|\n",
      "|    0 | HG00096     | HG00096         |\n",
      "|    1 | HG00097     | HG00097         |\n",
      "|    3 | HG00099     | HG00099         |\n",
      "|    4 | HG00100     | HG00100         |\n",
      "|    5 | HG00101     | HG00101         |\n",
      "|    6 | HG00102     | HG00102         |\n",
      "|    7 | HG00103     | HG00103         |\n",
      "|    9 | HG00105     | HG00105         |\n",
      "|   10 | HG00106     | HG00106         |\n",
      "|   11 | HG00107     | HG00107         |\n",
      "|   12 | HG00108     | HG00108         |\n",
      "|   13 | HG00109     | HG00109         |\n",
      "|   14 | HG00110     | HG00110         |\n",
      "|   15 | HG00111     | HG00111         |\n",
      "|   16 | HG00112     | HG00112         |\n",
      "|   17 | HG00113     | HG00113         |\n",
      "|   18 | HG00114     | HG00114         |\n",
      "|   19 | HG00115     | HG00115         |\n",
      "|   20 | HG00116     | HG00116         |\n",
      "|   21 | HG00117     | HG00117         |\n",
      "|   22 | HG00118     | HG00118         |\n",
      "|   23 | HG00119     | HG00119         |\n",
      "|   24 | HG00120     | HG00120         |\n",
      "|   25 | HG00121     | HG00121         |\n",
      "|   26 | HG00122     | HG00122         |\n",
      "|   27 | HG00123     | HG00123         |\n",
      "|   29 | HG00125     | HG00125         |\n",
      "|   30 | HG00126     | HG00126         |\n",
      "|   31 | HG00127     | HG00127         |\n",
      "|   32 | HG00128     | HG00128         |\n",
      "|   33 | HG00129     | HG00129         |\n",
      "|   34 | HG00130     | HG00130         |\n",
      "|   35 | HG00131     | HG00131         |\n",
      "|   36 | HG00132     | HG00132         |\n",
      "|   37 | HG00133     | HG00133         |\n",
      "|   40 | HG00136     | HG00136         |\n",
      "|   41 | HG00137     | HG00137         |\n",
      "|   42 | HG00138     | HG00138         |\n",
      "|   43 | HG00139     | HG00139         |\n",
      "|   44 | HG00140     | HG00140         |\n",
      "|   45 | HG00141     | HG00141         |\n",
      "|   46 | HG00142     | HG00142         |\n",
      "|   47 | HG00143     | HG00143         |\n",
      "|   49 | HG00145     | HG00145         |\n",
      "|   50 | GBR002      | HG00146         |\n",
      "|   52 | HG00148     | HG00148         |\n",
      "|   53 | HG00149     | HG00149         |\n",
      "|   54 | HG00150     | HG00150         |\n",
      "|   55 | HG00151     | HG00151         |\n",
      "|   58 | HG00154     | HG00154         |\n",
      "|   59 | GBR001      | HG00155         |\n",
      "|   61 | HG00157     | HG00157         |\n",
      "|   62 | GBR003      | HG00158         |\n",
      "|   63 | HG00159     | HG00159         |\n",
      "|   64 | HG00160     | HG00160         |\n",
      "|   65 | HG00171     | HG00171         |\n",
      "|   66 | HG00173     | HG00173         |\n",
      "|   67 | HG00174     | HG00174         |\n",
      "|   68 | HG00176     | HG00176         |\n",
      "|   69 | HG00177     | HG00177         |\n",
      "|   70 | HG00178     | HG00178         |\n",
      "|   71 | HG00179     | HG00179         |\n",
      "|   72 | HG00180     | HG00180         |\n",
      "|   73 | HG00181     | HG00181         |\n",
      "|   74 | HG00182     | HG00182         |\n",
      "|   75 | HG00183     | HG00183         |\n",
      "|   76 | HG00185     | HG00185         |\n",
      "|   77 | HG00186     | HG00186         |\n",
      "|   78 | HG00187     | HG00187         |\n",
      "|   79 | HG00188     | HG00188         |\n",
      "|   80 | HG00189     | HG00189         |\n",
      "|   81 | HG00190     | HG00190         |\n",
      "|   82 | HG00231     | HG00231         |\n",
      "|   83 | HG00232     | HG00232         |\n",
      "|   84 | HG00233     | HG00233         |\n",
      "|   85 | HG00234     | HG00234         |\n",
      "|   86 | HG00235     | HG00235         |\n",
      "|   87 | HG00236     | HG00236         |\n",
      "|   88 | HG00237     | HG00237         |\n",
      "|   89 | HG00238     | HG00238         |\n",
      "|   90 | HG00239     | HG00239         |\n",
      "|   91 | HG00240     | HG00240         |\n",
      "|   92 | HG00242     | HG00242         |\n",
      "|   93 | HG00243     | HG00243         |\n",
      "|   94 | HG00244     | HG00244         |\n",
      "|   95 | HG00245     | HG00245         |\n",
      "|   96 | HG00246     | HG00246         |\n",
      "|  100 | HG00250     | HG00250         |\n",
      "|  101 | HG00251     | HG00251         |\n",
      "|  102 | HG00252     | HG00252         |\n",
      "|  103 | HG00253     | HG00253         |\n",
      "|  104 | HG00254     | HG00254         |\n",
      "|  105 | HG00255     | HG00255         |\n",
      "|  106 | HG00256     | HG00256         |\n",
      "|  107 | HG00257     | HG00257         |\n",
      "|  108 | HG00258     | HG00258         |\n",
      "|  109 | HG00259     | HG00259         |\n",
      "|  110 | HG00260     | HG00260         |\n",
      "|  111 | HG00261     | HG00261         |\n",
      "|  112 | HG00262     | HG00262         |\n",
      "|  113 | HG00263     | HG00263         |\n",
      "|  114 | HG00264     | HG00264         |\n",
      "|  115 | HG00265     | HG00265         |\n",
      "|  116 | HG00266     | HG00266         |\n",
      "|  117 | HG00267     | HG00267         |\n",
      "|  118 | HG00268     | HG00268         |\n",
      "|  119 | HG00269     | HG00269         |\n",
      "|  121 | HG00271     | HG00271         |\n",
      "|  122 | HG00272     | HG00272         |\n",
      "|  123 | HG00273     | HG00273         |\n",
      "|  124 | HG00274     | HG00274         |\n",
      "|  125 | HG00275     | HG00275         |\n",
      "|  126 | HG00276     | HG00276         |\n",
      "|  127 | HG00277     | HG00277         |\n",
      "|  128 | HG00278     | HG00278         |\n",
      "|  129 | HG00280     | HG00280         |\n",
      "|  130 | HG00281     | HG00281         |\n",
      "|  131 | HG00282     | HG00282         |\n",
      "|  132 | HG00284     | HG00284         |\n",
      "|  133 | HG00285     | HG00285         |\n",
      "|  134 | HG00288     | HG00288         |\n",
      "|  135 | HG00290     | HG00290         |\n",
      "|  138 | HG00304     | HG00304         |\n",
      "|  139 | HG00306     | HG00306         |\n",
      "|  140 | HG00308     | HG00308         |\n",
      "|  141 | HG00309     | HG00309         |\n",
      "|  142 | HG00310     | HG00310         |\n",
      "|  143 | HG00311     | HG00311         |\n",
      "|  145 | HG00313     | HG00313         |\n",
      "|  146 | HG00315     | HG00315         |\n",
      "|  147 | HG00318     | HG00318         |\n",
      "|  148 | HG00319     | HG00319         |\n",
      "|  149 | HG00320     | HG00320         |\n",
      "|  150 | HG00321     | HG00321         |\n",
      "|  151 | HG00323     | HG00323         |\n",
      "|  152 | HG00324     | HG00324         |\n",
      "|  153 | HG00325     | HG00325         |\n",
      "|  154 | HG00326     | HG00326         |\n",
      "|  155 | HG00327     | HG00327         |\n",
      "|  156 | HG00328     | HG00328         |\n",
      "|  157 | HG00329     | HG00329         |\n",
      "|  158 | HG00330     | HG00330         |\n",
      "|  159 | HG00331     | HG00331         |\n",
      "|  160 | HG00332     | HG00332         |\n",
      "|  161 | HG00334     | HG00334         |\n",
      "|  162 | HG00335     | HG00335         |\n",
      "|  163 | HG00336     | HG00336         |\n",
      "|  164 | HG00337     | HG00337         |\n",
      "|  165 | HG00338     | HG00338         |\n",
      "|  166 | HG00339     | HG00339         |\n",
      "|  167 | HG00341     | HG00341         |\n",
      "|  168 | HG00342     | HG00342         |\n",
      "|  169 | HG00343     | HG00343         |\n",
      "|  170 | HG00344     | HG00344         |\n",
      "|  171 | HG00345     | HG00345         |\n",
      "|  172 | HG00346     | HG00346         |\n",
      "|  173 | HG00349     | HG00349         |\n",
      "|  174 | HG00350     | HG00350         |\n",
      "|  175 | HG00351     | HG00351         |\n",
      "|  176 | HG00353     | HG00353         |\n",
      "|  177 | HG00355     | HG00355         |\n",
      "|  178 | HG00356     | HG00356         |\n",
      "|  179 | HG00357     | HG00357         |\n",
      "|  180 | HG00358     | HG00358         |\n",
      "|  182 | HG00360     | HG00360         |\n",
      "|  183 | HG00361     | HG00361         |\n",
      "|  184 | HG00362     | HG00362         |\n",
      "|  185 | HG00364     | HG00364         |\n",
      "|  186 | HG00365     | HG00365         |\n",
      "|  187 | HG00366     | HG00366         |\n",
      "|  188 | HG00367     | HG00367         |\n",
      "|  189 | HG00368     | HG00368         |\n",
      "|  190 | HG00369     | HG00369         |\n",
      "|  191 | HG00371     | HG00371         |\n",
      "|  192 | HG00372     | HG00372         |\n",
      "|  193 | HG00373     | HG00373         |\n",
      "|  194 | HG00375     | HG00375         |\n",
      "|  195 | HG00376     | HG00376         |\n",
      "|  197 | HG00378     | HG00378         |\n",
      "|  198 | HG00379     | HG00379         |\n",
      "|  199 | HG00380     | HG00380         |\n",
      "|  200 | HG00381     | HG00381         |\n",
      "|  201 | HG00382     | HG00382         |\n",
      "|  202 | HG00383     | HG00383         |\n",
      "|  203 | HG00384     | HG00384         |\n",
      "|  576 | HG01334     | HG01334         |\n",
      "|  687 | IBS001      | HG01500         |\n",
      "|  688 | IBS001      | HG01501         |\n",
      "|  690 | IBS002      | HG01503         |\n",
      "|  691 | IBS002      | HG01504         |\n",
      "|  693 | IBS003      | HG01506         |\n",
      "|  694 | IBS003      | HG01507         |\n",
      "|  696 | IBS004      | HG01509         |\n",
      "|  697 | IBS004      | HG01510         |\n",
      "|  699 | IBS005      | HG01512         |\n",
      "|  700 | IBS005      | HG01513         |\n",
      "|  702 | IBS006      | HG01515         |\n",
      "|  703 | IBS006      | HG01516         |\n",
      "|  705 | IBS007      | HG01518         |\n",
      "|  706 | IBS007      | HG01519         |\n",
      "|  708 | IBS008      | HG01521         |\n",
      "|  709 | IBS008      | HG01522         |\n",
      "|  711 | IBS009      | HG01524         |\n",
      "|  712 | IBS009      | HG01525         |\n",
      "|  714 | IBS010      | HG01527         |\n",
      "|  715 | IBS010      | HG01528         |\n",
      "|  717 | IBS011      | HG01530         |\n",
      "|  718 | IBS011      | HG01531         |\n",
      "|  720 | IBS013      | HG01536         |\n",
      "|  721 | IBS013      | HG01537         |\n",
      "|  749 | IBS024      | HG01602         |\n",
      "|  750 | IBS024      | HG01603         |\n",
      "|  752 | IBS023      | HG01605         |\n",
      "|  753 | IBS023      | HG01606         |\n",
      "|  754 | IBS022      | HG01607         |\n",
      "|  755 | IBS022      | HG01608         |\n",
      "|  757 | IBS021      | HG01610         |\n",
      "|  759 | IBS021      | HG01612         |\n",
      "|  760 | IBS020      | HG01613         |\n",
      "|  762 | IBS020      | HG01615         |\n",
      "|  764 | IBS019      | HG01617         |\n",
      "|  765 | IBS019      | HG01618         |\n",
      "|  766 | IBS018      | HG01619         |\n",
      "|  767 | IBS018      | HG01620         |\n",
      "|  770 | IBS017      | HG01623         |\n",
      "|  771 | IBS017      | HG01624         |\n",
      "|  772 | IBS016      | HG01625         |\n",
      "|  773 | IBS016      | HG01626         |\n",
      "|  775 | IBS015      | HG01628         |\n",
      "|  777 | IBS015      | HG01630         |\n",
      "|  778 | IBS014      | HG01631         |\n",
      "|  779 | IBS014      | HG01632         |\n",
      "|  782 | IBS025      | HG01668         |\n",
      "|  783 | IBS025      | HG01669         |\n",
      "|  784 | IBS026      | HG01670         |\n",
      "|  786 | IBS026      | HG01672         |\n",
      "|  787 | IBS027      | HG01673         |\n",
      "|  789 | IBS027      | HG01675         |\n",
      "|  790 | IBS028      | HG01676         |\n",
      "|  792 | IBS028      | HG01678         |\n",
      "|  793 | IBS029      | HG01679         |\n",
      "|  794 | IBS029      | HG01680         |\n",
      "|  796 | IBS030      | HG01682         |\n",
      "|  798 | IBS030      | HG01684         |\n",
      "|  799 | IBS031      | HG01685         |\n",
      "|  800 | IBS031      | HG01686         |\n",
      "|  802 | IBS034      | HG01694         |\n",
      "|  803 | IBS034      | HG01695         |\n",
      "|  805 | IBS035      | HG01697         |\n",
      "|  807 | IBS035      | HG01699         |\n",
      "|  808 | IBS036      | HG01700         |\n",
      "|  810 | IBS036      | HG01702         |\n",
      "|  812 | IBS037      | HG01704         |\n",
      "|  813 | IBS037      | HG01705         |\n",
      "|  815 | IBS038      | HG01707         |\n",
      "|  816 | IBS038      | HG01708         |\n",
      "|  817 | IBS039      | HG01709         |\n",
      "|  818 | IBS039      | HG01710         |\n",
      "|  820 | IBS041      | HG01746         |\n",
      "|  821 | IBS041      | HG01747         |\n",
      "|  824 | IBS044      | HG01756         |\n",
      "|  825 | IBS044      | HG01757         |\n",
      "|  826 | IBS046      | HG01761         |\n",
      "|  827 | IBS046      | HG01762         |\n",
      "|  830 | IBS047      | HG01765         |\n",
      "|  831 | IBS047      | HG01766         |\n",
      "|  832 | IBS048      | HG01767         |\n",
      "|  833 | IBS048      | HG01768         |\n",
      "|  835 | IBS049      | HG01770         |\n",
      "|  836 | IBS049      | HG01771         |\n",
      "|  838 | IBS050      | HG01773         |\n",
      "|  840 | IBS050      | HG01775         |\n",
      "|  841 | IBS051      | HG01776         |\n",
      "|  842 | IBS051      | HG01777         |\n",
      "|  844 | IBS052      | HG01779         |\n",
      "|  846 | IBS052      | HG01781         |\n",
      "|  848 | IBS053      | HG01783         |\n",
      "|  849 | IBS053      | HG01784         |\n",
      "|  850 | IBS054      | HG01785         |\n",
      "|  851 | IBS054      | HG01786         |\n",
      "|  853 | HG01789     | HG01789         |\n",
      "|  854 | HG01790     | HG01790         |\n",
      "|  855 | HG01791     | HG01791         |\n",
      "| 1143 | HG02215     | HG02215         |\n",
      "| 1146 | IBS055      | HG02219         |\n",
      "| 1147 | IBS056      | HG02220         |\n",
      "| 1148 | IBS056      | HG02221         |\n",
      "| 1150 | IBS057      | HG02223         |\n",
      "| 1151 | IBS057      | HG02224         |\n",
      "| 1154 | IBS059      | HG02230         |\n",
      "| 1155 | IBS059      | HG02231         |\n",
      "| 1156 | IBS060      | HG02232         |\n",
      "| 1157 | IBS060      | HG02233         |\n",
      "| 1159 | IBS061      | HG02235         |\n",
      "| 1160 | IBS061      | HG02236         |\n",
      "| 1162 | IBS062      | HG02238         |\n",
      "| 1163 | IBS062      | HG02239         |\n",
      "| 2358 | 1328        | NA06984         |\n",
      "| 2359 | 1341        | NA06985         |\n",
      "| 2360 | 13291       | NA06986         |\n",
      "| 2361 | 1328        | NA06989         |\n",
      "| 2364 | 1340        | NA06994         |\n",
      "| 2367 | 1340        | NA07000         |\n",
      "| 2374 | 13291       | NA07037         |\n",
      "| 2376 | 1341        | NA07048         |\n",
      "| 2377 | 13292       | NA07051         |\n",
      "| 2379 | 1340        | NA07056         |\n",
      "| 2383 | 1345        | NA07347         |\n",
      "| 2386 | 1345        | NA07357         |\n",
      "| 2400 | 1334        | NA10847         |\n",
      "| 2402 | 1344        | NA10851         |\n",
      "| 2415 | 1350        | NA11829         |\n",
      "| 2416 | 1350        | NA11830         |\n",
      "| 2417 | 1350        | NA11831         |\n",
      "| 2418 | 1350        | NA11832         |\n",
      "| 2420 | 1349        | NA11840         |\n",
      "| 2421 | 1349        | NA11843         |\n",
      "| 2422 | 1347        | NA11881         |\n",
      "| 2425 | 1377        | NA11892         |\n",
      "| 2426 | 1377        | NA11893         |\n",
      "| 2427 | 1377        | NA11894         |\n",
      "| 2429 | 1423        | NA11918         |\n",
      "| 2430 | 1423        | NA11919         |\n",
      "| 2431 | 1423        | NA11920         |\n",
      "| 2432 | 1424        | NA11930         |\n",
      "| 2433 | 1424        | NA11931         |\n",
      "| 2434 | 1424a       | NA11932         |\n",
      "| 2435 | 1424a       | NA11933         |\n",
      "| 2436 | 1362        | NA11992         |\n",
      "| 2438 | 1362        | NA11994         |\n",
      "| 2439 | 1362        | NA11995         |\n",
      "| 2440 | 1420        | NA12003         |\n",
      "| 2441 | 1420        | NA12004         |\n",
      "| 2442 | 1420        | NA12005         |\n",
      "| 2443 | 1420        | NA12006         |\n",
      "| 2444 | 1346        | NA12043         |\n",
      "| 2445 | 1346        | NA12044         |\n",
      "| 2446 | 1346        | NA12045         |\n",
      "| 2447 | 1346        | NA12046         |\n",
      "| 2450 | 1344        | NA12058         |\n",
      "| 2451 | 1334        | NA12144         |\n",
      "| 2454 | 1408        | NA12154         |\n",
      "| 2455 | 1408        | NA12155         |\n",
      "| 2456 | 1408        | NA12156         |\n",
      "| 2457 | 1375        | NA12234         |\n",
      "| 2461 | 1416        | NA12249         |\n",
      "| 2463 | 1418        | NA12272         |\n",
      "| 2464 | 1418        | NA12273         |\n",
      "| 2466 | 1418        | NA12275         |\n",
      "| 2467 | 1421a       | NA12282         |\n",
      "| 2468 | 1421a       | NA12283         |\n",
      "| 2469 | 1421        | NA12286         |\n",
      "| 2470 | 1421        | NA12287         |\n",
      "| 2474 | 1330        | NA12340         |\n",
      "| 2475 | 1330        | NA12341         |\n",
      "| 2476 | 1330        | NA12342         |\n",
      "| 2479 | 13281       | NA12347         |\n",
      "| 2480 | 13281       | NA12348         |\n",
      "| 2483 | 1353        | NA12383         |\n",
      "| 2485 | 1354        | NA12399         |\n",
      "| 2486 | 1354        | NA12400         |\n",
      "| 2487 | 1355        | NA12413         |\n",
      "| 2488 | 1355        | NA12414         |\n",
      "| 2490 | 1353        | NA12489         |\n",
      "| 2491 | 1353        | NA12546         |\n",
      "| 2494 | 1358        | NA12716         |\n",
      "| 2495 | 1358        | NA12717         |\n",
      "| 2496 | 1358        | NA12718         |\n",
      "| 2499 | 1444        | NA12748         |\n",
      "| 2500 | 1444        | NA12749         |\n",
      "| 2501 | 1444        | NA12750         |\n",
      "| 2502 | 1444        | NA12751         |\n",
      "| 2505 | 1447        | NA12760         |\n",
      "| 2506 | 1447        | NA12761         |\n",
      "| 2507 | 1447        | NA12762         |\n",
      "| 2508 | 1447        | NA12763         |\n",
      "| 2511 | 1451        | NA12775         |\n",
      "| 2512 | 1451        | NA12776         |\n",
      "| 2513 | 1451        | NA12777         |\n",
      "| 2514 | 1451        | NA12778         |\n",
      "| 2517 | 1454        | NA12812         |\n",
      "| 2518 | 1454        | NA12813         |\n",
      "| 2519 | 1454        | NA12814         |\n",
      "| 2520 | 1454        | NA12815         |\n",
      "| 2523 | 1456        | NA12827         |\n",
      "| 2524 | 1456        | NA12828         |\n",
      "| 2525 | 1456        | NA12829         |\n",
      "| 2526 | 1456        | NA12830         |\n",
      "| 2528 | 1458        | NA12842         |\n",
      "| 2529 | 1458        | NA12843         |\n",
      "| 2532 | 1459        | NA12872         |\n",
      "| 2533 | 1459        | NA12873         |\n",
      "| 2534 | 1459        | NA12874         |\n",
      "| 2537 | 1463        | NA12878         |\n",
      "| 2538 | 1463        | NA12889         |\n",
      "| 2539 | 1463        | NA12890         |\n",
      "| 3466 | NA20502     | NA20502         |\n",
      "| 3467 | NA20503     | NA20503         |\n",
      "| 3468 | NA20504     | NA20504         |\n",
      "| 3469 | NA20505     | NA20505         |\n",
      "| 3470 | NA20506     | NA20506         |\n",
      "| 3471 | NA20507     | NA20507         |\n",
      "| 3472 | NA20508     | NA20508         |\n",
      "| 3473 | NA20509     | NA20509         |\n",
      "| 3474 | NA20510     | NA20510         |\n",
      "| 3475 | NA20511     | NA20511         |\n",
      "| 3476 | NA20512     | NA20512         |\n",
      "| 3477 | NA20513     | NA20513         |\n",
      "| 3478 | NA20514     | NA20514         |\n",
      "| 3479 | NA20515     | NA20515         |\n",
      "| 3480 | NA20516     | NA20516         |\n",
      "| 3481 | NA20517     | NA20517         |\n",
      "| 3482 | NA20518     | NA20518         |\n",
      "| 3483 | NA20519     | NA20519         |\n",
      "| 3484 | NA20520     | NA20520         |\n",
      "| 3485 | NA20521     | NA20521         |\n",
      "| 3486 | NA20522     | NA20522         |\n",
      "| 3487 | NA20524     | NA20524         |\n",
      "| 3488 | NA20525     | NA20525         |\n",
      "| 3490 | NA20527     | NA20527         |\n",
      "| 3491 | NA20528     | NA20528         |\n",
      "| 3492 | NA20529     | NA20529         |\n",
      "| 3493 | NA20530     | NA20530         |\n",
      "| 3494 | NA20531     | NA20531         |\n",
      "| 3495 | NA20532     | NA20532         |\n",
      "| 3496 | NA20533     | NA20533         |\n",
      "| 3497 | NA20534     | NA20534         |\n",
      "| 3498 | NA20535     | NA20535         |\n",
      "| 3499 | NA20536     | NA20536         |\n",
      "| 3501 | NA20538     | NA20538         |\n",
      "| 3502 | NA20539     | NA20539         |\n",
      "| 3503 | NA20540     | NA20540         |\n",
      "| 3504 | NA20541     | NA20541         |\n",
      "| 3505 | NA20542     | NA20542         |\n",
      "| 3506 | NA20543     | NA20543         |\n",
      "| 3507 | NA20544     | NA20544         |\n",
      "| 3508 | NA20581     | NA20581         |\n",
      "| 3509 | NA20582     | NA20582         |\n",
      "| 3510 | NA20585     | NA20585         |\n",
      "| 3511 | NA20586     | NA20586         |\n",
      "| 3512 | NA20587     | NA20587         |\n",
      "| 3513 | NA20588     | NA20588         |\n",
      "| 3514 | NA20589     | NA20589         |\n",
      "| 3515 | NA20752     | NA20752         |\n",
      "| 3516 | NA20753     | NA20753         |\n",
      "| 3517 | NA20754     | NA20754         |\n",
      "| 3518 | NA20755     | NA20755         |\n",
      "| 3519 | NA20756     | NA20756         |\n",
      "| 3520 | NA20757     | NA20757         |\n",
      "| 3521 | NA20758     | NA20758         |\n",
      "| 3522 | NA20759     | NA20759         |\n",
      "| 3523 | NA20760     | NA20760         |\n",
      "| 3524 | NA20761     | NA20761         |\n",
      "| 3525 | NA20762     | NA20762         |\n",
      "| 3526 | NA20763     | NA20763         |\n",
      "| 3527 | NA20764     | NA20764         |\n",
      "| 3528 | NA20765     | NA20765         |\n",
      "| 3529 | NA20766     | NA20766         |\n",
      "| 3530 | NA20767     | NA20767         |\n",
      "| 3531 | NA20768     | NA20768         |\n",
      "| 3532 | NA20769     | NA20769         |\n",
      "| 3533 | NA20770     | NA20770         |\n",
      "| 3534 | NA20771     | NA20771         |\n",
      "| 3535 | NA20772     | NA20772         |\n",
      "| 3536 | NA20773     | NA20773         |\n",
      "| 3537 | NA20774     | NA20774         |\n",
      "| 3538 | NA20775     | NA20775         |\n",
      "| 3539 | NA20778     | NA20778         |\n",
      "| 3540 | NA20783     | NA20783         |\n",
      "| 3541 | NA20785     | NA20785         |\n",
      "| 3542 | NA20786     | NA20786         |\n",
      "| 3543 | NA20787     | NA20787         |\n",
      "| 3544 | NA20790     | NA20790         |\n",
      "| 3545 | NA20792     | NA20792         |\n",
      "| 3546 | NA20795     | NA20795         |\n",
      "| 3547 | NA20796     | NA20796         |\n",
      "| 3548 | NA20797     | NA20797         |\n",
      "| 3549 | NA20798     | NA20798         |\n",
      "| 3550 | NA20799     | NA20799         |\n",
      "| 3551 | NA20800     | NA20800         |\n",
      "| 3552 | NA20801     | NA20801         |\n",
      "| 3553 | NA20802     | NA20802         |\n",
      "| 3554 | NA20803     | NA20803         |\n",
      "| 3555 | NA20804     | NA20804         |\n",
      "| 3556 | NA20805     | NA20805         |\n",
      "| 3557 | NA20806     | NA20806         |\n",
      "| 3558 | NA20807     | NA20807         |\n",
      "| 3559 | NA20808     | NA20808         |\n",
      "| 3560 | NA20809     | NA20809         |\n",
      "| 3561 | NA20810     | NA20810         |\n",
      "| 3562 | NA20811     | NA20811         |\n",
      "| 3563 | NA20812     | NA20812         |\n",
      "| 3564 | NA20813     | NA20813         |\n",
      "| 3565 | NA20814     | NA20814         |\n",
      "| 3566 | NA20815     | NA20815         |\n",
      "| 3568 | NA20818     | NA20818         |\n",
      "| 3569 | NA20819     | NA20819         |\n",
      "| 3570 | NA20821     | NA20821         |\n",
      "| 3571 | NA20822     | NA20822         |\n",
      "| 3572 | NA20826     | NA20826         |\n",
      "| 3573 | NA20827     | NA20827         |\n",
      "| 3574 | NA20828     | NA20828         |\n",
      "| 3577 | NA20832     | NA20832         |\n"
     ]
    }
   ],
   "source": [
    "#download file containing metadata and filter for EUR super population \n",
    "int_call_samples_meta = pd.read_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/integrated_call_samples_v3.20130502.ALL.panel.txt\",sep=\"\\t\")\n",
    "european_individuals = int_call_samples_meta.loc[int_call_samples_meta.super_pop == \"EUR\"]\n",
    "european_individuals\n",
    "\n",
    "#download file containing Family ID and Individual ID for all individuals, filter for individual ID in european_individuals metadata \n",
    "int_call_samples_individuals = pd.read_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/integrated_call_samples_v3.20200731.ALL.ped.txt\",sep=\"\\t\")\n",
    "european_individuals_ids = int_call_samples_individuals.iloc[np.isin(int_call_samples_individuals['Individual ID'].values, european_individuals['sample'].values)]\n",
    "eur_individuals_ids = european_individuals_ids[[\"Family ID\",\"Individual ID\"]]\n",
    "print(eur_individuals_ids.to_markdown())\n",
    "\n",
    "#write file containing family and individual IDs to .txt file to be used in the filtering step next\n",
    "eur_individuals_ids.to_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/eur_individuals.txt\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96fe2ca6-dacd-4ce8-a846-40c1aad87ef7",
   "metadata": {},
   "source": [
    "##### 1.2 Run this command to get the plink triplet files from the 1000 genomes data for chromosome 22 filtered for european individuals\n",
    "\n",
    "plink --vcf /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz --keep /Users/tanvibansal/Documents/GitHub/Capstone/eur_individuals.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/22_1000g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2372b2ec-d418-42fd-ab4d-84224cd49a49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050075</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050115</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050213</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050527</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103542</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51241342</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103543</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51241386</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103544</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244163</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103545</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244205</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103546</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244237</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1103547 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR SNP  CM       POS A1 A2\n",
       "0         22   .   0  16050075  G  A\n",
       "1         22   .   0  16050115  A  G\n",
       "2         22   .   0  16050213  T  C\n",
       "3         22   .   0  16050319  T  C\n",
       "4         22   .   0  16050527  A  C\n",
       "...      ...  ..  ..       ... .. ..\n",
       "1103542   22   .   0  51241342  A  C\n",
       "1103543   22   .   0  51241386  G  C\n",
       "1103544   22   .   0  51244163  G  A\n",
       "1103545   22   .   0  51244205  T  C\n",
       "1103546   22   .   0  51244237  T  C\n",
       "\n",
       "[1103547 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read and display our generated .bim file \n",
    "bim = pd.read_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur/22_1000g.bim\",sep=\"\\t\",header=None,names=[\"CHR\",\"SNP\",\"CM\",\"POS\",\"A1\",\"A2\"])\n",
    "bim = bim.astype({\"POS\":\"int32\"})\n",
    "bim"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ad9f246-4cf4-4bc1-bd84-31a92e45a2e8",
   "metadata": {},
   "source": [
    "#### 2. SNP match 1000 genomes to GWAS and assign SNP IDs\n",
    "\n",
    "The 1000 genomes data we generated the .bim file does not include SNP ID's, so we need to create a process here to add them to the .bim file so we can filter by ID's in the HapMap3 list. But, sometimes our generated .bim file has multiple records for mutations at a given base pair location/chromosome to keep track of multiple versions of a mutation at that location (see cell below). The GWAS correlations are computed only for specific versions of these mutations (i.e. Allele 1 = X -> Allele 1 = Y), so we'll need to filter the SNP's in the .bim file to match the SNP's in the GWAS df. \n",
    "\n",
    "The tricky part is that A1 and A2 may be swapped as different sources prefer different standards with minor allele assignment, so we'll need to find cases where SNP's in the .bim file match the chromosome and position of SNP's in the GWAS file, then further downselect to cases where A1 = A1 and A2 = A2 OR  A1 = A2 and A2 = A1. Then we can name the SNPs in accordance with the GWAS file, generate random ID's for SNP's without a matching GWAS ID, and output a valid .bim file format to regenerate the plink triplet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "eac08f9d-2234-4262-a909-c2a05b5f5b7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050075</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050115</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050213</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16050527</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103873</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51241342</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103874</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51241386</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103875</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244163</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103876</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244205</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103877</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51244237</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1103878 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR SNP  CM       POS A1 A2  ... beta  se  tstat pval  ref  alt\n",
       "0         22   .   0  16050075  G  A  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "1         22   .   0  16050115  A  G  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "2         22   .   0  16050213  T  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "3         22   .   0  16050319  T  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "4         22   .   0  16050527  A  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "...      ...  ..  ..       ... .. ..  ...  ...  ..    ...  ...  ...  ...\n",
       "1103873   22   .   0  51241342  A  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "1103874   22   .   0  51241386  G  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "1103875   22   .   0  51244163  G  A  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "1103876   22   .   0  51244205  T  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "1103877   22   .   0  51244237  T  C  ...  NaN NaN    NaN  NaN  NaN  NaN\n",
       "\n",
       "[1103878 rows x 19 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#download modified GWAS file with HM's modifications to naming convention \n",
    "#gwas_imputed = pd.read_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/HEIGHT_irnt.gwas.imputed_v3.both_sexes.tsv\",sep=\"\\t\")\n",
    "gwas_df = gwas_df.rename(columns={\"chr\":\"CHR\",\"pos\":\"POS\"})\n",
    "\n",
    "#filter for chromosome of interest and cast position column to integer\n",
    "gwas_22 = gwas_df.loc[(gwas_df.CHR ==22)]\n",
    "gwas_22 = gwas_22.astype({\"POS\":\"int32\"})\n",
    "\n",
    "#now join the 1000g generated bim file to the modified gwas file\n",
    "bim_gwas_df = bim.merge(gwas_22, on = [\"CHR\",\"POS\"],suffixes=(\".bim\",\".gwas\"),how=\"left\")\n",
    "bim_gwas_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3fde47f8-0d12-4340-8182-fd22221db3da",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/z0/1hhlpnfn2mz3vk7t57zdcy0w0000gn/T/ipykernel_46541/3492587372.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bim_gwas_df_no_null[\"allele_match\"] = bim_gwas_df_no_null.apply(lambda x: allele_match_finder(x),axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>allele_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5419</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16464274</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>22:16464274:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.084046</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>60578.3</td>\n",
       "      <td>-262.5120</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.003183</td>\n",
       "      <td>1.072040</td>\n",
       "      <td>0.283702</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5834</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16488635</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>22:16488635:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.080987</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>58373.6</td>\n",
       "      <td>-220.8840</td>\n",
       "      <td>0.003806</td>\n",
       "      <td>0.003111</td>\n",
       "      <td>1.223410</td>\n",
       "      <td>0.221175</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5837</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16488702</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>22:16488702:G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.080078</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>57718.1</td>\n",
       "      <td>-250.3560</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>0.003121</td>\n",
       "      <td>1.197910</td>\n",
       "      <td>0.230954</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6013</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16495833</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>22:16495833:C:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078102</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>56294.0</td>\n",
       "      <td>-249.4170</td>\n",
       "      <td>0.004095</td>\n",
       "      <td>0.003069</td>\n",
       "      <td>1.334020</td>\n",
       "      <td>0.182199</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6069</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16498458</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>22:16498458:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>53654.4</td>\n",
       "      <td>-203.4670</td>\n",
       "      <td>0.004117</td>\n",
       "      <td>0.003224</td>\n",
       "      <td>1.277110</td>\n",
       "      <td>0.201566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103673</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51229717</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>22:51229717:A:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.296120</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>213436.0</td>\n",
       "      <td>-437.6630</td>\n",
       "      <td>-0.002442</td>\n",
       "      <td>0.002004</td>\n",
       "      <td>-1.218610</td>\n",
       "      <td>0.222992</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103679</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51229805</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>22:51229805:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.072999</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>52615.7</td>\n",
       "      <td>-172.5280</td>\n",
       "      <td>0.001794</td>\n",
       "      <td>0.003184</td>\n",
       "      <td>0.563533</td>\n",
       "      <td>0.573073</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103716</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51231220</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>22:51231220:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.053676</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>38688.2</td>\n",
       "      <td>-20.4974</td>\n",
       "      <td>0.002084</td>\n",
       "      <td>0.003916</td>\n",
       "      <td>0.532177</td>\n",
       "      <td>0.594604</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103804</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51237063</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.299014</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>215522.0</td>\n",
       "      <td>-260.2040</td>\n",
       "      <td>-0.001741</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>-0.895286</td>\n",
       "      <td>0.370635</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103831</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51237712</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>22:51237712:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.055150</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>39750.5</td>\n",
       "      <td>-23.8471</td>\n",
       "      <td>0.001338</td>\n",
       "      <td>0.003885</td>\n",
       "      <td>0.344285</td>\n",
       "      <td>0.730632</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174549 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR SNP  CM       POS A1  ...     tstat      pval ref  alt allele_match\n",
       "5419      22   .   0  16464274  C  ...  1.072040  0.283702   A    C            1\n",
       "5834      22   .   0  16488635  A  ...  1.223410  0.221175   C    A            1\n",
       "5837      22   .   0  16488702  C  ...  1.197910  0.230954   G    C            1\n",
       "6013      22   .   0  16495833  A  ...  1.334020  0.182199   C    A            1\n",
       "6069      22   .   0  16498458  A  ...  1.277110  0.201566   G    A            1\n",
       "...      ...  ..  ..       ... ..  ...       ...       ...  ..  ...          ...\n",
       "1103673   22   .   0  51229717  T  ... -1.218610  0.222992   A    T            1\n",
       "1103679   22   .   0  51229805  C  ...  0.563533  0.573073   T    C            1\n",
       "1103716   22   .   0  51231220  G  ...  0.532177  0.594604   A    G            1\n",
       "1103804   22   .   0  51237063  C  ... -0.895286  0.370635   T    C            1\n",
       "1103831   22   .   0  51237712  A  ...  0.344285  0.730632   G    A            1\n",
       "\n",
       "[174549 rows x 20 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#start with the subset of our .bim dataframe that has matching GWAS data populated. The rest of the df will get random ID's generated, we'll come back to that later\n",
    "bim_gwas_df_no_null = bim_gwas_df.loc[~bim_gwas_df.ref.isnull()]\n",
    "\n",
    "#get positions of SNPs that are duplicated in the non-null df. These SNP's will need to be filtered based on true allele matches in GWAS df\n",
    "duplicate_snp_pos = bim_gwas_df_no_null.loc[bim_gwas_df_no_null.POS.duplicated(),\"POS\"].values\n",
    "def duplicate_variant_finder(df):\n",
    "    duplicate_variants = collections.Counter(bim_gwas_df_allele_match.variant.values) \n",
    "    dvs = [k  for k in duplicate_variants.keys() if duplicate_variants[k] > 1]\n",
    "    return(dvs)\n",
    "    \n",
    "bim_gwas_df_duplicate = bim_gwas_df_no_null.iloc[np.isin(bim_gwas_df_no_null.POS.values,duplicate_snp_pos)].sort_values(by = \"POS\")\n",
    "\n",
    "#define function to return 1 if the alleles in .bim are the same as in gwas (regardless of order), 0 if not\n",
    "import collections \n",
    "def allele_match_finder(x):\n",
    "    bim_alleles = x[['A1','A2']]\n",
    "    gwas_alleles = x[['ref','alt']]\n",
    "    same = collections.Counter(bim_alleles) == collections.Counter(gwas_alleles)\n",
    "    if same:\n",
    "        out = 1\n",
    "    else:\n",
    "        out = 0\n",
    "    return out\n",
    "bim_gwas_df_duplicate[\"allele_match\"] = bim_gwas_df_duplicate.apply(lambda x: allele_match_finder(x),axis=1) \n",
    "\n",
    "#that works, lets just run allele_match_finder over the whole non-null dataset for simplicity and ignore the rows with allele_match = 0. \n",
    "#we'll need to generate random IDs for these later\n",
    "bim_gwas_df_no_null[\"allele_match\"] = bim_gwas_df_no_null.apply(lambda x: allele_match_finder(x),axis=1) \n",
    "bim_gwas_df_no_null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8fac145b-e656-4e7b-b98a-4e3fa5844629",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>allele_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19710</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>17102425</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>22:17102425:CCT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.288595</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>208013.00</td>\n",
       "      <td>-323.8390</td>\n",
       "      <td>0.000658</td>\n",
       "      <td>0.001835</td>\n",
       "      <td>0.358479</td>\n",
       "      <td>0.719985</td>\n",
       "      <td>CCT</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19711</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>17102425</td>\n",
       "      <td>C</td>\n",
       "      <td>CCT</td>\n",
       "      <td>22:17102425:CCT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.288595</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>208013.00</td>\n",
       "      <td>-323.8390</td>\n",
       "      <td>0.000658</td>\n",
       "      <td>0.001835</td>\n",
       "      <td>0.358479</td>\n",
       "      <td>0.719985</td>\n",
       "      <td>CCT</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20885</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>17136274</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>22:17136274:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>3384.72</td>\n",
       "      <td>-40.6283</td>\n",
       "      <td>0.001257</td>\n",
       "      <td>0.013476</td>\n",
       "      <td>0.093280</td>\n",
       "      <td>0.925681</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20886</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>17136274</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>22:17136274:T:TA</td>\n",
       "      <td>TA</td>\n",
       "      <td>0.004837</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>3486.29</td>\n",
       "      <td>-42.7700</td>\n",
       "      <td>0.001997</td>\n",
       "      <td>0.012902</td>\n",
       "      <td>0.154784</td>\n",
       "      <td>0.876991</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21099</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>17141436</td>\n",
       "      <td>CCGAACG</td>\n",
       "      <td>C</td>\n",
       "      <td>22:17141436:C:CCGAA</td>\n",
       "      <td>CCGAA</td>\n",
       "      <td>0.294499</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>212268.00</td>\n",
       "      <td>-287.7320</td>\n",
       "      <td>0.000939</td>\n",
       "      <td>0.001826</td>\n",
       "      <td>0.514520</td>\n",
       "      <td>0.606889</td>\n",
       "      <td>C</td>\n",
       "      <td>CCGAA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097697</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51041570</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>22:51041570:C:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.046123</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>33244.50</td>\n",
       "      <td>-121.2010</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.003951</td>\n",
       "      <td>0.300844</td>\n",
       "      <td>0.763534</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099169</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51085928</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>22:51085928:AAAAAGAAAAG:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.100454</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>72404.80</td>\n",
       "      <td>-180.4010</td>\n",
       "      <td>-0.001366</td>\n",
       "      <td>0.002859</td>\n",
       "      <td>-0.477722</td>\n",
       "      <td>0.632848</td>\n",
       "      <td>AAAAAGAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099170</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51085928</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAGAAAAG</td>\n",
       "      <td>22:51085928:A:AAAAAG</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>0.015195</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>10952.40</td>\n",
       "      <td>-64.5595</td>\n",
       "      <td>-0.009348</td>\n",
       "      <td>0.006906</td>\n",
       "      <td>-1.353550</td>\n",
       "      <td>0.175882</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099168</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51085928</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>22:51085928:A:AAAAAG</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>0.015195</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>10952.40</td>\n",
       "      <td>-64.5595</td>\n",
       "      <td>-0.009348</td>\n",
       "      <td>0.006906</td>\n",
       "      <td>-1.353550</td>\n",
       "      <td>0.175882</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099171</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>51085928</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAGAAAAG</td>\n",
       "      <td>22:51085928:AAAAAGAAAAG:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.100454</td>\n",
       "      <td>False</td>\n",
       "      <td>360388.0</td>\n",
       "      <td>72404.80</td>\n",
       "      <td>-180.4010</td>\n",
       "      <td>-0.001366</td>\n",
       "      <td>0.002859</td>\n",
       "      <td>-0.477722</td>\n",
       "      <td>0.632848</td>\n",
       "      <td>AAAAAGAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1498 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR SNP  CM       POS  ...      pval          ref     alt allele_match\n",
       "19710     22   .   0  17102425  ...  0.719985          CCT       C            0\n",
       "19711     22   .   0  17102425  ...  0.719985          CCT       C            1\n",
       "20885     22   .   0  17136274  ...  0.925681            T       A            1\n",
       "20886     22   .   0  17136274  ...  0.876991            T      TA            0\n",
       "21099     22   .   0  17141436  ...  0.606889            C   CCGAA            0\n",
       "...      ...  ..  ..       ...  ...       ...          ...     ...          ...\n",
       "1097697   22   .   0  51041570  ...  0.763534            C       G            1\n",
       "1099169   22   .   0  51085928  ...  0.632848  AAAAAGAAAAG       A            0\n",
       "1099170   22   .   0  51085928  ...  0.175882            A  AAAAAG            0\n",
       "1099168   22   .   0  51085928  ...  0.175882            A  AAAAAG            1\n",
       "1099171   22   .   0  51085928  ...  0.632848  AAAAAGAAAAG       A            1\n",
       "\n",
       "[1498 rows x 20 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_gwas_df_duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4d8a4d5f-fa50-4a8c-81fa-80b9da03448e",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'bim_gwas_df_no_null_drop_dupl' is not defined",
     "output_type": "error",
     "traceback": [
      "Traceback \u001b[0;36m(most recent call last)\u001b[0m:\n",
      "\u001b[0;36m  Cell \u001b[0;32mIn[31], line 6\u001b[0;36m\n\u001b[0;31m    duplicate_variants_df = bim_gwas_df_no_null_drop_dupl.iloc[np.isin(bim_gwas_df_no_null_drop_dupl.variant,dvs)]\u001b[0;36m\n",
      "\u001b[0;31mNameError\u001b[0m\u001b[0;31m:\u001b[0m name 'bim_gwas_df_no_null_drop_dupl' is not defined\n"
     ]
    }
   ],
   "source": [
    "bim_gwas_df_allele_match = bim_gwas_df_no_null.loc[bim_gwas_df_no_null.allele_match ==1] \n",
    "\n",
    "#now lets check for duplicate variant IDs in our remaining set as the plink extraction will fail if any SNP IDs are repeated\n",
    "duplicate_variants = collections.Counter(bim_gwas_df_allele_match.variant.values) \n",
    "dvs = [k  for k in duplicate_variants.keys() if duplicate_variants[k] > 1]\n",
    "duplicate_variants_df = bim_gwas_df_no_null_drop_dupl.iloc[np.isin(bim_gwas_df_no_null_drop_dupl.variant,dvs)]\n",
    "print(\"Variant ID's with duplicates remaining: %s\"%(len(dvs))) \n",
    "duplicate_variants_df.head(15)\n",
    "\n",
    "#we can see that variants with 2 instances are all replicated because A1 and A2 in the .bim file are present in both orders. \n",
    "#lets just take the cases where A2 == ref for now\n",
    "x = duplicate_variants_df.iloc[1]\n",
    "def allele_order_finder(x):\n",
    "    if x.A2 == x.ref:\n",
    "        out = 1\n",
    "    else:\n",
    "        out = 0\n",
    "    return out\n",
    "bim_gwas_df_no_null[\"allele_ordered\"] = bim_gwas_df_no_null.apply(lambda x: allele_order_finder(x),axis=1)\n",
    "bim_gwas_df_no_null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "86f74ded-bece-4e32-8c90-518f1488c9b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CM</th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5419</th>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>16464274</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHR SNP  CM       POS A1 A2\n",
       "5419   22   .   0  16464274  C  A"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim.iloc[np.isin(bim.POS.values, 16464274)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "90f92d6e-0ced-4dfa-8616-4ef391be85e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#now lets test again how many duplicate variant IDs we have\n",
    "#bim_gwas_df_allele_match_ordered = bim_gwas_df_no_null.loc[(bim_gwas_df_no_null.allele_match ==1) & (bim_gwas_df_no_null.allele_ordered == 1)] \n",
    "#duplicate_variants = collections.Counter(bim_gwas_df_allele_match_ordered.variant.values)\n",
    "#dvs = [k for k in duplicate_variants.keys() if duplicate_variants[k] > 1]\n",
    "#print(\"Variant ID's with duplicates remaining: %s\"%(len(dvs)))\n",
    "\n",
    "#no more duplicate variant IDs, now dump this all into a function to run over each row of the .bim file we can parallelize\n",
    "def SNP_id_finder(x, gwas_df):\n",
    "    x_gwas_df = gwas_df.loc[(gwas_df.CHR == x.CHR) & (gwas_df.POS == x.POS)]\n",
    "    n_matches = len(x_gwas_df)\n",
    "    alternate_id = \"%s:%s\"%(x.POS,x.name)\n",
    "    if n_matches == 0:\n",
    "        return alternate_id\n",
    "    else:\n",
    "        #allel match check\n",
    "        allele_match = [collections.Counter(x[['A1','A2']]) == collections.Counter(x_gwas_df.iloc[i][['ref','alt']]) for i in range(len(x_gwas_df))]\n",
    "        if sum(allele_match) == 0:\n",
    "            return alternate_id\n",
    "        else:\n",
    "            matches = x_gwas_df.loc[allele_match]\n",
    "            #if len(matches) == 1:\n",
    "             #   return matches.variant.values.item()\n",
    "            #else:\n",
    "            #allele order check\n",
    "            allele_ordered = [x.A2 == matches.iloc[i]['ref'] for i in range(len(matches))]\n",
    "            if sum(allele_ordered) == 0: # this should never happen but handling it here just in case\n",
    "                return alternate_id\n",
    "            else:\n",
    "                matches_remaining = matches.loc[allele_ordered]\n",
    "                if len(matches_remaining) == 1:\n",
    "                    return matches_remaining.variant.values.item() \n",
    "                else:\n",
    "                    return alternate_id # this should never happen but handling it here just in case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d098e1ba-4551-4b05-8dc3-055ccae29e06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 10 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n",
      "Duplicate SNP ID's: 0\n"
     ]
    }
   ],
   "source": [
    "#run function over each row of the bim dataset\n",
    "from pandarallel import pandarallel\n",
    "import gc\n",
    "pandarallel.initialize(progress_bar = False)\n",
    "bim_ids = bim.parallel_apply(lambda x: SNP_id_finder(x, gwas_22),axis=1)\n",
    "gc.collect()\n",
    "duplicated_bim_ids = collections.Counter(bim_ids)\n",
    "duplicate_ids = [k for k in duplicated_bim_ids.keys() if duplicated_bim_ids[k] > 1]\n",
    "print(\"Duplicate SNP ID's: %s\"%(len(duplicate_ids)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "7f07fc3f-4b8e-4528-9695-30ed097f860b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#no duplicate ids, we're good to replace the SNP ID column with our computed IDs and write the whole modified .bim file out to txt\n",
    "bim.loc[:,\"SNP\"]=bim_ids\n",
    "bim.to_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/modified.bim\",sep=\"\\t\",header=None,index=False)\n",
    "\n",
    "#find the common SNP ids between annot_df and .bim now and write out to .txt file to ensure we can run LDSC in the next stage\n",
    "matches = bim_ids.apply(lambda x: len(x.split(\":\"))).sort_values(ascending=False)\n",
    "matches = matches.loc[matches.values > 2]\n",
    "unique_snps = bim_ids.loc[matches.index].to_frame().sort_values(by=0).rename(columns={0:\"SNP\"}).merge(annot_df.SNP).values\n",
    "with open('/Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/filtered_snps.txt', 'w') as f:\n",
    "    for snp in unique_snps:\n",
    "        f.write(f\"{snp[0]}\\t\") \n",
    "\n",
    "#filter annotation df for exact order and content match for SNP IDs and write to disk\n",
    "annot_df_prep = annot_df.merge(bim_ids.loc[matches.index].to_frame(),left_on=[\"SNP\"],right_on=[0]).sort_values(by=\"SNP\").drop(columns=[0])\n",
    "annot_df_prep[\"CM\"] = 0.0 \n",
    "annot_df_prep = annot_df_prep[[\"CHR\",\"BP\",\"SNP\", \"CM\", \"s_het_w.f1\", \"s_het_w.f2\", \"s_het_w.f3\", \"s_het_w.f4\", \"s_het_w.f5\"]]\n",
    "annot_df_prep.to_csv(\"/Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/s_het_weighted_ch22.annot\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12f37f55-f69e-41ec-922c-b4cd2f82facb",
   "metadata": {},
   "source": [
    "##### Run this command to run the plink extraction\n",
    "plink --bfile /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/22_1000g_modified --extract /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/filtered_snps.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/22_1000g_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ab3ff6d-7467-44dd-94fc-d36d1ad1a354",
   "metadata": {},
   "source": [
    "plink --vcf /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz --keep /Users/tanvibansal/Documents/GitHub/Capstone/eur_individuals.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/22_1000g\n",
    "\n",
    "\n",
    "plink --bfile /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur --extract /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22/filtered_snps.txt --make-bed --out /Users/tanvibansal/Documents/GitHub/ldsc/1kg_eur/s_het_22"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f447270-1783-45e8-bdd6-9d3202f6ef22",
   "metadata": {},
   "source": [
    "##### Run this comment to run the ld score computation \n",
    "python ldsc.py --l2 --bfile /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/22_1000g_filt --ld-wind-cm 1 --annot /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/s_het_weighted_ch22.annot --out /Users/tanvibansal/Documents/GitHub/Capstone/generated_plink_files/chr22/1000g_eur_hm3/s_het_w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29aea24e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot histograms for each difference column\n",
    "def plot_histograms(df, columns):\n",
    "    plt.figure(figsize=(14, 10))\n",
    "    \n",
    "    for i, col in enumerate(columns):\n",
    "        plt.subplot(2, 3, i + 1)  # Create a subplot for each column\n",
    "        df[col].hist(bins=50, color='skyblue', edgecolor='black')\n",
    "        plt.title(f'Distribution of {col}')\n",
    "        plt.xlabel('Value')\n",
    "        plt.ylabel('Frequency')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "# Define the columns to plot\n",
    "columns_to_plot = ['diff_f1', 'diff_f2', 'diff_f3', 'diff_b1', 'diff_b2', 'diff_b3']\n",
    "# Call the function to plot histograms\n",
    "plot_histograms(f_b_final_with_differences, columns_to_plot)\n",
    "\n",
    "columns_to_plot = [\"f1_dist\",\"f2_dist\",\"f3_dist\",\"f4_dist\",\"f5_dist\"]\n",
    "plot_histograms(nearest_genes_sample_df, columns_to_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e4837c-ffa1-491f-989b-a6cd2611bd33",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_het_info = pd.read_excel('s_het_info.xlsx',sheet_name=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4f1fa143",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to perform the merge for all forward and backward GeneSymbols and rename the 'post_mean' columns\n",
    "def merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges=3):\n",
    "    # Rename 'ensg' column in s_het_info to match the GeneSymbol columns in f_b_final_with_differences\n",
    "    s_het_info = s_het_info.rename(columns={'ensg': 'GeneSymbol'})\n",
    "\n",
    "    # Iterate over both forward and backward gene symbols\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Forward merge for each GeneSymbol_f\n",
    "        f_b_final_with_differences = pd.merge(\n",
    "            f_b_final_with_differences, \n",
    "            s_het_info[['GeneSymbol', 'post_mean']], \n",
    "            left_on=f'GeneSymbol_f{i}', \n",
    "            right_on='GeneSymbol', \n",
    "            how='left',\n",
    "            suffixes=('', f'_f{i}')\n",
    "        )\n",
    "        \n",
    "        # Rename the post_mean column to indicate the forward direction\n",
    "        f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "            columns={'post_mean': f's_het_post_f{i}'}\n",
    "        )\n",
    "        \n",
    "        # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "        f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "        # Backward merge for each GeneSymbol_b\n",
    "        f_b_final_with_differences = pd.merge(\n",
    "            f_b_final_with_differences, \n",
    "            s_het_info[['GeneSymbol', 'post_mean']], \n",
    "            left_on=f'GeneSymbol_b{i}', \n",
    "            right_on='GeneSymbol', \n",
    "            how='left',\n",
    "            suffixes=('', f'_b{i}')\n",
    "        )\n",
    "        \n",
    "        # Rename the post_mean column to indicate the backward direction\n",
    "        f_b_final_with_differences = f_b_final_with_differences.rename(\n",
    "            columns={'post_mean': f's_het_post_b{i}'}\n",
    "        )\n",
    "        \n",
    "        # Drop the duplicated 'GeneSymbol' column created during the merge\n",
    "        f_b_final_with_differences = f_b_final_with_differences.drop(columns='GeneSymbol', errors='ignore')\n",
    "\n",
    "    return f_b_final_with_differences\n",
    "\n",
    "# Example usage:\n",
    "f_b_final_merged_s_het = merge_s_het_info_with_all(f_b_final_with_differences, s_het_info, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db752cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f7cc0427",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# f_b_final_merged_s_het[['diff_b1']] = abs(f_b_final_merged_s_het[['diff_b1']])\n",
    "# f_b_final_merged_s_het[['diff_b2']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# f_b_final_merged_s_het[['diff_b3']] = abs(f_b_final_merged_s_het[['diff_b3']])\n",
    "# Loop to apply the absolute value for all diff_b columns\n",
    "def apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3):\n",
    "    for i in range(1, num_merges + 1):\n",
    "        f_b_final_merged_s_het[f'diff_b{i}'] = f_b_final_merged_s_het[f'diff_b{i}'].abs()\n",
    "\n",
    "    return f_b_final_merged_s_het\n",
    "\n",
    "# Example usage:\n",
    "f_b_final_merged_s_het = apply_abs_diff_b(f_b_final_merged_s_het, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4c6da80",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_b_final_merged_s_het"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3566073f",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to keep only the distance and s_het measures from the dataframe\n",
    "def keep_dist_and_s_het_measures(df, num_merges=3):\n",
    "    # List to store column names for distance and s_het measures\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the diff and s_het_post column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'diff_f{i}')\n",
    "        columns_to_keep.append(f'diff_b{i}')\n",
    "        columns_to_keep.append(f's_het_post_f{i}')\n",
    "        columns_to_keep.append(f's_het_post_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "filtered_f_b_final = keep_dist_and_s_het_measures(f_b_final_merged_s_het, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9536b18f",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8efcecb5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to compute s_het weighted by 1/distance per gene-SNP pair\n",
    "def compute_weighted_s_het(df, num_merges=3):\n",
    "    # Loop to compute the weighted s_het for both forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        # Calculate inverse distance for forward and backward directions\n",
    "        df[f'inv_dist_f{i}'] = 1 / df[f'diff_f{i}']\n",
    "        df[f'inv_dist_b{i}'] = 1 / df[f'diff_b{i}']\n",
    "        \n",
    "        # Compute weighted s_het by multiplying s_het by inverse distance\n",
    "        df[f'weighted_s_het_f{i}'] = df[f'inv_dist_f{i}'] * df[f's_het_post_f{i}']\n",
    "        df[f'weighted_s_het_b{i}'] = df[f'inv_dist_b{i}'] * df[f's_het_post_b{i}']\n",
    "\n",
    "    return df\n",
    "\n",
    "# Example usage:\n",
    "weighted_f_b_final = compute_weighted_s_het(filtered_f_b_final, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b3df79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_f_b_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0b3796b6",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Function to keep only the weighted s_het values from the dataframe\n",
    "def keep_weighted_s_het(df, num_merges=3):\n",
    "    # List to store column names for weighted s_het values\n",
    "    columns_to_keep = ['variant', 'chr', 'pos']\n",
    "\n",
    "    # Loop to collect the weighted_s_het column names for forward and backward directions\n",
    "    for i in range(1, num_merges + 1):\n",
    "        columns_to_keep.append(f'weighted_s_het_f{i}')\n",
    "        columns_to_keep.append(f'weighted_s_het_b{i}')\n",
    "\n",
    "    # Keep only the relevant columns\n",
    "    df_filtered = df[columns_to_keep]\n",
    "\n",
    "    return df_filtered\n",
    "\n",
    "# Example usage:\n",
    "weighted_s_het_only = keep_weighted_s_het(weighted_f_b_final, num_merges=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f06a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_s_het_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381a5b0e-240d-4840-8232-2655389eb38c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#oliver's results \n",
    "weighted_s_het_only.iloc[np.argwhere(np.isin(weighted_s_het_only.variant, m[\"variant.mut\"].values)).flatten()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8f8edd-ed90-4661-bd3a-a067dec7af2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (Spyder)",
   "language": "python3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
